--- a/src/main/java/org/bukkit/inventory/InventoryView.java
+++ b/src/main/java/org/bukkit/inventory/InventoryView.java
@@ -323,6 +_,18 @@
     @Deprecated(since = "1.21.1") // Paper
     public void setTitle(@NotNull String title);
 
+    // Daydream start - Add packet title update to inventory
+    /**
+     * 클라이언트에게 패킷을 보내 새 타이틀로 인벤토리를 다시 엽니다.
+     * <p>
+     * 인벤토리의 ID와 내용은 동일하게 유지됩니다. 서버 측 인벤토리 타이틀에는 영향을 미치지 않습니다.
+     * {@link #title()}와 같은 메서드는 여전히 이전 타이틀을 반환합니다.
+     *
+     * @param title 새로운 타이틀
+     */
+    public void sendTitleUpdate(@NotNull net.kyori.adventure.text.Component title);
+    // Daydream end - Add packet title update to inventory
+
     /**
      * Gets the menu type of the inventory view if applicable.
      * <p>
