--- a/src/main/java/org/bukkit/GameRule.java
+++ b/src/main/java/org/bukkit/GameRule.java
@@ -229,6 +_,12 @@
      * Whether primed tnt explodes.
      */
     public static final GameRule<Boolean> TNT_EXPLODES = new GameRule<>("tntExplodes", Boolean.class);
+    // Daydream start - Add more GameRule
+    /**
+     * 불과 관련된 블록에 의해 플레이어가 피해를 입을지 여부
+     */
+    public static final GameRule<Boolean> BLOCK_FIRE_DAMAGE = new GameRule<>("blockFireDamage", Boolean.class);
+    // Daydream end - Add more GameRule
 
     // Numerical rules
     /**
