From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Thu, 5 Aug 2021 15:44:59 +0900
Subject: [PATCH] Rebrand to Daydream


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 8d8fe04e6b09d2a5b1cc05002073df5c58cdcb96..471eb9dfd6f4f6494415795e7d9420a15ded0aa0 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2121,6 +2121,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Daydream start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getDaydreamConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Daydream end
+
         /**
          * Sends the component to the player
          *
