From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Thu, 5 Aug 2021 15:44:59 +0900
Subject: [PATCH] Rebrand to Daydream


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index f1fa97d12f97baf97beb92ca0719cf3cf906b225..929859b9de0856e7686abbaa8ccdccd99341f0ba 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2190,6 +2190,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Daydream start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getDaydreamConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Daydream end
+
         /**
          * Sends the component to the player
          *
