From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Fri, 17 May 2024 23:20:56 +0900
Subject: [PATCH] Rebrand to Daydream


diff --git a/build.gradle.kts b/build.gradle.kts
index 8d76f0a8b6b67f0cae0b2fa23f0f79fc9fe97f30..ab7fa63820837ea6c72e66bfaf5a65c7590f968b 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -39,7 +39,7 @@ abstract class MockitoAgentProvider : CommandLineArgumentProvider {
 // Paper end - configure mockito agent that is needed in newer java versions
 
 dependencies {
-    api("com.mojang:brigadier:1.2.9") // Paper - Brigadier command api
+    api("com.mojang:brigadier:1.3.10") // Paper - Brigadier command api // Daydream - Rebrand to Daydream (Keep in sync with major MC versions)
     // api dependencies are listed transitively to API consumers
     api("com.google.guava:guava:33.3.1-jre")
     api("com.google.code.gson:gson:2.11.0")
diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 652ff54e7c50412503725d628bfe72ed03059790..cdf1a0aab7d8dbe18161b6731a3f1f68e03eccee 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -18,6 +18,12 @@ public interface ServerBuildInfo {
      * The brand id for Paper.
      */
     Key BRAND_PAPER_ID = Key.key("papermc", "paper");
+    // Daydream start - Rebrand to Daydream
+    /**
+     * The brand id for Daydream.
+     */
+    Key BRAND_DAYDREAM_ID = Key.key("caramel", "daydream");
+    // Daydream end - Rebrand to Daydream
 
     /**
      * Gets the {@code ServerBuildInfo}.
@@ -70,6 +76,15 @@ public interface ServerBuildInfo {
      */
     String minecraftVersionName();
 
+    // Daydream start - Rebrand to Daydream (Expose protocol version)
+    /**
+     * Gets the Minecraft protocol version.
+     *
+     * @return the Minecraft protocol version
+     */
+    default int protocolVersion() { throw new UnsupportedOperationException(); }
+    // Daydream end - Rebrand to Daydream (Expose protocol version)
+
     /**
      * Gets the build number.
      *
