From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Thu, 5 Aug 2021 15:44:59 +0900
Subject: [PATCH] Rebrand to Daydream


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1d1a1d087dabc9794e0062a064da2cced4062309..f1a252881c979c4d800a06806d464d0d804d8042 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1994,6 +1994,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Daydream start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getDaydreamConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Daydream end
+
         /**
          * Sends the component to the player
          *
