From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Thu, 5 Aug 2021 15:44:59 +0900
Subject: [PATCH] Add Daydream Config


diff --git a/src/main/java/co/aikar/timings/TimingsManager.java b/src/main/java/co/aikar/timings/TimingsManager.java
index e72ad05abada04426e32a73d02b21cb69079d268..f15fa2abcc04d7d1b6f49a77a023dbd04dbe0260 100644
--- a/src/main/java/co/aikar/timings/TimingsManager.java
+++ b/src/main/java/co/aikar/timings/TimingsManager.java
@@ -155,7 +155,8 @@ public final class TimingsManager {
         final Server server = Bukkit.getServer();
         if (!(  server == null || pluginName == null ||
                 "minecraft".equals(pluginName) || "bukkit".equals(pluginName) ||
-                "spigot".equalsIgnoreCase(pluginName) || "paper".equals(pluginName)
+                "spigot".equalsIgnoreCase(pluginName) || "paper".equals(pluginName) ||
+                "daydream".equals(pluginName) // Daydream - Add Daydream Config (Rebrand to Daydream)
         )) {
             plugin = server.getPluginManager().getPlugin(pluginName);
         }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index ad816538b30079c62d5e1eb98c6f4b61e12e8d47..6455b54397f8d50a1e859b8d23a3b5d65a9ddd62 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2283,6 +2283,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Daydream start - Add Daydream Config
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getDaydreamConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Daydream end - Add Daydream Config
+
         /**
          * Sends the component to the player
          *
diff --git a/src/main/java/org/bukkit/plugin/java/LibraryLoader.java b/src/main/java/org/bukkit/plugin/java/LibraryLoader.java
index c66252802c51174bc26f266cb5cdecdd856ff220..2a836bbf71601ef917e0d0c8829d3f1a99faf1bf 100644
--- a/src/main/java/org/bukkit/plugin/java/LibraryLoader.java
+++ b/src/main/java/org/bukkit/plugin/java/LibraryLoader.java
@@ -144,6 +144,7 @@ public class LibraryLoader
             }
 
             jarFiles.add( url );
+            if (!org.bukkit.Bukkit.spigot().getDaydreamConfig().getBoolean("logger.ignore-library-loader-log", false)) // Daydream - Add Daydream Config
             logger.log( Level.INFO, "[{0}] Loaded library {1}", new Object[]
             {
                 java.util.Objects.requireNonNullElseGet(desc.getPrefix(), desc::getName), file // Paper - use configured log prefix
