From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Thu, 5 Aug 2021 17:50:02 +0900
Subject: [PATCH] Add Daydream Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1db87711dd9b67bcba42e731c8c1a81502abf56c..628792c12aaa51310950b27e04293f14474af5ec 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1715,6 +1715,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Daydream start - add config to timings report
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getDaydreamConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Daydream end - add config to timings report
+
         /**
          * Sends the component to the player
          *
