From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Fri, 6 Aug 2021 17:47:12 +0900
Subject: [PATCH] Setting Daydream command


diff --git a/src/main/java/co/aikar/timings/TimingsManager.java b/src/main/java/co/aikar/timings/TimingsManager.java
index a92925d41110226f7fda055b71ce7be60eedd038..45c7400e3f2d34b9b16ecadcde376e7251bf81ac 100644
--- a/src/main/java/co/aikar/timings/TimingsManager.java
+++ b/src/main/java/co/aikar/timings/TimingsManager.java
@@ -150,8 +150,7 @@ public final class TimingsManager {
         Plugin plugin = null;
         final Server server = Bukkit.getServer();
         if (!(  server == null || pluginName == null ||
-                "minecraft".equals(pluginName) || "bukkit".equals(pluginName) ||
-                "spigot".equalsIgnoreCase(pluginName) || "paper".equals(pluginName)
+                "minecraft".equals(pluginName) || "daydream".equals(pluginName) // Daydream
         )) {
             plugin = server.getPluginManager().getPlugin(pluginName);
         }
diff --git a/src/main/java/org/bukkit/command/SimpleCommandMap.java b/src/main/java/org/bukkit/command/SimpleCommandMap.java
index 74252236b138969560e6513f24e7ecc6dc4a4127..9b4dd02a3eb8bb588e6b2cf448c842542f5d8bc8 100644
--- a/src/main/java/org/bukkit/command/SimpleCommandMap.java
+++ b/src/main/java/org/bukkit/command/SimpleCommandMap.java
@@ -35,6 +35,10 @@ public class SimpleCommandMap implements CommandMap {
     }
 
     private void setDefaultCommands() {
+        register("daydream", new VersionCommand("version")); // Daydream
+        register("daydream", new ReloadCommand("reload")); // Daydream
+        register("daydream", new co.aikar.timings.TimingsCommand("timings")); // Daydream
+        if (true) return; // Daydream - Disable bukkit prefix commands
         register("bukkit", new VersionCommand("version"));
         register("bukkit", new ReloadCommand("reload"));
         register("bukkit", new PluginsCommand("plugins"));
@@ -42,6 +46,7 @@ public class SimpleCommandMap implements CommandMap {
     }
 
     public void setFallbackCommands() {
+        if (true) return; // Daydream - Disable unused help command
         register("bukkit", new HelpCommand());
     }
 
@@ -257,6 +262,7 @@ public class SimpleCommandMap implements CommandMap {
     }
 
     public void registerServerAliases() {
+        if (true) return; // Daydream - Disable unused command aliases
         Map<String, String[]> values = server.getCommandAliases();
 
         for (Map.Entry<String, String[]> entry : values.entrySet()) {
diff --git a/src/main/java/org/bukkit/command/defaults/ReloadCommand.java b/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
index 0c7ba0718de2b93d013968ca0fec34ffd423990f..0a941826f20e189872b7ed4a87d1c91af30ed6e6 100644
--- a/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
@@ -14,8 +14,12 @@ public class ReloadCommand extends BukkitCommand {
         super(name);
         this.description = "Reloads the server configuration and plugins";
         this.usageMessage = "/reload [permissions|commands|confirm]"; // Paper
+        this.usageMessage = "/reload [permissions|confirm]"; // Daydream
+        /*
         this.setPermission("bukkit.command.reload");
         this.setAliases(Arrays.asList("rl"));
+        */
+        this.setPermission("daydream.command.reload"); // Daydream
     }
 
     @Override
@@ -29,6 +33,8 @@ public class ReloadCommand extends BukkitCommand {
                 Bukkit.getServer().reloadPermissions();
                 Command.broadcastCommandMessage(sender, ChatColor.GREEN + "Permissions successfully reloaded.");
                 return true;
+            // Daydream start
+            /*
             } else if ("commands".equalsIgnoreCase(args[0])) {
                 if (Bukkit.getServer().reloadCommandAliases()) {
                     Command.broadcastCommandMessage(sender, ChatColor.GREEN + "Command aliases successfully reloaded.");
@@ -36,6 +42,8 @@ public class ReloadCommand extends BukkitCommand {
                     Command.broadcastCommandMessage(sender, ChatColor.RED + "An error occurred while trying to reload command aliases.");
                 }
                 return true;
+            */
+            // Daydream end
             } else if ("confirm".equalsIgnoreCase(args[0])) {
                 confirmed = true;
             } else {
@@ -60,6 +68,7 @@ public class ReloadCommand extends BukkitCommand {
     @NotNull
     @Override
     public List<String> tabComplete(@NotNull CommandSender sender, @NotNull String alias, @NotNull String[] args) throws IllegalArgumentException {
+        if (true) return com.google.common.collect.Lists.newArrayList("permissions", "confirm"); // Daydream
         return com.google.common.collect.Lists.newArrayList("permissions", "commands"); // Paper
     }
 }
diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
index 2f58de226594b2f0811458ceb5345b8e85bdb8eb..38260ad8b139df0fd1442db17d2ec3c0e2d2afd1 100644
--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -42,15 +42,15 @@ public class VersionCommand extends BukkitCommand {
 
         this.description = "Gets the version of this server including any plugins in use";
         this.usageMessage = "/version [plugin name]";
-        this.setPermission("bukkit.command.version");
-        this.setAliases(Arrays.asList("ver", "about"));
+        //this.setPermission("bukkit.command.version"); // Daydream
+        this.setAliases(Arrays.asList("ver"/*, "about"*/)); // Daydream
     }
 
     @Override
     public boolean execute(@NotNull CommandSender sender, @NotNull String currentAlias, @NotNull String[] args) {
-        if (!testPermission(sender)) return true;
+        //if (!testPermission(sender)) return true; // Daydream
 
-        if (args.length == 0) {
+        if (args.length == 0 || !sender.hasPermission("daydream.command.version")) { // Daydream
             // Daydream start
             /*
             sendVersion(sender);
