From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Fri, 6 Aug 2021 20:08:04 +0900
Subject: [PATCH] Add more Player API


diff --git a/build.gradle.kts b/build.gradle.kts
index 4629e17d2af492d78581736bc390e5cb30386990..546e50adabf87d7779fd38eeaf1fcb0818d65c9d 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -52,6 +52,7 @@ dependencies {
     implementation("org.ow2.asm:asm-commons:9.4")
     // Paper end
     api("com.mojang:brigadier:1.2.9") // Daydream - Add Brigadier Command API
+    api("com.mojang:authlib:5.0.51") // Daydream - Mojang Authlib (keep in sync with major MC versions)
 
     api("org.apache.maven:maven-resolver-provider:3.8.5")
     compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.3")
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 9d76747d18910ef11085c4b23d21ff091e2698ab..aea99dee77bedd98b6b53d716a081ba02ee9d4ce 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3054,6 +3054,15 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @Nullable
     org.bukkit.event.player.PlayerResourcePackStatusEvent.Status getResourcePackStatus();
 
+    // Daydream start
+    /**
+     * Sets the player's resource pack status.
+     *
+     * @param status new resource pack status
+     */
+    void setResourcePackStatus(@NotNull org.bukkit.event.player.PlayerResourcePackStatusEvent.Status status);
+    // Daydream end
+
     /**
      * @return the most recent resource pack hash received from the player,
      *         or null if no hash has ever been received from this player.
@@ -3077,6 +3086,16 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @NotNull
     com.destroystokyo.paper.profile.PlayerProfile getPlayerProfile();
 
+    // Daydream start
+    /**
+     * Gets this player's game profile.
+     *
+     * @return player's game profile object
+     */
+    @NotNull
+    com.mojang.authlib.GameProfile getProfile();
+    // Daydream end
+
     /**
      * Changes the PlayerProfile for this player. This will cause this player
      * to be reregistered to all clients that can currently see this player.
