From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Tue, 12 Oct 2021 20:40:48 +0900
Subject: [PATCH] Add Brigadier Command Registerer


diff --git a/build.gradle.kts b/build.gradle.kts
index ea6c1b996f0df109dcb71428c21b81c6a80011f0..2f67716a3bdb82c96fc7e6541f8a07b5daa42da0 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -40,6 +40,11 @@ dependencies {
     api("org.apache.logging.log4j:log4j-api:2.17.1") // Paper // Daydream
     api("org.slf4j:slf4j-api:1.7.30") // Paper
     api("net.kyori:adventure-text-minimessage:4.2.0-SNAPSHOT") // Daydream
+    // Daydream start
+    api("io.papermc.paper:paper-mojangapi:1.17.1-R0.1-SNAPSHOT") {
+        exclude("io.papermc.paper", "paper-api")
+    }
+    // Daydream end
 
     implementation("org.ow2.asm:asm:9.1")
     implementation("org.ow2.asm:asm-commons:9.1")
diff --git a/src/main/java/moe/caramel/daydream/brigadier/AbstractCommand.java b/src/main/java/moe/caramel/daydream/brigadier/AbstractCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..d71ffa1b304c1ba42f089982b991378cadd600c1
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/brigadier/AbstractCommand.java
@@ -0,0 +1,411 @@
+package moe.caramel.daydream.brigadier;
+
+import com.destroystokyo.paper.brigadier.BukkitBrigadierCommandSource;
+import com.mojang.brigadier.LiteralMessage;
+import com.mojang.brigadier.arguments.ArgumentType;
+import com.mojang.brigadier.builder.LiteralArgumentBuilder;
+import com.mojang.brigadier.builder.RequiredArgumentBuilder;
+import com.mojang.brigadier.suggestion.SuggestionProvider;
+import com.mojang.brigadier.suggestion.Suggestions;
+import com.mojang.brigadier.suggestion.SuggestionsBuilder;
+import net.kyori.adventure.text.Component;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+@SuppressWarnings("unused")
+public abstract class AbstractCommand {
+
+    private String[] aliases;
+    private Component permissionMessage;
+    private String description, permission, usageMessage;
+
+    /**
+     * Make a command!
+     *
+     * @param builder Make it freely using the builder.
+     */
+    public abstract void createCommand(@NotNull LiteralArgumentBuilder<BukkitBrigadierCommandSource> builder);
+
+
+    /*
+        Aliases
+     */
+
+    /**
+     * Returns array of active aliases of this command.
+     *
+     * @return Array of aliases
+     */
+    @Nullable
+    public String[] getAliases() {
+        return aliases;
+    }
+
+    /**
+     * Sets the list of aliases to request on registration for this command.
+     *
+     * @param aliases aliases to register to this command
+     * @return this command object, for chaining
+     */
+    @NotNull
+    public AbstractCommand setAliases(@NotNull String... aliases) {
+        this.aliases = aliases;
+        return this;
+    }
+
+
+    /*
+        Description
+     */
+
+    /**
+     * Gets a brief description of this command.
+     *
+     * @return Description of this command
+     */
+    @Nullable
+    public String getDescription() {
+        return description;
+    }
+
+    /**
+     * Sets a brief description of this command.
+     *
+     * @param description command description
+     * @return this command object, for chaining
+     */
+    @NotNull
+    public AbstractCommand setDescription(@NotNull String description) {
+        this.description = description;
+        return this;
+    }
+
+
+    /*
+        Permission
+     */
+
+    /**
+     * Gets the permission required by users to be able to perform this command.
+     *
+     * @return Permission name, or null if none
+     */
+    @Nullable
+    public String getPermission() {
+        return permission;
+    }
+
+    /**
+     * Sets the permission required by users to be able to perform this command.
+     *
+     * @param permission Permission name or null
+     * @return this command object, for chaining
+     */
+    @NotNull
+    public AbstractCommand setPermission(@Nullable String permission) {
+        this.permission = permission;
+        return this;
+    }
+
+
+    /*
+        PermissionMessage
+     */
+
+    /**
+     * Gets the permission message.
+     *
+     * @return the permission message
+     */
+    public @Nullable Component permissionMessage() {
+        return this.permissionMessage;
+    }
+
+    /**
+     * Returns a message to be displayed on a failed permission check for this command
+     *
+     * @return Permission check failed message
+     * @deprecated use {@link #permissionMessage()}
+     */
+    @Nullable
+    @Deprecated
+    public String getPermissionMessage() {
+        return io.papermc.paper.text.PaperComponents.legacySectionSerializer().serializeOrNull(permissionMessage);
+    }
+
+    /**
+     * Sets the permission message.
+     *
+     * @param permissionMessage the permission message
+     */
+    @NotNull
+    public AbstractCommand permissionMessage(@Nullable Component permissionMessage) {
+        this.permissionMessage = permissionMessage;
+        return this;
+    }
+
+    /**
+     * Sets the message sent when a permission check fails.
+     *
+     * @param permissionMessage new permission message, null to indicate default message, or an empty string to indicate no message
+     * @return this command object, for chaining
+     * @deprecated use {@link #permissionMessage(Component)}
+     */
+    @NotNull
+    @Deprecated
+    public AbstractCommand setPermissionMessage(@Nullable String permissionMessage) {
+        this.permissionMessage = io.papermc.paper.text.PaperComponents.legacySectionSerializer().deserializeOrNull(permissionMessage);
+        return this;
+    }
+
+
+    /*
+       Usage
+     */
+
+    /**
+     * Gets an example usage of this command
+     *
+     * @return One or more example usages
+     */
+    @Nullable
+    public String getUsage() {
+        return usageMessage;
+    }
+
+    /**
+     * Sets the example usage of this command
+     *
+     * @param usage new example usage
+     * @return this command object, for chaining
+     */
+    @NotNull
+    public AbstractCommand setUsage(@NotNull String usage) {
+        this.usageMessage = usage;
+        return this;
+    }
+
+
+    /*
+       Utilities
+     */
+    /**
+     * Creating a literal argument
+     *
+     * @param literal The name of the literal argument
+     * @return A literal argument
+     */
+    @NotNull
+    protected LiteralArgumentBuilder<BukkitBrigadierCommandSource> literal(@NotNull String literal) {
+        return LiteralArgumentBuilder.literal(literal);
+    }
+
+    /**
+     * Creating a required argument
+     *
+     * @param name The name of the argument
+     * @param type The ArgumentType for the type
+     * @param <T>  The type
+     * @return A required argument for the type
+     */
+    @NotNull
+    protected <T> RequiredArgumentBuilder<BukkitBrigadierCommandSource, T> argument(@NotNull String name, @NotNull ArgumentType<T> type) {
+        return RequiredArgumentBuilder.argument(name, type);
+    }
+
+    /**
+     * Utility method for creating custom suggestions quickly
+     *
+     * @param strings The string to suggest
+     * @return The suggestion provider of the inputted strings
+     */
+    @NotNull
+    protected SuggestionProvider<BukkitBrigadierCommandSource> suggest(@NotNull String... strings) {
+        return suggest(Arrays.asList(strings));
+    }
+
+    /**
+     * Utility method for creating custom suggestions quickly (multiple)
+     * <p>
+     * IMPORTANT!
+     * This suggestion only works when Argument is {@link com.mojang.brigadier.arguments.StringArgumentType#greedyString()}.
+     *
+     * @param strings The string to suggest
+     * @return The suggestion provider of the inputted strings
+     */
+    @NotNull
+    protected SuggestionProvider<BukkitBrigadierCommandSource> multipleSuggest(@NotNull String... strings) {
+        return suggest(Arrays.asList(strings));
+    }
+
+    /**
+     * Utility method for creating custom mutable suggestions quickly
+     *
+     * @param strings The mutable string to suggest
+     * @return The suggestion provider of the inputted strings
+     */
+    @NotNull
+    protected SuggestionProvider<BukkitBrigadierCommandSource> mutableSuggestArray(@NotNull Supplier<String[]> strings) {
+        return suggest(strings.get());
+    }
+
+    /**
+     * Utility method for creating custom mutable suggestions quickly
+     * <p>
+     * IMPORTANT!
+     * This suggestion only works when Argument is {@link com.mojang.brigadier.arguments.StringArgumentType#greedyString()}.
+     *
+     * @param strings The mutable string to suggest
+     * @return The suggestion provider of the inputted strings
+     */
+    @NotNull
+    protected SuggestionProvider<BukkitBrigadierCommandSource> mutableMultipleSuggestArray(@NotNull Supplier<String[]> strings) {
+        return multipleSuggest(Arrays.asList(strings.get()));
+    }
+
+    /**
+     * Creating custom suggestions
+     *
+     * @param strings The strings to suggest
+     * @return The SuggestionProvider of the inputted string collection
+     */
+    @NotNull
+    protected SuggestionProvider<BukkitBrigadierCommandSource> suggest(@NotNull Collection<String> strings) {
+        return (c, b) -> suggestMatching(b, strings, false);
+    }
+
+    /**
+     * Creating custom suggestions (multiple)
+     * <p>
+     * IMPORTANT!
+     * This suggestion only works when Argument is {@link com.mojang.brigadier.arguments.StringArgumentType#greedyString()}.
+     *
+     * @param strings The strings to suggest
+     * @return The SuggestionProvider of the inputted string collection
+     */
+    @NotNull
+    protected SuggestionProvider<BukkitBrigadierCommandSource> multipleSuggest(@NotNull Collection<String> strings) {
+        return (c, b) -> suggestMatching(b, strings, true);
+    }
+
+    /**
+     * Creating custom mutable suggestions
+     *
+     * @param strings The mutable strings to suggest
+     * @return The SuggestionProvider of the inputted string collection
+     */
+    @NotNull
+    protected SuggestionProvider<BukkitBrigadierCommandSource> mutableSuggestCollection(@NotNull Function<Collection<String>, Collection<String>> strings) {
+        return (c, b) -> suggestMatching(b, strings.apply(new ArrayList<>()), false);
+    }
+
+    /**
+     * Creating custom mutable suggestions (multiple)
+     * <p>
+     * IMPORTANT!
+     * This suggestion only works when Argument is {@link com.mojang.brigadier.arguments.StringArgumentType#greedyString()}.
+     *
+     * @param strings The mutable strings to suggest
+     * @return The SuggestionProvider of the inputted string collection
+     */
+    @NotNull
+    protected SuggestionProvider<BukkitBrigadierCommandSource> mutableMultipleSuggestCollection(@NotNull Function<Collection<String>, Collection<String>> strings) {
+        return (c, b) -> suggestMatching(b, strings.apply(new ArrayList<>()), true);
+    }
+
+    /**
+     * Creating suggestions with tooltips
+     *
+     * @param suggestions The map of suggestions
+     * @return The suggestion provider for the inputted map
+     */
+    @NotNull
+    protected SuggestionProvider<BukkitBrigadierCommandSource> suggest(@NotNull Map<String, String> suggestions) {
+        return (c, b) -> suggestMatching(b, suggestions, false);
+    }
+
+    /**
+     * Creating suggestions with tooltips (multiple)
+     * <p>
+     * IMPORTANT!
+     * This suggestion only works when Argument is {@link com.mojang.brigadier.arguments.StringArgumentType#greedyString()}.
+     *
+     * @param suggestions The map of suggestions
+     * @return The suggestion provider for the inputted map
+     */
+    @NotNull
+    protected SuggestionProvider<BukkitBrigadierCommandSource> multipleSuggest(@NotNull Map<String, String> suggestions) {
+        return (c, b) -> suggestMatching(b, suggestions, true);
+    }
+
+    /**
+     * Creating mutable suggestions with tooltips
+     *
+     * @param suggestions The map of mutable suggestions
+     * @return The suggestion provider for the inputted map
+     */
+    @NotNull
+    protected SuggestionProvider<BukkitBrigadierCommandSource> mutableSuggestMap(@NotNull Function<Map<String, String>, Map<String, String>> suggestions) {
+        return (c, b) -> suggestMatching(b, suggestions.apply(new HashMap<>()), false);
+    }
+
+    /**
+     * Creating mutable suggestions with tooltips (multiple)
+     * <p>
+     * IMPORTANT!
+     * This suggestion only works when Argument is {@link com.mojang.brigadier.arguments.StringArgumentType#greedyString()}.
+     *
+     * @param suggestions The map of mutable suggestions
+     * @return The suggestion provider for the inputted map
+     */
+    @NotNull
+    protected SuggestionProvider<BukkitBrigadierCommandSource> mutableMultipleSuggestMap(@NotNull Function<Map<String, String>, Map<String, String>> suggestions) {
+        return (c, b) -> suggestMatching(b, suggestions.apply(new HashMap<>()), true);
+    }
+
+    /**
+     * Suggest all matching strings into the given SuggestionsBuilder
+     *
+     * @param builder     The builder to give suggestions to
+     * @param suggestions The suggestions to pick from
+     * @param multiple    is Multiple?
+     * @return The built suggestions
+     */
+    @NotNull
+    private static CompletableFuture<Suggestions> suggestMatching(@NotNull SuggestionsBuilder builder, @NotNull Iterable<String> suggestions, boolean multiple) {
+        if (multiple) builder = builder.createOffset(builder.getInput().lastIndexOf(' ') + 1);
+        String token = builder.getRemainingLowerCase();
+        for (String s : suggestions) if (s.toLowerCase().startsWith(token)) builder.suggest(s);
+
+        return builder.buildFuture();
+    }
+
+    /**
+     * Suggest all matching strings into the given SuggestionsBuilder and add a Message tooltip
+     *
+     * @param builder     The builder to give suggestions to
+     * @param suggestions The suggestions to pick from
+     * @param multiple    is Multiple?
+     * @return The built suggestions
+     */
+    @NotNull
+    private static CompletableFuture<Suggestions> suggestMatching(@NotNull SuggestionsBuilder builder, @NotNull Map<String, String> suggestions, boolean multiple) {
+        if (multiple) builder = builder.createOffset(builder.getInput().lastIndexOf(' ') + 1);
+        String token = builder.getRemainingLowerCase();
+
+        for (Map.Entry<String, String> entry : suggestions.entrySet()) {
+            if (entry.getKey().toLowerCase().startsWith(token))
+                builder.suggest(entry.getKey(), new LiteralMessage(entry.getValue()));
+        }
+
+        return builder.buildFuture();
+    }
+}
diff --git a/src/main/java/moe/caramel/daydream/brigadier/Arguments.java b/src/main/java/moe/caramel/daydream/brigadier/Arguments.java
new file mode 100644
index 0000000000000000000000000000000000000000..8d7d2d56c1c435080c71d6080819a556c5712032
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/brigadier/Arguments.java
@@ -0,0 +1,123 @@
+package moe.caramel.daydream.brigadier;
+
+import com.destroystokyo.paper.brigadier.BukkitBrigadierCommandSource;
+import com.mojang.authlib.GameProfile;
+import moe.caramel.daydream.brigadier.argument.BlockPosArgument;
+import moe.caramel.daydream.brigadier.argument.BlockStateArgument;
+import moe.caramel.daydream.brigadier.argument.SimpleArgument;
+import moe.caramel.daydream.brigadier.argument.SimpleArgumentSuggestion;
+import moe.caramel.daydream.brigadier.argument.EntityAnchorArgument;
+import moe.caramel.daydream.brigadier.argument.EntityArgument;
+import moe.caramel.daydream.brigadier.argument.ItemArgument;
+import moe.caramel.daydream.brigadier.data.Operation;
+import moe.caramel.daydream.brigadier.data.Position;
+import moe.caramel.daydream.brigadier.argument.ResourceLocationArgument;
+import moe.caramel.daydream.brigadier.argument.ScoreHolderArgument;
+import moe.caramel.daydream.brigadier.argument.VecArgument;
+import moe.caramel.daydream.util.NumberRange;
+import net.kyori.adventure.nbt.api.BinaryTagHolder;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.format.TextColor;
+import org.bukkit.Axis;
+import org.bukkit.NamespacedKey;
+import org.bukkit.Particle;
+import org.bukkit.World;
+import org.bukkit.block.Block;
+import org.bukkit.enchantments.Enchantment;
+import org.bukkit.entity.EntityType;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.potion.PotionEffectType;
+import org.bukkit.scoreboard.DisplaySlot;
+import org.bukkit.scoreboard.Objective;
+import org.bukkit.scoreboard.Team;
+
+import java.util.Collection;
+import java.util.EnumSet;
+import java.util.UUID;
+import java.util.function.Predicate;
+
+/**
+ * WARNING!
+ * <p>
+ * This is intended API design.
+ * It was designed in this way for internal debugging.
+ * Don't ever change the value.
+ */
+public final class Arguments {
+
+    public static EntityArgument ENTITY; // "Player", "0123", "@e", "@e[type=foo]", "dd12be42-52a9-4a91-a8a1-11c01849e498"
+
+    public static SimpleArgument<Collection<GameProfile>> GAME_PROFILES; // "Player", "0123", "dd12be42-52a9-4a91-a8a1-11c01849e498", "@e"
+
+    public static BlockPosArgument BLOCK_POS; // "0 0 0", "~ ~ ~", "^ ^ ^", "^1 ^ ^-5", "~0.5 ~1 ~-5"
+
+    public static SimpleArgument<Position<BukkitBrigadierCommandSource>> COLUMN_POS; // "0 0", "~ ~", "~1 ~-2", "^ ^", "^-1 ^0"
+
+    public static VecArgument VEC3; // "0 0 0", "~ ~ ~", "^ ^ ^", "^1 ^ ^-5", "0.1 -0.5 .9", "~0.5 ~1 ~-5"
+
+    public static VecArgument VEC2; // "0 0", "~ ~", "0.1 -0.5", "~1 ~-2"
+
+    public static BlockStateArgument BLOCK_STATE; // "stone", "minecraft:stone", "stone[foo=bar]", "foo{bar=baz}"
+
+    public static SimpleArgument<Predicate<Block>> BLOCK_PREDICATE; // "stone", "minecraft:stone", "stone[foo=bar]", "#stone", "#stone[foo=bar]{baz=nbt}"
+
+    public static ItemArgument ITEM; // "stick", "minecraft:stick", "stick{foo=bar}"
+
+    public static SimpleArgument<Predicate<ItemStack>> ITEM_PREDICATE; // "stick", "minecraft:stick", "#stick", "#stick{foo=bar}"
+
+    public static SimpleArgument<TextColor> COLOR; // "red", "green"
+
+    public static SimpleArgument<Component> COMPONENT; // "\"hello world\"", "\"\"", "\"{\"text\":\"hello world\"}", "[\"\"]"
+
+    public static SimpleArgument<Component> MESSAGE; // "Hello world!", "foo", "@e", "Hello @p :)"
+
+    public static SimpleArgument<BinaryTagHolder> COMPOUND_TAG; // "{}", "{foo=bar}"
+
+    // NbtTagArgument
+
+    // NbtPathArgument
+
+    public static SimpleArgument<Objective> OBJECTIVE; // "foo", "*", "012"
+
+    public static SimpleArgument<String> OBJECTIVE_CRITERIA; // "foo", "foo.bar.baz", "minecraft:foo"
+
+    public static SimpleArgument<Operation> OPERATION; // "=", ">", "<"
+
+    public static SimpleArgument<Particle> PARTICLE; // "foo", "foo:bar", "particle with options"
+
+    public static SimpleArgument<Float> ANGLE; // "0", "~", "~-5"
+
+    public static SimpleArgument<Position<BukkitBrigadierCommandSource>> ROTATION; // "0 0", "~ ~", "~-5 ~5"
+
+    public static SimpleArgument<DisplaySlot> SCOREBOARD_SLOT; // "sidebar", "sidebar.team.light_purple"
+
+    public static ScoreHolderArgument SCORE_HOLDER; // "Player", "0123", "*", "@e"
+
+    public static SimpleArgument<EnumSet<Axis>> SWIZZLE; // "xyz", "x"
+
+    public static SimpleArgument<Team> TEAM; // "foo", "123"
+
+    public static SimpleArgument<Integer> SLOT; // "container.5", "12", "weapon"
+
+    public static ResourceLocationArgument<NamespacedKey> RESOURCE_LOCATION; // "0..5", "0", "-5", "-100..", "..100"
+
+    public static SimpleArgument<PotionEffectType> MOB_EFFECT; // "spooky", "effect"
+
+    // FunctionArgument
+
+    public static SimpleArgument<EntityAnchorArgument> ENTITY_ANCHOR; // "eyes", "feet"
+
+    public static SimpleArgument<NumberRange<Integer>> RANGE_INTEGER; // "0..5", "0", "-5", "-100..", "..100"
+
+    public static SimpleArgument<NumberRange<Double>> RANGE_FLOATS; // "0..5.2", "0", "-5.4", "-100.76..", "..100"
+
+    public static SimpleArgument<Enchantment> ITEM_ENCHANTMENT; // "unbreaking", "silk_touch"
+
+    public static SimpleArgumentSuggestion<EntityType> ENTITY_SUMMON; // "minecraft:pig", "cow"
+
+    public static SimpleArgument<World> DIMENSION; // "minecraft:overworld", "minecraft:the_nether"
+
+    public static SimpleArgument<Integer> TIME; // "0d", "0s", "0t", "0"
+
+    public static SimpleArgument<UUID> UUID; // "dd12be42-52a9-4a91-a8a1-11c01849e498"
+}
diff --git a/src/main/java/moe/caramel/daydream/brigadier/Suggestions.java b/src/main/java/moe/caramel/daydream/brigadier/Suggestions.java
new file mode 100644
index 0000000000000000000000000000000000000000..bfac782a59b64b851515e7307a9d0926e6fbd67f
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/brigadier/Suggestions.java
@@ -0,0 +1,21 @@
+package moe.caramel.daydream.brigadier;
+
+import moe.caramel.daydream.brigadier.suggestion.SimpleSuggestion;
+
+/**
+ * WARNING!
+ * <p>
+ * This is intended API design.
+ * It was designed in this way for internal debugging.
+ * Don't ever change the value.
+ */
+public final class Suggestions {
+
+    public static SimpleSuggestion ALL_RECIPES;
+
+    public static SimpleSuggestion AVAILABLE_SOUNDS;
+
+    public static SimpleSuggestion AVAILABLE_BIOMES;
+
+    public static SimpleSuggestion SUMMONABLE_ENTITIES;
+}
diff --git a/src/main/java/moe/caramel/daydream/brigadier/argument/ArgumentInf.java b/src/main/java/moe/caramel/daydream/brigadier/argument/ArgumentInf.java
new file mode 100644
index 0000000000000000000000000000000000000000..e3009743cec0a4d233628b754f619354570525ef
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/brigadier/argument/ArgumentInf.java
@@ -0,0 +1,9 @@
+package moe.caramel.daydream.brigadier.argument;
+
+import com.mojang.brigadier.arguments.ArgumentType;
+import org.jetbrains.annotations.NotNull;
+
+public interface ArgumentInf {
+
+    @NotNull ArgumentType<?> get();
+}
diff --git a/src/main/java/moe/caramel/daydream/brigadier/argument/BlockPosArgument.java b/src/main/java/moe/caramel/daydream/brigadier/argument/BlockPosArgument.java
new file mode 100644
index 0000000000000000000000000000000000000000..e564721974edfe3b7f34012cbd921e8205e191f2
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/brigadier/argument/BlockPosArgument.java
@@ -0,0 +1,16 @@
+package moe.caramel.daydream.brigadier.argument;
+
+import com.destroystokyo.paper.brigadier.BukkitBrigadierCommandSource;
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import org.bukkit.block.Block;
+import org.jetbrains.annotations.NotNull;
+
+public interface BlockPosArgument extends ArgumentInf {
+
+    @NotNull
+    Block getLoadedBlockPos(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+
+    @NotNull
+    Block getSpawnablePos(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+}
diff --git a/src/main/java/moe/caramel/daydream/brigadier/argument/BlockStateArgument.java b/src/main/java/moe/caramel/daydream/brigadier/argument/BlockStateArgument.java
new file mode 100644
index 0000000000000000000000000000000000000000..94e19bb08f9a37be2133e0f3712a47f5b361409c
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/brigadier/argument/BlockStateArgument.java
@@ -0,0 +1,23 @@
+package moe.caramel.daydream.brigadier.argument;
+
+import com.destroystokyo.paper.brigadier.BukkitBrigadierCommandSource;
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import org.bukkit.Location;
+import org.bukkit.World;
+import org.bukkit.block.Block;
+import org.bukkit.block.data.BlockData;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.function.Predicate;
+
+public interface BlockStateArgument extends ArgumentInf {
+
+    @NotNull
+    BlockData getState(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+
+    boolean place(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name, @NotNull World world, @NotNull Location location, int flags);
+
+    @NotNull
+    Predicate<Block> getPredicate(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+}
diff --git a/src/main/java/moe/caramel/daydream/brigadier/argument/EntityAnchorArgument.java b/src/main/java/moe/caramel/daydream/brigadier/argument/EntityAnchorArgument.java
new file mode 100644
index 0000000000000000000000000000000000000000..cfd681a43f4b4f10d5c55e4c0a2fe1bbc40b7dad
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/brigadier/argument/EntityAnchorArgument.java
@@ -0,0 +1,10 @@
+package moe.caramel.daydream.brigadier.argument;
+
+import org.bukkit.entity.Entity;
+import org.bukkit.util.Vector;
+import org.jetbrains.annotations.NotNull;
+
+public interface EntityAnchorArgument {
+
+    @NotNull Vector apply(@NotNull Entity entity);
+}
diff --git a/src/main/java/moe/caramel/daydream/brigadier/argument/EntityArgument.java b/src/main/java/moe/caramel/daydream/brigadier/argument/EntityArgument.java
new file mode 100644
index 0000000000000000000000000000000000000000..1f343e849551acae44fbd1b6922ec65a72834935
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/brigadier/argument/EntityArgument.java
@@ -0,0 +1,32 @@
+package moe.caramel.daydream.brigadier.argument;
+
+import com.destroystokyo.paper.brigadier.BukkitBrigadierCommandSource;
+import com.mojang.brigadier.arguments.ArgumentType;
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.Player;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+
+public interface EntityArgument {
+
+    @NotNull ArgumentType<?> entity();
+
+    @NotNull ArgumentType<?> entities();
+
+    @NotNull ArgumentType<?> player();
+
+    @NotNull ArgumentType<?> players();
+
+    @NotNull Entity getEntity(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+
+    @NotNull Collection<? extends Entity> getEntities(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+
+    @NotNull Collection<Player> getOptionalPlayers(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+
+    @NotNull Player getPlayer(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+
+    @NotNull Collection<Player> getPlayers(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+}
diff --git a/src/main/java/moe/caramel/daydream/brigadier/argument/ItemArgument.java b/src/main/java/moe/caramel/daydream/brigadier/argument/ItemArgument.java
new file mode 100644
index 0000000000000000000000000000000000000000..0bcfb4a2118d2c2743bcd86d77b528021c1769e4
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/brigadier/argument/ItemArgument.java
@@ -0,0 +1,16 @@
+package moe.caramel.daydream.brigadier.argument;
+
+import com.destroystokyo.paper.brigadier.BukkitBrigadierCommandSource;
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.NotNull;
+
+public interface ItemArgument extends ArgumentInf {
+
+    @NotNull
+    ItemStack getItem(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+
+    @NotNull
+    ItemStack createItemStack(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name, int amount, boolean checkOverstack) throws CommandSyntaxException;
+}
diff --git a/src/main/java/moe/caramel/daydream/brigadier/argument/ResourceLocationArgument.java b/src/main/java/moe/caramel/daydream/brigadier/argument/ResourceLocationArgument.java
new file mode 100644
index 0000000000000000000000000000000000000000..ce65655aa3f3192ed12f6d0f52d00fbff72edc51
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/brigadier/argument/ResourceLocationArgument.java
@@ -0,0 +1,30 @@
+package moe.caramel.daydream.brigadier.argument;
+
+import com.destroystokyo.paper.brigadier.BukkitBrigadierCommandSource;
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import org.bukkit.advancement.Advancement;
+import org.bukkit.attribute.Attribute;
+import org.jetbrains.annotations.NotNull;
+
+public interface ResourceLocationArgument<T> extends SimpleArgument<T> {
+
+    @NotNull
+    Advancement getAdvancement(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+
+    @NotNull
+    T getRecipe(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+
+    /*
+    @NotNull
+    T getPredicate(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+     */
+
+    @NotNull
+    Attribute getAttribute(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+
+    /*
+    @NotNull
+    T getItemModifier(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+    */
+}
diff --git a/src/main/java/moe/caramel/daydream/brigadier/argument/ScoreHolderArgument.java b/src/main/java/moe/caramel/daydream/brigadier/argument/ScoreHolderArgument.java
new file mode 100644
index 0000000000000000000000000000000000000000..e5ac0f101a76c8c6d8058397af753865c381ae7d
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/brigadier/argument/ScoreHolderArgument.java
@@ -0,0 +1,28 @@
+package moe.caramel.daydream.brigadier.argument;
+
+import com.destroystokyo.paper.brigadier.BukkitBrigadierCommandSource;
+import com.mojang.brigadier.arguments.ArgumentType;
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import com.mojang.brigadier.suggestion.SuggestionProvider;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+import java.util.function.Supplier;
+
+public interface ScoreHolderArgument {
+
+    @NotNull ArgumentType<?> getHolder();
+
+    @NotNull ArgumentType<?> getHolders();
+
+    @NotNull String getName(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+
+    @NotNull Collection<String> getNames(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+
+    @NotNull Collection<String> getNamesWithDefaultWildcard(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+
+    @NotNull Collection<String> getNamesSupplier(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name, @NotNull Supplier<Collection<String>> players) throws CommandSyntaxException;
+
+    @NotNull SuggestionProvider<BukkitBrigadierCommandSource> getSuggestion();
+}
diff --git a/src/main/java/moe/caramel/daydream/brigadier/argument/SimpleArgument.java b/src/main/java/moe/caramel/daydream/brigadier/argument/SimpleArgument.java
new file mode 100644
index 0000000000000000000000000000000000000000..378810282864c9a3ceb7ecb72cad20f0e7ff2e3e
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/brigadier/argument/SimpleArgument.java
@@ -0,0 +1,12 @@
+package moe.caramel.daydream.brigadier.argument;
+
+import com.destroystokyo.paper.brigadier.BukkitBrigadierCommandSource;
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import org.jetbrains.annotations.NotNull;
+
+public interface SimpleArgument<T> extends ArgumentInf {
+
+    @NotNull
+    T getData(final @NotNull CommandContext<? extends BukkitBrigadierCommandSource> context, @NotNull String name) throws CommandSyntaxException;
+}
diff --git a/src/main/java/moe/caramel/daydream/brigadier/argument/SimpleArgumentSuggestion.java b/src/main/java/moe/caramel/daydream/brigadier/argument/SimpleArgumentSuggestion.java
new file mode 100644
index 0000000000000000000000000000000000000000..fc49c907007458c21ad13e7b9e1c55380647e4b7
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/brigadier/argument/SimpleArgumentSuggestion.java
@@ -0,0 +1,5 @@
+package moe.caramel.daydream.brigadier.argument;
+
+import moe.caramel.daydream.brigadier.suggestion.SimpleSuggestion;
+
+public interface SimpleArgumentSuggestion<T> extends SimpleArgument<T>, SimpleSuggestion {}
diff --git a/src/main/java/moe/caramel/daydream/brigadier/argument/VecArgument.java b/src/main/java/moe/caramel/daydream/brigadier/argument/VecArgument.java
new file mode 100644
index 0000000000000000000000000000000000000000..30ad53e262f8f4bc9746edcab9bf79f475bb29bc
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/brigadier/argument/VecArgument.java
@@ -0,0 +1,11 @@
+package moe.caramel.daydream.brigadier.argument;
+
+import com.destroystokyo.paper.brigadier.BukkitBrigadierCommandSource;
+import com.mojang.brigadier.arguments.ArgumentType;
+import moe.caramel.daydream.brigadier.data.Position;
+import org.jetbrains.annotations.NotNull;
+
+public interface VecArgument extends SimpleArgument<Position<BukkitBrigadierCommandSource>> {
+
+    @NotNull ArgumentType<?> get(boolean centerIntegers);
+}
diff --git a/src/main/java/moe/caramel/daydream/brigadier/data/Operation.java b/src/main/java/moe/caramel/daydream/brigadier/data/Operation.java
new file mode 100644
index 0000000000000000000000000000000000000000..2b11e6a5c9a841dabd5d91632a0945a1a128515c
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/brigadier/data/Operation.java
@@ -0,0 +1,9 @@
+package moe.caramel.daydream.brigadier.data;
+
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+
+@FunctionalInterface
+public interface Operation {
+
+    int apply(int score, int score2) throws CommandSyntaxException;
+}
diff --git a/src/main/java/moe/caramel/daydream/brigadier/data/Position.java b/src/main/java/moe/caramel/daydream/brigadier/data/Position.java
new file mode 100644
index 0000000000000000000000000000000000000000..1711a2159548ea1686a02a7225a6145d9f0fb76c
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/brigadier/data/Position.java
@@ -0,0 +1,23 @@
+package moe.caramel.daydream.brigadier.data;
+
+import com.destroystokyo.paper.brigadier.BukkitBrigadierCommandSource;
+import org.bukkit.Location;
+import org.bukkit.util.Vector;
+import org.jetbrains.annotations.NotNull;
+
+public interface Position<C extends BukkitBrigadierCommandSource> {
+
+    @NotNull Vector getPositionVector(@NotNull C source);
+
+    float getXRot(@NotNull C source);
+
+    float getYRot(@NotNull C source);
+
+    @NotNull Location getLocation(@NotNull C source);
+
+    boolean isXRelative();
+
+    boolean isYRelative();
+
+    boolean isZRelative();
+}
diff --git a/src/main/java/moe/caramel/daydream/brigadier/suggestion/SimpleSuggestion.java b/src/main/java/moe/caramel/daydream/brigadier/suggestion/SimpleSuggestion.java
new file mode 100644
index 0000000000000000000000000000000000000000..01496ae337322cd91fe42444554a2ca36068e145
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/brigadier/suggestion/SimpleSuggestion.java
@@ -0,0 +1,11 @@
+package moe.caramel.daydream.brigadier.suggestion;
+
+import com.destroystokyo.paper.brigadier.BukkitBrigadierCommandSource;
+import com.mojang.brigadier.suggestion.SuggestionProvider;
+import org.jetbrains.annotations.NotNull;
+
+public interface SimpleSuggestion {
+
+    @NotNull
+    SuggestionProvider<BukkitBrigadierCommandSource> getSuggestion();
+}
diff --git a/src/main/java/moe/caramel/daydream/util/NumberRange.java b/src/main/java/moe/caramel/daydream/util/NumberRange.java
new file mode 100644
index 0000000000000000000000000000000000000000..068c6ed9299935b24cafe5b826bcd9f9822ac4af
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/util/NumberRange.java
@@ -0,0 +1,5 @@
+package moe.caramel.daydream.util;
+
+import org.jetbrains.annotations.NotNull;
+
+public record NumberRange<T extends Number>(@NotNull T min, @NotNull T max) {}
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index c551010f84ac5d3ebc626c253a8f8282924152c4..e5bba7b80bc577ace301b77900f49d8ef2a97108 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -820,6 +820,19 @@ public final class Bukkit {
         return server.getPluginCommand(name);
     }
 
+    // Daydream start
+    /**
+     * Register command using Brigadier.
+     *
+     * @param plugin Plugin
+     * @param label command label
+     * @param command Command node to register
+     */
+    public static void registerCommand(@NotNull org.bukkit.plugin.Plugin plugin, @NotNull String label, @NotNull moe.caramel.daydream.brigadier.AbstractCommand command) {
+        server.registerCommand(plugin, label, command);
+    }
+    // Daydream end
+
     /**
      * Writes loaded players to disk.
      */
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 72b82efe67c4917216407e9aad86ba9e7e9efce5..c4d8e3fee567a378d0a44b351a0d0fb407439abd 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -680,6 +680,17 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @Nullable
     public PluginCommand getPluginCommand(@NotNull String name);
 
+    // Daydream start
+    /**
+     * Register command using Brigadier.
+     *
+     * @param plugin Plugin
+     * @param label command label
+     * @param command Command node to register
+     */
+    public void registerCommand(@NotNull org.bukkit.plugin.Plugin plugin, @NotNull String label, @NotNull moe.caramel.daydream.brigadier.AbstractCommand command);
+    // Daydream end
+
     /**
      * Writes loaded players to disk.
      */
diff --git a/src/main/java/org/bukkit/command/Command.java b/src/main/java/org/bukkit/command/Command.java
index 95847fcb1ac2e430ee192f4f7ac94e981151c5db..9f1b4141aae7f745083a73a4e204a473d3d0ff8f 100644
--- a/src/main/java/org/bukkit/command/Command.java
+++ b/src/main/java/org/bukkit/command/Command.java
@@ -185,7 +185,7 @@ public abstract class Command {
         }
 
             // Paper start - use components for permissionMessage
-        net.kyori.adventure.text.Component permissionMessage = this.permissionMessage != null ? this.permissionMessage : io.papermc.paper.text.PaperComponents.legacySectionSerializer().deserialize(Bukkit.getPermissionMessage());
+        net.kyori.adventure.text.Component permissionMessage = this.permissionMessage != null ? this.permissionMessage : (Bukkit.getPermissionMessage().length() == 0 ? net.kyori.adventure.text.Component.translatable("command.failed", net.kyori.adventure.text.format.NamedTextColor.RED) : io.papermc.paper.text.PaperComponents.legacySectionSerializer().deserialize(Bukkit.getPermissionMessage())); // Daydream
         if (!permissionMessage.equals(net.kyori.adventure.text.Component.empty())) {
             target.sendMessage(permissionMessage.replaceText(net.kyori.adventure.text.TextReplacementConfig.builder().matchLiteral("<permission>").replacement(permission).build()));
             // Paper end
