From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 9 Apr 2023 14:33:20 +0900
Subject: [PATCH] Add more Event API


diff --git a/src/main/java/org/bukkit/event/Event.java b/src/main/java/org/bukkit/event/Event.java
index 8ec56cd6b8e0f5c5dd8c7c88b4671e18dcf109d0..3df96b64ad5ce31a69229f6fd14ca750f041e422 100644
--- a/src/main/java/org/bukkit/event/Event.java
+++ b/src/main/java/org/bukkit/event/Event.java
@@ -51,6 +51,29 @@ public abstract class Event {
     }
     // Paper end
 
+    // Daydream start - Add more Event API
+    private final java.util.Map<String, Object> metadata = new java.util.HashMap<>();
+
+    /**
+     * Gets the metadata of this event.
+     *
+     * @return The event metadata map.
+    */
+    @NotNull
+    public java.util.Map<String, Object> getEventMetadata() {
+        return metadata;
+    }
+
+    /**
+     * 이 이벤트가 인터럽트를 지원하는지 여부를 가져옵니다.
+     *
+     * @return 인터럽트 지원 여부
+     */
+    public boolean isSupportInterrupt() {
+        return false;
+    }
+    // Daydream end - Add more Event API
+
     /**
      * Convenience method for providing a user-friendly identifier. By
      * default, it is the event's class's {@linkplain Class#getSimpleName()
