From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Fri, 6 Aug 2021 20:08:04 +0900
Subject: [PATCH] Add more ResourcePack API


diff --git a/src/main/java/moe/caramel/daydream/player/ResourcePack.java b/src/main/java/moe/caramel/daydream/player/ResourcePack.java
new file mode 100644
index 0000000000000000000000000000000000000000..46702d7db63fca173b78a507b4900e031b89a4a0
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/player/ResourcePack.java
@@ -0,0 +1,114 @@
+package moe.caramel.daydream.player;
+
+import net.kyori.adventure.resource.ResourcePackInfo;
+import net.kyori.adventure.resource.ResourcePackStatus;
+import org.bukkit.event.player.PlayerResourcePackStatusEvent.Status;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import java.net.URI;
+import java.util.UUID;
+
+/**
+ * 리소스팩 데이터
+ */
+public interface ResourcePack {
+
+    /**
+     * 리소스팩 정보를 가져옵니다.
+     *
+     * @return 리소스팩 정보
+     */
+    @NotNull
+    ResourcePackInfo info();
+
+    /**
+     * 리소스팩의 UUID를 가져옵니다.
+     *
+     * @return 리소스팩의 UUID
+     */
+    @NotNull
+    default UUID id() {
+        return info().id();
+    }
+
+    /**
+     * 리소스팩의 URI를 가져옵니다.
+     *
+     * @return 리소스팩의 URI
+     */
+    @NotNull
+    default URI uri() {
+        return info().uri();
+    }
+
+    /**
+     * 리소스팩의 해시를 가져옵니다.
+     *
+     * @return 리소스팩의 해시
+     */
+    @NotNull
+    default String hash() {
+        return info().hash();
+    }
+
+    /**
+     * 리소스팩 적용 상태를 Adventure 형식으로 가져옵니다.
+     *
+     * @return 리소스팩 적용 상태
+     */
+    @Nullable
+    ResourcePackStatus status();
+
+    /**
+     * 리소스팩 적용 상태를 Bukkit 형식으로 가져옵니다.
+     *
+     * @return 리소스팩 적용 상태
+     */
+    @Nullable
+    default Status bukkitStatus() {
+        return adventureToBukkit(status());
+    }
+
+    // ===========================
+
+    /**
+     * Adventure 리소스팩 상태를 Bukkit 형식으로 변환합니다.
+     *
+     * @param status Adventure 상태
+     * @return Bukkit 상태
+     */
+    @Contract("!null -> !null; null -> null")
+    static @Nullable Status adventureToBukkit(final @Nullable ResourcePackStatus status) {
+        return (status == null) ? null : switch (status) {
+            case SUCCESSFULLY_LOADED -> Status.SUCCESSFULLY_LOADED;
+            case DECLINED -> Status.DECLINED;
+            case FAILED_DOWNLOAD -> Status.FAILED_DOWNLOAD;
+            case ACCEPTED -> Status.ACCEPTED;
+            case DOWNLOADED -> Status.DOWNLOADED;
+            case INVALID_URL -> Status.INVALID_URL;
+            case FAILED_RELOAD -> Status.FAILED_RELOAD;
+            case DISCARDED -> Status.DISCARDED;
+        };
+    }
+
+    /**
+     *  Bukkit 리소스팩 상태를 Adventure 형식으로 변환합니다.
+     *
+     * @param status Bukkit 상태
+     * @return Adventure 상태
+     */
+    @Contract("!null -> !null; null -> null")
+    static @Nullable ResourcePackStatus bukkitToAdventure(final @Nullable Status status) {
+        return (status == null) ? null : switch (status) {
+            case SUCCESSFULLY_LOADED -> ResourcePackStatus.SUCCESSFULLY_LOADED;
+            case DECLINED -> ResourcePackStatus.DECLINED;
+            case FAILED_DOWNLOAD -> ResourcePackStatus.FAILED_DOWNLOAD;
+            case ACCEPTED -> ResourcePackStatus.ACCEPTED;
+            case DOWNLOADED -> ResourcePackStatus.DOWNLOADED;
+            case INVALID_URL -> ResourcePackStatus.INVALID_URL;
+            case FAILED_RELOAD -> ResourcePackStatus.FAILED_RELOAD;
+            case DISCARDED -> ResourcePackStatus.DISCARDED;
+        };
+    }
+}
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index a4e0f613058f96db94f2fd91d015a87dda356a70..9ea8be5a9ab65c1f1f117b581d75ecb6f1c44155 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2764,6 +2764,23 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     org.bukkit.event.player.PlayerResourcePackStatusEvent.@Nullable Status getResourcePackStatus();
 
+    // Daydream start - Add more ResourcePack API
+    /**
+     * 적용된 리소스팩 목록을 가져옵니다.
+     *
+     * @return 리소스팩 목록
+     */
+    @NotNull Collection<? extends moe.caramel.daydream.player.ResourcePack> getAppliedResourcePacks();
+
+    /**
+     * 리소스팩 적용 상태를 가져옵니다.
+     *
+     * @param uuid 리소스팩의 고유 ID
+     * @return 적용을 하지 않았거나, 클라이언트에게 응답을 받지 않은 경우 {@code null}을 반환할 수도 있습니다.
+     */
+    @Nullable org.bukkit.event.player.PlayerResourcePackStatusEvent.Status getResourcePackStatus(@NotNull UUID uuid);
+    // Daydream end - Add more ResourcePack API
+
     /**
      * Gets the most recent pack hash from the player.
      *
