From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Fri, 6 Aug 2021 18:55:00 +0900
Subject: [PATCH] Bring back server name


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index e5bba7b80bc577ace301b77900f49d8ef2a97108..e0c3355e2ce93e6a6f682526967aaae75358661c 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -232,6 +232,18 @@ public final class Bukkit {
         return server.getIp();
     }
 
+    // Daydream start
+    /**
+     * Get the name of this server.
+     *
+     * @return the name of this server
+     */
+    @NotNull
+    public static String getServerName() {
+        return server.getServerName();
+    }
+    // Daydream end
+
     /**
      * Get world type (level-type setting) for default world.
      *
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index c4d8e3fee567a378d0a44b351a0d0fb407439abd..b8a2283123822dccb8b7d7228413000d043287f5 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -178,6 +178,16 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     public String getIp();
 
+    // Daydream start
+    /**
+     * Get the name of this server.
+     *
+     * @return the name of this server
+     */
+    @NotNull
+    public String getServerName();
+    // Daydream end
+
     /**
      * Get world type (level-type setting) for default world.
      *
