From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 8 Aug 2021 13:12:57 +0900
Subject: [PATCH] Add PlayerSendIllegalChatEvent


diff --git a/src/main/java/moe/caramel/daydream/event/packet/PlayerSendIllegalChatEvent.java b/src/main/java/moe/caramel/daydream/event/packet/PlayerSendIllegalChatEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..6c210fd6aabf28e2aea6aff27c779a0f57ad3988
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/event/packet/PlayerSendIllegalChatEvent.java
@@ -0,0 +1,53 @@
+package moe.caramel.daydream.event.packet;
+
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * 플레이어가 잘못된 채팅 패킷을 서버에 전송했을 때 호출됩니다.
+ */
+public final class PlayerSendIllegalChatEvent extends PlayerEvent {
+
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+    public @NotNull HandlerList getHandlers() { return HANDLER_LIST; }
+    public static @NotNull HandlerList getHandlerList() { return HANDLER_LIST; }
+
+    private final Action action;
+
+    @ApiStatus.Internal
+    public PlayerSendIllegalChatEvent(final @NotNull Player player, final @NotNull Action action) {
+        super(player);
+        this.action = action;
+    }
+
+    /**
+     * 잘못된 채팅 행동을 가져옵니다.
+     *
+     * @return 잘못된 채팅 행동
+     */
+    @NotNull
+    public Action getAction() {
+        return action;
+    }
+
+    /**
+     * 잘못된 채팅 행동 목록
+     */
+    public enum Action {
+        /**
+         * 플레이어가 빈 메시지를 전송할 때
+         */
+        EMPTY,
+        /**
+         * 플레이어가 사망 상태일 때
+         */
+        PLAYER_DEAD,
+        /**
+         * 채팅 표시 상태가 숨김일 때
+         */
+        VISIBILITY_IS_HIDDEN
+    }
+}
