From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Thu, 22 Jul 2021 16:19:39 +0900
Subject: [PATCH] Add caramelPaper API (Player)


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 27848ad0eee667e515685a5f1cef3e9bfc7a3f53..d667137301e830b761ba179dcbd0b51ee3277633 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -17,6 +17,7 @@ import org.bukkit.GameMode;
 import org.bukkit.Instrument;
 import org.bukkit.Location;
 import org.bukkit.Material;
+import org.bukkit.NamespacedKey; // Daydream
 import org.bukkit.Note;
 import org.bukkit.OfflinePlayer;
 import org.bukkit.Particle;
@@ -883,7 +884,7 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @param fadeInTicks  ticks to fade-in
      * @param stayTicks    ticks to stay visible
      * @param fadeOutTicks ticks to fade-out
-     * @deprecated Use {@link #showTitle(net.kyori.adventure.title.Title)} 
+     * @deprecated Use {@link #showTitle(net.kyori.adventure.title.Title)}
      */
     @Deprecated
     public void setTitleTimes(int fadeInTicks, int stayTicks, int fadeOutTicks);
@@ -1872,6 +1873,24 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
         return net.kyori.adventure.text.event.HoverEvent.showEntity(op.apply(net.kyori.adventure.text.event.HoverEvent.ShowEntity.of(this.getType().getKey(), this.getUniqueId(), this.displayName())));
     }
 
+    // Daydream start
+    /**
+     * Send a change game state packet for this player
+     *
+     * @param reason Reason Value
+     * @param value Value
+     */
+    void sendChangeGameState(int reason, int value);
+
+    /**
+     * Send a CustomPayload packet for this player
+     *
+     * @param channel channel Value
+     * @param value Value
+     */
+    void sendCustomPayload(@NotNull NamespacedKey channel, @NotNull String value);
+    // Daydream end
+
     /**
      * Request that the player's client download and switch resource packs.
      * <p>
@@ -1910,6 +1929,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @Nullable
     org.bukkit.event.player.PlayerResourcePackStatusEvent.Status getResourcePackStatus();
 
+    // Daydream start
+    /**
+     * @param status Sets the status of the player's resourcepack.
+     */
+    void setResourcePackStatus(@NotNull org.bukkit.event.player.PlayerResourcePackStatusEvent.Status status);
+    // Daydream end
+
     /**
      * @return the most recent resource pack hash received from the player,
      *         or null if no hash has ever been received from this player.
@@ -1933,6 +1959,15 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @NotNull
     PlayerProfile getPlayerProfile();
 
+    // Daydream start
+    /**
+     * Gets this players profile
+     * @return The players profile object
+     */
+    @NotNull
+    com.mojang.authlib.GameProfile getProfile();
+    // Daydream end
+
     /**
      * Changes the PlayerProfile for this player. This will cause this player
      * to be reregistered to all clients that can currently see this player
@@ -1940,6 +1975,16 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     void setPlayerProfile(@NotNull PlayerProfile profile);
 
+    // Daydream start
+    /**
+     * Changes the PlayerProfile for this player. This will cause this player
+     * to be reregistered to all clients that can currently see this player
+     * @param profile The new profile to use
+     * @param refreshSelf refresh player info this player?
+     */
+    void setPlayerProfile(@NotNull PlayerProfile profile, boolean refreshSelf);
+    // Daydream end
+
     /**
      * Returns the amount of ticks the current cooldown lasts
      *
