From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 8 Aug 2021 15:45:15 +0900
Subject: [PATCH] Add PlayerStatisticOpenEvent


diff --git a/src/main/java/moe/caramel/daydream/event/packet/PlayerStatisticOpenEvent.java b/src/main/java/moe/caramel/daydream/event/packet/PlayerStatisticOpenEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..73d0f6f9bcb155288674bb1103fb28cf6881eed1
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/event/packet/PlayerStatisticOpenEvent.java
@@ -0,0 +1,47 @@
+package moe.caramel.daydream.event.packet;
+
+import org.bukkit.entity.Player;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * 플레이어가 통계 화면을 열었을 경우 호출됩니다.
+ */
+public final class PlayerStatisticOpenEvent extends PlayerEvent implements Cancellable {
+
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+    public @NotNull HandlerList getHandlers() { return HANDLER_LIST; }
+    public static @NotNull HandlerList getHandlerList() { return HANDLER_LIST; }
+
+    private boolean cancelled;
+
+    public PlayerStatisticOpenEvent(@NotNull Player player) {
+        super(player);
+    }
+
+    /**
+     * 통계 패킷을 전송할지의 여부를 가져옵니다.
+     *
+     * @return 통계 패킷을 전송할지의 여부
+     */
+    @Override
+    public boolean isCancelled() {
+        return this.cancelled;
+    }
+
+    /**
+     * 통계 패킷을 전송하지 않습니다.
+     * <p>
+     * 주의: 이벤트를 취소해도 클라이언트에는 통계 로드 화면이 계속 출력됩니다.
+     * 인벤토리를 열고 닫는 방법을 사용하여 로드 화면을 제거할 수 있습니다.
+     * </p>
+     *
+     * @param cancel {@code true}인 경우 통계 패킷을 전송하지 않습니다.
+     */
+    @Override
+    public void setCancelled(boolean cancel) {
+        this.cancelled = cancel;
+    }
+}
