From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Fri, 6 Aug 2021 18:55:00 +0900
Subject: [PATCH] Bring back server name


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index e8414592b3afeb1e5db2b817b8fb7c13e073b9aa..2accb726a5eabacbcc2e142d568be8a072c4a3dc 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -212,6 +212,20 @@ public final class Bukkit {
         return server.getIp();
     }
 
+    // Daydream start
+    /**
+     * Get the IP that this server is bound to, or empty string if not
+     * specified.
+     *
+     * @return the IP string that this server is bound to, otherwise empty
+     *     string
+     */
+    @NotNull
+    public static String getServerName() {
+        return server.getServerName();
+    }
+    // Daydream end
+
     /**
      * Get world type (level-type setting) for default world.
      *
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1aa2e0384ac3b1e48ebf442d1ec9f075d8341c68..7aec792d9de7b8008d700e8acf526db47164cd57 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -177,6 +177,18 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     public String getIp();
 
+    // Daydream start
+    /**
+     * Get the IP that this server is bound to, or empty string if not
+     * specified.
+     *
+     * @return the IP string that this server is bound to, otherwise empty
+     *     string
+     */
+    @NotNull
+    public String getServerName();
+    // Daydream end
+
     /**
      * Get world type (level-type setting) for default world.
      *
