From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Fri, 6 Aug 2021 18:55:00 +0900
Subject: [PATCH] Bring back server name


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 94c8b4c118c8f7fe73f0a44ccc0b367786251c99..7e6be8a8bdcd4c87fe938499e4a6dbc806d5bd16 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -212,6 +212,20 @@ public final class Bukkit {
         return server.getIp();
     }
 
+    // Daydream start
+    /**
+     * Get the IP that this server is bound to, or empty string if not
+     * specified.
+     *
+     * @return the IP string that this server is bound to, otherwise empty
+     *     string
+     */
+    @NotNull
+    public static String getServerName() {
+        return server.getServerName();
+    }
+    // Daydream end
+
     /**
      * Get world type (level-type setting) for default world.
      *
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index aeca69ac96441bc64aaa1f85cec1a3c1bc05616f..258e508d5c48a562ceb5ebf80ac82d48a93f1214 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -177,6 +177,18 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     public String getIp();
 
+    // Daydream start
+    /**
+     * Get the IP that this server is bound to, or empty string if not
+     * specified.
+     *
+     * @return the IP string that this server is bound to, otherwise empty
+     *     string
+     */
+    @NotNull
+    public String getServerName();
+    // Daydream end
+
     /**
      * Get world type (level-type setting) for default world.
      *
