From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 8 Aug 2021 21:16:23 +0900
Subject: [PATCH] Add PlayerAdvancementScreenEvent


diff --git a/src/main/java/moe/caramel/daydream/event/packet/PlayerAdvancementScreenEvent.java b/src/main/java/moe/caramel/daydream/event/packet/PlayerAdvancementScreenEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..c28145485940fd9c8114355fb0106c4515491eae
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/event/packet/PlayerAdvancementScreenEvent.java
@@ -0,0 +1,84 @@
+package moe.caramel.daydream.event.packet;
+
+import org.bukkit.NamespacedKey;
+import org.bukkit.entity.Player;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * 플레이어가 발전 과제 화면을 열거나 닫았을 경우 호출됩니다.
+ */
+public final class PlayerAdvancementScreenEvent extends PlayerEvent implements Cancellable {
+
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+    public @NotNull HandlerList getHandlers() { return HANDLER_LIST; }
+    public static @NotNull HandlerList getHandlerList() { return HANDLER_LIST; }
+
+    private final Action action;
+    private final NamespacedKey key;
+    private boolean cancelled;
+
+    public PlayerAdvancementScreenEvent(@NotNull Player player, @NotNull Action action, @Nullable NamespacedKey key) {
+        super(player);
+        this.action = action;
+        this.key = key;
+    }
+
+    /**
+     * 발전 과제 화면을 닫을지의 여부를 가져옵니다.
+     *
+     * @return {@code true}인 경우 스크린을 닫습니다.
+     * {@link Action#CLOSED_SCREEN}인 경우 작동하지 않습니다.
+     */
+    @Override
+    public boolean isCancelled() {
+        return this.cancelled;
+    }
+
+    /**
+     * 스크린을 닫을지의 여부를 가져옵니다.
+     * 인수가 {@code true}인 경우 스크린을 닫으나, {@link Action#CLOSED_SCREEN}인 경우 작동하지 않습니다.
+     */
+    @Override
+    public void setCancelled(boolean cancel) {
+        this.cancelled = cancel;
+    }
+
+    /**
+     * 플레이어가 요청한 액션을 가져옵니다.
+     *
+     * @return 플레이어의 요청
+     */
+    @NotNull
+    public Action getAction() {
+        return action;
+    }
+
+    /**
+     * 플레이어가 요청한 탭의 {@link NamespacedKey}를 가져옵니다.
+     *
+     * @return 플레이어가 요청한 탭의 {@link NamespacedKey}.
+     * {@link Action#CLOSED_SCREEN}인 경우 {@code null}을 반환합니다.
+     */
+    @Nullable
+    public NamespacedKey getKey() {
+        return key;
+    }
+
+    /**
+     * 발전 과제 요청 목록
+     */
+    public enum Action {
+        /**
+         * 새로운 발전 과제 탭을 오픈하는 경우
+         */
+        OPENED_TAB,
+        /**
+         * 발전 과제 화면을 닫는 경우
+         */
+        CLOSED_SCREEN
+    }
+}
