From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Fri, 6 Aug 2021 20:08:04 +0900
Subject: [PATCH] Add more Player API


diff --git a/pom.xml b/pom.xml
index 7477da645959f639ec34001dc736c05d3686d45d..bda76b4b9f183c7515b3cd8bc7c13a4133ade37e 100644
--- a/pom.xml
+++ b/pom.xml
@@ -31,6 +31,10 @@
             <url>https://oss.sonatype.org/content/groups/public/</url>
         </repository>
         <!-- Daydream start -->
+        <repository>
+            <id>minecraft-repo</id>
+            <url>https://libraries.minecraft.net/</url>
+        </repository>
         <repository>
             <id>papermc</id>
             <url>https://papermc.io/repo/repository/maven-public/</url>
@@ -102,6 +106,12 @@
             <version>1.0.18</version>
             <scope>provided</scope>
         </dependency>
+        <dependency>
+            <groupId>com.mojang</groupId>
+            <artifactId>authlib</artifactId>
+            <version>2.1.28</version> <!-- keep in sync with major MC versions -->
+            <scope>compile</scope> <!-- expose to Plugins -->
+        </dependency>
         <!-- Daydream end -->
         <dependency>
             <groupId>commons-lang</groupId>
diff --git a/src/main/java/moe/caramel/daydream/network/PlayerInfoTypes.java b/src/main/java/moe/caramel/daydream/network/PlayerInfoTypes.java
new file mode 100644
index 0000000000000000000000000000000000000000..00b548c05d21bb0d898ebb39439d68c982b1cdbc
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/network/PlayerInfoTypes.java
@@ -0,0 +1,13 @@
+package moe.caramel.daydream.network;
+
+/**
+ * This is the enum type used for PlayerInfo packets.
+ */
+public enum PlayerInfoTypes {
+
+    ADD_PLAYER,
+    UPDATE_GAME_MODE,
+    UPDATE_LATENCY,
+    UPDATE_DISPLAY_NAME,
+    REMOVE_PLAYER
+}
\ No newline at end of file
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 4abbc042e9f75440a0d0d4d6cabdb329432ff028..37d0777e2d4b8c6af9389402c09f9d21b5a891c0 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1228,6 +1228,17 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     public void showPlayer(@NotNull Plugin plugin, @NotNull Player player);
 
+    // Daydream start
+    /**
+     * Resend the target's info packet to this player.
+     * It's useful when changing a player's skin.
+     *
+     * @param type The type of packet
+     * @param targets Target players
+     */
+    public void sendPlayerInfo(@NotNull moe.caramel.daydream.network.PlayerInfoTypes type, @NotNull java.util.List<Player> targets);
+    // Daydream end
+
     /**
      * Checks to see if a player has been hidden from this player
      *
@@ -1874,6 +1885,24 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
         return net.kyori.adventure.text.event.HoverEvent.showEntity(op.apply(net.kyori.adventure.text.event.HoverEvent.ShowEntity.of(this.getType().getKey(), this.getUniqueId(), this.displayName())));
     }
 
+    // Daydream start
+    /**
+     * Send a change game state packet to this player
+     *
+     * @param reason Reason Value
+     * @param value Value
+     */
+    void sendChangeGameState(int reason, int value);
+
+    /**
+     * Send a CustomPayload packet to this player
+     *
+     * @param channel channel Value
+     * @param value Value
+     */
+    void sendCustomPayload(@NotNull org.bukkit.NamespacedKey channel, @NotNull String value);
+    // Daydream end
+
     /**
      * Request that the player's client download and switch resource packs.
      * <p>
@@ -1912,6 +1941,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @Nullable
     org.bukkit.event.player.PlayerResourcePackStatusEvent.Status getResourcePackStatus();
 
+    // Daydream start
+    /**
+     * @param status resource pack status
+     */
+    void setResourcePackStatus(@NotNull org.bukkit.event.player.PlayerResourcePackStatusEvent.Status status);
+    // Daydream end
+
     /**
      * @return the most recent resource pack hash received from the player,
      *         or null if no hash has ever been received from this player.
@@ -1935,6 +1971,15 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @NotNull
     PlayerProfile getPlayerProfile();
 
+    // Daydream start
+    /**
+     * Gets this players profile
+     * @return The players profile object
+     */
+    @NotNull
+    com.mojang.authlib.GameProfile getProfile();
+    // Daydream end
+
     /**
      * Changes the PlayerProfile for this player. This will cause this player
      * to be reregistered to all clients that can currently see this player
