From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 8 Aug 2021 21:57:20 +0900
Subject: [PATCH] Allows cancellation of chunk unload packet


diff --git a/src/main/java/io/papermc/paper/event/packet/PlayerChunkUnloadEvent.java b/src/main/java/io/papermc/paper/event/packet/PlayerChunkUnloadEvent.java
index 12163a7b0591a7d022dc7eb9ee6608a1b6c39d9b..20ee966d82210a9b070fe4f9234973700055b925 100644
--- a/src/main/java/io/papermc/paper/event/packet/PlayerChunkUnloadEvent.java
+++ b/src/main/java/io/papermc/paper/event/packet/PlayerChunkUnloadEvent.java
@@ -12,7 +12,7 @@ import org.jetbrains.annotations.NotNull;
  * Should only be used for packet/clientside related stuff.
  * Not intended for modifying server side.
  */
-public class PlayerChunkUnloadEvent extends ChunkEvent {
+public class PlayerChunkUnloadEvent extends ChunkEvent implements org.bukkit.event.Cancellable { // Daydream
 
     private static final HandlerList handlers = new HandlerList();
     private final Player player;
@@ -22,6 +22,30 @@ public class PlayerChunkUnloadEvent extends ChunkEvent {
         this.player = player;
     }
 
+    // Daydream start
+    private boolean cancelled;
+
+    /**
+     * 청크 언로드 패킷이 차단 되었는지의 여부입니다.
+     *
+     * @return 청크 언로드 패킷이 차단 되었는지의 여부
+     */
+    @Override
+    public boolean isCancelled() {
+        return cancelled;
+    }
+
+    /**
+     * 청크 언로드 패킷의 차단 여부를 결정합니다.
+     *
+     * @param cancel {@code true}인 경우 청크 언로드 패킷을 차단합니다.
+     */
+    @Override
+    public void setCancelled(boolean cancel) {
+        this.cancelled = cancel;
+    }
+    // Daydream end
+
     @NotNull
     @Override
     public HandlerList getHandlers() {
