From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 8 Aug 2021 21:59:37 +0900
Subject: [PATCH] Add more PlayerTeleportEvent reason


diff --git a/src/main/java/org/bukkit/event/player/PlayerTeleportEvent.java b/src/main/java/org/bukkit/event/player/PlayerTeleportEvent.java
index 553d7740489fe729166c8ca8ef8c7834db3663ad..c614b9263f5aa9cfad20e9403d4662dd29f61cf0 100644
--- a/src/main/java/org/bukkit/event/player/PlayerTeleportEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerTeleportEvent.java
@@ -73,6 +73,12 @@ public class PlayerTeleportEvent extends PlayerMoveEvent {
          * fruit
          */
         CHORUS_FRUIT,
+        // Daydream start
+        /**
+         * Indicates the teleportation was caused by a player dismounting an entity
+         */
+        DISMOUNT,
+        // Daydream end
         /**
          * Indicates the teleportation was caused by an event not covered by
          * this enum
