From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 8 Aug 2021 03:05:54 +0900
Subject: [PATCH] Add fixed no damage tick option


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 31353bd20404a8c2acf6bf0df524dc3cae324272..32677555cd9e3fed1fb83df666df30858e942b79 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -415,6 +415,16 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      */
     public void setNoDamageTicks(int ticks);
 
+    // Daydream start
+    /**
+     * Sets the living entity's current no damage ticks.
+     *
+     * @param ticks amount of no damage ticks
+     * @param force is force?
+     */
+    public void setNoDamageTicks(int ticks, boolean force);
+    // Daydream end
+
     /**
      * Gets the player identified as the killer of the living entity.
      * <p>
