From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 8 Aug 2021 21:36:50 +0900
Subject: [PATCH] Add PlayerCPSChangeEvent


diff --git a/src/main/java/moe/caramel/daydream/event/player/PlayerCPSChangeEvent.java b/src/main/java/moe/caramel/daydream/event/player/PlayerCPSChangeEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..7aa507be636132660bd8735c855306cdc7b7cefe
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/event/player/PlayerCPSChangeEvent.java
@@ -0,0 +1,53 @@
+package moe.caramel.daydream.event.player;
+
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.jetbrains.annotations.NotNull;
+
+public class PlayerCPSChangeEvent extends PlayerEvent {
+
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+    private final int currentCPS;
+    private final int maxCPS;
+    private final KeyBind key;
+
+    public PlayerCPSChangeEvent(@NotNull Player who, int currentCPS, int maxCPS, @NotNull KeyBind key) {
+        super(who);
+        this.currentCPS = currentCPS;
+        this.maxCPS = maxCPS;
+        this.key = key;
+    }
+
+    public int getCurrentCPS() {
+        return this.currentCPS;
+    }
+
+    public int getMaxCPS() {
+        return maxCPS;
+    }
+
+    public boolean isMaxCPS() {
+        return this.currentCPS == this.maxCPS;
+    }
+
+    @NotNull
+    public KeyBind getKey() {
+        return this.key;
+    }
+
+    @Override
+    public @NotNull HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+
+    public enum KeyBind {
+        LEFT_CLICK,
+        DROP_KEY
+    }
+}
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index db81a52b2be54595ab4b52226f9563143895aa79..5b165f922de0d8bdf8fee9f81c3634008325c2ea 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2217,6 +2217,38 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     void sendOpLevel(byte level);
     // Paper end
 
+    // Daydream start
+    /**
+     * @param maxCPS set Player's max left click KPS
+     */
+    void setMaxLeftClickCPS(int maxCPS);
+
+    /**
+     * @param maxKPS set Player's max Drop key KPS
+     */
+    void setMaxDropKeyKPS(int maxKPS);
+
+    /**
+     * @return Return Player's max left click CPS
+     */
+    int getMaxLeftClickCPS();
+
+    /**
+     * @return Return Player's max drop key KPS
+     */
+    int getMaxDropKeyKPS();
+
+    /**
+     * @param delay set Player's swap key delay
+     */
+    void setSwapKeyDelay(int delay);
+
+    /**
+     * @return Return Player's swap key delay
+     */
+    int getSwapKeyDelay();
+    // Daydream end
+
     // Spigot start
     public class Spigot extends Entity.Spigot {
 
