From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Tue, 24 Aug 2021 00:10:34 +0900
Subject: [PATCH] Add caramel.moe Network Support


diff --git a/build.gradle.kts b/build.gradle.kts
index 90282b4598c4483fbdecdb50eefd90cc81f4234a..4f2b9285a6c60cd69a3b8d668f39f6d35f7e9133 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -52,6 +52,7 @@ dependencies {
     }
     // Daydream end - Add Brigadier Command API
     api("com.mojang:authlib:3.3.39") // Daydream - Mojang Authlib (keep in sync with major MC versions)
+    api("moe.caramel:acacia-user:1.0.1-proto") // Daydream - caramel.moe Network
 
     compileOnly("org.apache.maven:maven-resolver-provider:3.8.4")
     compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.2")
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index edc6e3f653a3ec7a6a2ad568b5c2b6d6c07b95ac..de7567937d85d7b896c0ac06caa794d05badc7b3 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -50,6 +50,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     @NotNull
     moe.caramel.daydream.player.InteractManager getInteractManager();
+
+    /**
+     * caramel.moe Network 유저 데이터를 가져옵니다.
+     *
+     * @return caramel.moe Network 유저 데이터
+     */
+    @NotNull moe.caramel.acacia.user.User getUserData();
     // Daydream end
 
     // Paper start
diff --git a/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java b/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java
index 635b8787fc235b61c0d5677def034656e4ec4cef..f9b83e5e53d1190869f9bf6f16b2cb08119baedf 100644
--- a/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java
+++ b/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java
@@ -240,6 +240,29 @@ public class AsyncPlayerPreLoginEvent extends Event {
         this.message = net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().deserialize(message); // Paper
     }
 
+    // Daydream start
+    private moe.caramel.acacia.user.User userData;
+
+    /**
+     * caramel.moe 네트워크 유저 데이터를 가져옵니다.
+     *
+     * @return caramel.moe 네트워크 유저 데이터
+     */
+    @NotNull
+    public moe.caramel.acacia.user.User getUserData() {
+        return userData;
+    }
+
+    /**
+     * caramel.moe 네트워크 유저 데이터를 설정합니다.
+     *
+     * @param userData caramel.moe 네트워크 유저 데이터
+     */
+    public void setUserData(@NotNull moe.caramel.acacia.user.User userData) {
+        this.userData = userData;
+    }
+    // Daydream end
+
     /**
      * Gets the player's name.
      *
diff --git a/src/main/java/org/bukkit/event/server/ServerLoadEvent.java b/src/main/java/org/bukkit/event/server/ServerLoadEvent.java
index c9a252d7e2f8c82b8ce10c02a92ada4529495f71..3876f150f428f77bffbf1b4753a5e514f7b6e2a1 100644
--- a/src/main/java/org/bukkit/event/server/ServerLoadEvent.java
+++ b/src/main/java/org/bukkit/event/server/ServerLoadEvent.java
@@ -37,6 +37,29 @@ public class ServerLoadEvent extends ServerEvent {
         return type;
     }
 
+    // Daydream start
+    private Class<? extends moe.caramel.acacia.user.User> deserializeClass;
+
+    /**
+     * Returns a class for deserialization.
+     *
+     * @return a class for deserialize
+     */
+    @org.jetbrains.annotations.Nullable
+    public Class<? extends moe.caramel.acacia.user.User> getDeserializeClass() {
+        return deserializeClass;
+    }
+
+    /**
+     * Sets the class for deserialization.
+     *
+     * @param deserializeClass class for deserialize.
+     */
+    public void setDeserializeClass(@NotNull Class<? extends moe.caramel.acacia.user.User> deserializeClass) {
+        this.deserializeClass = deserializeClass;
+    }
+    // Daydream end
+
     @NotNull
     @Override
     public HandlerList getHandlers() {
