From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 8 Aug 2021 13:37:09 +0900
Subject: [PATCH] Add Adventure interact block ignore option & Fix Adventure
 interact event


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 2accb726a5eabacbcc2e142d568be8a072c4a3dc..f3480f140784def3e0b87f4610faec9377b8475e 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1973,6 +1973,26 @@ public final class Bukkit {
     }
     // Paper end
 
+    // Daydream start
+    /**
+     * Adventure 모드에서 클릭 상호작용 이벤트를 작동하였을 때 블록 무시의 여부
+     *
+     * @return 블록을 무시하도록 설정이 되어있다면 true 를 반환합니다.
+     */
+    public static boolean isAdventureInteractBlockIgnore() {
+        return server.isAdventureInteractBlockIgnore();
+    }
+
+    /**
+     * Adventure 모드에서 클릭 상호작용 이벤트를 작동하였을 때 블록 무시의 여부를 설정합니다
+     *
+     * @param ignore 블록을 무시하도록 설정하려면 true
+     */
+    public static void setAdventureInteractBlockIgnore(boolean ignore) {
+        server.setAdventureInteractBlockIgnore(ignore);
+    }
+    // Daydream end
+
     @NotNull
     public static Server.Spigot spigot() {
         return server.spigot();
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 7aec792d9de7b8008d700e8acf526db47164cd57..919f8c8b3d59dc5016061a7b260d086c46a7b38a 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1744,4 +1744,20 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     io.papermc.paper.datapack.DatapackManager getDatapackManager();
     // Paper end
+
+    // Daydream start
+    /**
+     * Adventure 모드에서 클릭 상호작용 이벤트를 작동하였을 때 블록 무시의 여부
+     *
+     * @return 블록을 무시하도록 설정이 되어있다면 true 를 반환합니다.
+     */
+    boolean isAdventureInteractBlockIgnore();
+
+    /**
+     * Adventure 모드에서 클릭 상호작용 이벤트를 작동하였을 때 블록 무시의 여부를 설정합니다
+     *
+     * @param ignore 블록을 무시하도록 설정하려면 true
+     */
+    void setAdventureInteractBlockIgnore(boolean ignore);
+    // Daydream end
 }
