From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Mon, 18 Apr 2022 21:05:06 +0900
Subject: [PATCH] Add Network API


diff --git a/build.gradle.kts b/build.gradle.kts
index e06f984425f503bc0c4402917e91a80770f275fd..5dcd0fbdef32381256b5b517f142fb3ec42ce39c 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -69,6 +69,7 @@ dependencies {
     implementation("org.ow2.asm:asm-commons:9.7.1")
     // Paper end
     api("com.mojang:authlib:6.0.57") // Daydream - Add Brigadier Command API (Mojang Authlib; keep in sync with major MC versions)
+    api("io.netty:netty-handler:4.1.115.Final") // Daydream - Add Network API (keep in sync with major MC versions)
 
     api("org.apache.maven:maven-resolver-provider:3.9.6") // Paper - make API dependency for Paper Plugins
     compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.9.18")
diff --git a/src/main/java/com/destroystokyo/paper/network/NetworkClient.java b/src/main/java/com/destroystokyo/paper/network/NetworkClient.java
index c84ce3fc874eea3d8f0b1cf5273996d9b4af6225..f3e044b5529d7aa470c46535c2304ef2e4cc6a07 100644
--- a/src/main/java/com/destroystokyo/paper/network/NetworkClient.java
+++ b/src/main/java/com/destroystokyo/paper/network/NetworkClient.java
@@ -36,4 +36,14 @@ public interface NetworkClient {
      */
     @Nullable InetSocketAddress getVirtualHost();
 
+    // Daydream start - Add Network API
+    /**
+     * 이 클라이언트 연결의 netty 채널을 가져옵니다.
+     *
+     * @return netty 채널
+     */
+    default io.netty.channel.@Nullable Channel getChannel() {
+        return null;
+    }
+    // Daydream end - Add Network API
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 4bdb82e740c9b7f24c3f20be9174333e12dc6ce6..26fd11130544179f49cf5f684dcddbfe270dca20 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1076,6 +1076,24 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     public void sendEquipmentChange(@NotNull LivingEntity entity, @NotNull Map<EquipmentSlot, ItemStack> items);
 
+    // Daydream start - Add Network API
+    /**
+     * 이 플레이어에게 커스텀 페이로드를 전송합니다.
+     *
+     * @param channel 채널 이름
+     * @param value 데이터
+     */
+    public void sendCustomPayload(@NotNull org.bukkit.NamespacedKey channel, @NotNull byte[] value);
+
+    /**
+     * 이 플레이어에게 커스텀 페이로드를 전송합니다.
+     *
+     * @param channel 채널 이름
+     * @param buf bytebuf 오브젝트
+     */
+    public void sendCustomPayload(@NotNull org.bukkit.NamespacedKey channel, @NotNull io.netty.buffer.ByteBuf buf);
+    // Daydream end - Add Network API
+
     // Paper start
     /**
      * Send a sign change. This fakes a sign change packet for a user at
