From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Tue, 24 Aug 2021 00:10:34 +0900
Subject: [PATCH] Add caramel.moe Network Support


diff --git a/build.gradle.kts b/build.gradle.kts
index 4f8ad786591336e237d5646760dab6470b7bbea6..b8e0965479ca59023357b896d4f968f6a686d450 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -48,6 +48,8 @@ dependencies {
     // Paper end
     api("com.mojang:brigadier:1.0.18") // Daydream - Add Brigadier Command API
     api("com.mojang:authlib:3.17.30") // Daydream - Mojang Authlib (keep in sync with major MC versions)
+    compileOnly("moe.caramel:acacia-api:1.0.0-SNAPSHOT") // Daydream - caramel.moe Network
+    testRuntimeOnly("moe.caramel:acacia-api:1.0.0-SNAPSHOT") // Daydream - caramel.moe Network
 
     api("org.apache.maven:maven-resolver-provider:3.8.5") // Paper, expose
     compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.3")
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 35d93c173e63680891586b8e3f812d8ef4132d05..4bc2cd7498cebec7b152db0028fb5a00830bfcdf 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -51,6 +51,16 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     @NotNull
     moe.caramel.daydream.player.InteractManager getInteractManager();
+
+    /**
+     * caramel.moe Network 유저 데이터를 가져옵니다.
+     *
+     * @return caramel.moe Network 유저 데이터
+     */
+    @NotNull
+    default moe.caramel.acacia.user.User getUser() {
+        return moe.caramel.acacia.Acacia.getUser(this);
+    }
     // Daydream end
 
     // Paper start
