From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 8 Aug 2021 14:54:35 +0900
Subject: [PATCH] Add PlayerIllegalPacketEvent


diff --git a/src/main/java/moe/caramel/daydream/event/packet/PlayerIllegalPacketEvent.java b/src/main/java/moe/caramel/daydream/event/packet/PlayerIllegalPacketEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..092a98f21fef556ccd6eba93390be60bd6e9c6e5
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/event/packet/PlayerIllegalPacketEvent.java
@@ -0,0 +1,47 @@
+package moe.caramel.daydream.event.packet;
+
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class PlayerIllegalPacketEvent extends PlayerEvent {
+
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+    private final Action action;
+    private final String value;
+
+    public PlayerIllegalPacketEvent(@NotNull Player player, @NotNull Action action, @Nullable String value) {
+        super(player);
+        this.action = action;
+        this.value = value;
+    }
+
+    @Override
+    public @NotNull HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    public Action getAction() {
+        return action;
+    }
+
+    @Nullable
+    public String getValue() {
+        return value;
+    }
+
+    public enum Action {
+        BLOCK_DIG,
+        DIFFICULTY_CHANGE,
+        DIFFICULTY_LOCK,
+        TAB_COMPLETE
+    }
+}
\ No newline at end of file
