From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 8 Aug 2021 13:37:09 +0900
Subject: [PATCH] Add Adventure interact block ignore option & Fix Adventure
 interact event


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 7e6be8a8bdcd4c87fe938499e4a6dbc806d5bd16..d28d4b95dd7e1627920617571b9e681893b37f47 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1986,6 +1986,26 @@ public final class Bukkit {
     }
     // Paper end
 
+    // Daydream start
+    /**
+     * Adventure 모드에서 클릭 상호작용 이벤트를 작동하였을 때 블록 무시의 여부
+     *
+     * @return 블록을 무시하도록 설정이 되어있다면 true 를 반환합니다.
+     */
+    public static boolean isAdventureInteractBlockIgnore() {
+        return server.isAdventureInteractBlockIgnore();
+    }
+
+    /**
+     * Adventure 모드에서 클릭 상호작용 이벤트를 작동하였을 때 블록 무시의 여부를 설정합니다
+     *
+     * @param ignore 블록을 무시하도록 설정하려면 true
+     */
+    public static void setAdventureInteractBlockIgnore(boolean ignore) {
+        server.setAdventureInteractBlockIgnore(ignore);
+    }
+    // Daydream end
+
     @NotNull
     public static Server.Spigot spigot() {
         return server.spigot();
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 258e508d5c48a562ceb5ebf80ac82d48a93f1214..352cb1bc1080ddea7706b6964c7d8ad02713b1dc 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1755,4 +1755,20 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     io.papermc.paper.datapack.DatapackManager getDatapackManager();
     // Paper end
+
+    // Daydream start
+    /**
+     * Adventure 모드에서 클릭 상호작용 이벤트를 작동하였을 때 블록 무시의 여부
+     *
+     * @return 블록을 무시하도록 설정이 되어있다면 true 를 반환합니다.
+     */
+    boolean isAdventureInteractBlockIgnore();
+
+    /**
+     * Adventure 모드에서 클릭 상호작용 이벤트를 작동하였을 때 블록 무시의 여부를 설정합니다
+     *
+     * @param ignore 블록을 무시하도록 설정하려면 true
+     */
+    void setAdventureInteractBlockIgnore(boolean ignore);
+    // Daydream end
 }
