From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Tue, 24 Aug 2021 12:02:16 +0900
Subject: [PATCH] Add CancelCallback to BukkitRunnable


diff --git a/src/main/java/org/bukkit/scheduler/BukkitRunnable.java b/src/main/java/org/bukkit/scheduler/BukkitRunnable.java
index 35599aac72c50273abb9efe35bcb6406d9b1aba6..8345d3834e45d78b514f40065ac6bab3a111a418 100644
--- a/src/main/java/org/bukkit/scheduler/BukkitRunnable.java
+++ b/src/main/java/org/bukkit/scheduler/BukkitRunnable.java
@@ -30,6 +30,14 @@ public abstract class BukkitRunnable implements Runnable {
         Bukkit.getScheduler().cancelTask(getTaskId());
     }
 
+    // Daydream start - Add CancelCallback to BukkitRunnable
+    /**
+     * 어떠한 이유로든 스케줄이 취소되면 호출됩니다.
+     */
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public void cancelCallback() { }
+    // Daydream end - Add CancelCallback to BukkitRunnable
+
     /**
      * Schedules this in the Bukkit scheduler to run on next tick.
      *
