From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Thu, 22 Jul 2021 16:26:29 +0900
Subject: [PATCH] Add EntityRemovePotionEffectEvent


diff --git a/src/main/java/moe/caramel/daydream/event/entity/EntityRemovePotionEffectEvent.java b/src/main/java/moe/caramel/daydream/event/entity/EntityRemovePotionEffectEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..75c148c28abe3651855d56e64ff9bbc24b60fa7c
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/event/entity/EntityRemovePotionEffectEvent.java
@@ -0,0 +1,57 @@
+package moe.caramel.daydream.event.entity;
+
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.EntityType;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.entity.EntityEvent;
+import org.bukkit.potion.PotionEffectType;
+import org.jetbrains.annotations.NotNull;
+
+@Deprecated
+public class EntityRemovePotionEffectEvent extends EntityEvent implements Cancellable {
+
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+    private final PotionEffectType effect;
+    private boolean cancel = false;
+
+    public EntityRemovePotionEffectEvent(@NotNull Entity entity, @NotNull PotionEffectType effect) {
+        super(entity);
+        this.effect = effect;
+    }
+
+    @Override
+    public @NotNull Entity getEntity() {
+        return super.getEntity();
+    }
+
+    @Override
+    public @NotNull EntityType getEntityType() {
+        return super.getEntityType();
+    }
+
+    @NotNull
+    public PotionEffectType getEffect() {
+        return effect;
+    }
+
+    @Override
+    public void setCancelled(boolean cancel) {
+        this.cancel = cancel;
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return this.cancel;
+    }
+
+    @Override
+    public @NotNull HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+}
diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 5063b1c1b6710f36345b3172ac741598e59e4c7f..60fc32d8af34962303c728a862a2e2a9f873d6a9 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -473,6 +473,16 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      */
     public void removePotionEffect(@NotNull PotionEffectType type);
 
+    // Daydream start
+    /**
+     * Removes any effects present of the given {@link PotionEffectType}.
+     *
+     * @param type the potion type to remove
+     * @param force Forced regardless of true or false.
+     */
+    public void removePotionEffect(@NotNull PotionEffectType type, boolean force);
+    // Daydream end
+
     /**
      * Returns all currently active {@link PotionEffect}s on the living
      * entity.
