From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Mon, 18 Apr 2022 21:05:06 +0900
Subject: [PATCH] Add Network API


diff --git a/build.gradle.kts b/build.gradle.kts
index 3f45b3fe993e033e743e9f72b4ced20469ee1858..dabfcc0e05581219a08abe380ed37ab2e862273e 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -52,6 +52,7 @@ dependencies {
     // Paper end
     api("com.mojang:brigadier:1.1.8") // Daydream - Add Brigadier Command API
     api("com.mojang:authlib:5.0.47") // Daydream - Mojang Authlib (keep in sync with major MC versions)
+    api("io.netty:netty-all:4.1.97.Final") // Daydream - add Network API
 
     api("org.apache.maven:maven-resolver-provider:3.8.5")
     compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.3")
diff --git a/src/main/java/com/destroystokyo/paper/network/NetworkClient.java b/src/main/java/com/destroystokyo/paper/network/NetworkClient.java
index 7b2af1bd72dfbcf4e962a982940fc49b851aa04f..dad505d4e12bcdbb379485c59e476c98120d3ecd 100644
--- a/src/main/java/com/destroystokyo/paper/network/NetworkClient.java
+++ b/src/main/java/com/destroystokyo/paper/network/NetworkClient.java
@@ -38,4 +38,15 @@ public interface NetworkClient {
     @Nullable
     InetSocketAddress getVirtualHost();
 
+    // Daydream start - Add Network API
+    /**
+     * 이 클라이언트 연결의 netty 채널을 가져옵니다.
+     *
+     * @return netty 채널
+     */
+    @Nullable
+    default io.netty.channel.Channel getChannel() {
+        return null;
+    }
+    // Daydream end - Add Network API
 }
diff --git a/src/main/java/moe/caramel/daydream/network/MojangByteBuf.java b/src/main/java/moe/caramel/daydream/network/MojangByteBuf.java
new file mode 100644
index 0000000000000000000000000000000000000000..21fde661e12417b1af155cfcb747140b249e4c63
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/network/MojangByteBuf.java
@@ -0,0 +1,158 @@
+package moe.caramel.daydream.network;
+
+import io.netty.buffer.ByteBuf;
+import it.unimi.dsi.fastutil.ints.IntList;
+import net.kyori.adventure.text.Component;
+import org.bukkit.NamespacedKey;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import java.util.BitSet;
+import java.util.Collection;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.IntFunction;
+
+/**
+ * Mojang's ByteBuf wrapper class.
+ *
+ * Doesn't provide Javadoc; because the method name is specific.
+ */
+public abstract class MojangByteBuf extends ByteBuf {
+
+    public static final short MAX_STRING_LENGTH = 32767;
+    public static final int MAX_COMPONENT_STRING_LENGTH = 262144;
+
+    @NotNull
+    public abstract <T, C extends Collection<T>> C readCollection(@NotNull IntFunction<C> collectionFactory, @NotNull Function<MojangByteBuf, T> entryParser);
+
+    @NotNull
+    public abstract <T> MojangByteBuf writeCollection(@NotNull Collection<T> collection, @NotNull BiConsumer<MojangByteBuf, T> entrySerializer);
+
+    @NotNull
+    public abstract <T> List<T> readList(@NotNull Function<MojangByteBuf, T> entryParser);
+
+    @NotNull
+    public abstract IntList readIntIdList();
+
+    @NotNull
+    public abstract MojangByteBuf writeIntIdList(@NotNull IntList list);
+
+    @NotNull
+    public abstract <K, V, M extends Map<K, V>> M readMap(@NotNull IntFunction<M> mapFactory, @NotNull Function<MojangByteBuf, K> keyParser, @NotNull Function<MojangByteBuf, V> valueParser);
+
+    @NotNull
+    public abstract <K, V> Map<K, V> readMap(@NotNull Function<MojangByteBuf, K> keyParser, @NotNull Function<MojangByteBuf, V> valueParser);
+
+    @NotNull
+    public abstract <K, V> MojangByteBuf writeMap(@NotNull Map<K, V> map, @NotNull BiConsumer<MojangByteBuf, K> keySerializer, @NotNull BiConsumer<MojangByteBuf, V> valueSerializer);
+
+    @NotNull
+    public abstract void readWithCount(@NotNull Consumer<MojangByteBuf> consumer);
+
+    @NotNull
+    public abstract <T> MojangByteBuf writeOptional(@NotNull Optional<T> value, @NotNull BiConsumer<MojangByteBuf, T> serializer);
+
+    @NotNull
+    public abstract <T> Optional<T> readOptional(@NotNull Function<MojangByteBuf, T> parser);
+
+    @NotNull
+    public abstract byte[] readByteArray();
+
+    @NotNull
+    public abstract byte[] readByteArray(int maxSize);
+
+    @NotNull
+    public abstract MojangByteBuf writeByteArray(@NotNull byte[] array);
+
+    @NotNull
+    public abstract int[] readVarIntArray();
+
+    @NotNull
+    public abstract int[] readVarIntArray(int maxSize);
+
+    @NotNull
+    public abstract MojangByteBuf writeVarIntArray(@NotNull int[] array);
+
+    @NotNull
+    public abstract long[] readLongArray();
+
+    @NotNull
+    public abstract long[] readLongArray(@Nullable long[] toArray);
+
+    @NotNull
+    public abstract long[] readLongArray(@Nullable long[] toArray, int maxSize);
+
+    @NotNull
+    public abstract MojangByteBuf writeLongArray(@NotNull long[] array);
+
+    @NotNull
+    public abstract Component readComponent();
+
+    @NotNull
+    public abstract MojangByteBuf writeComponent(@NotNull Component component);
+
+    @NotNull
+    public abstract <E extends Enum<E>> E readEnum(@NotNull Class<E> enumClass);
+
+    @NotNull
+    public abstract MojangByteBuf writeEnum(@NotNull Enum<?> instance);
+
+    public abstract int readVarInt();
+
+    @NotNull
+    public abstract MojangByteBuf writeVarInt(int value);
+
+    public abstract long readVarLong();
+
+    @NotNull
+    public abstract MojangByteBuf writeVarLong(long value);
+
+    @NotNull
+    public abstract UUID readUUID();
+
+    @NotNull
+    public abstract MojangByteBuf writeUUID(@NotNull UUID uuid);
+
+    @NotNull
+    public abstract ItemStack readItem();
+
+    @NotNull
+    public abstract MojangByteBuf writeItem(@NotNull ItemStack stack);
+
+    @NotNull
+    public abstract String readUtf();
+
+    @NotNull
+    public abstract String readUtf(int maxLength);
+
+    @NotNull
+    public abstract MojangByteBuf writeUtf(@NotNull String string);
+
+    @NotNull
+    public abstract MojangByteBuf writeUtf(@NotNull String string, int maxLength);
+
+    @NotNull
+    public abstract NamespacedKey readResourceLocation();
+
+    @NotNull
+    public abstract MojangByteBuf writeResourceLocation(@NotNull NamespacedKey key);
+
+    @NotNull
+    public abstract Date readDate();
+
+    @NotNull
+    public abstract MojangByteBuf writeDate(@NotNull Date date);
+
+    @NotNull
+    public abstract BitSet readBitSet();
+
+    @NotNull
+    public abstract MojangByteBuf writeBitSet(@NotNull BitSet bitSet);
+}
diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index f569a1fd1b78ed3f3fd091bb3a0fee8c985630bb..e5861e716583d2e2ea2adafe3af57bd6209835d8 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -282,4 +282,15 @@ public interface UnsafeValues {
     // Paper end
 
     @NotNull moe.caramel.daydream.world.DamageMata getDamageMeta(@NotNull Object damageSource); // Daydream - Add More API for EntityDamageEvent
+
+    // Daydream start
+    /**
+     * Create a Mojang's ByteBuf wrapper.
+     *
+     * @param buf byteBuf object to be wrapped
+     * @return the Wrapped ByteBuf object
+     */
+    @org.jetbrains.annotations.NotNull
+    moe.caramel.daydream.network.MojangByteBuf createByteBuf(@org.jetbrains.annotations.NotNull io.netty.buffer.ByteBuf buf);
+    // Daydream end
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 6ce203131ffca7c87677d4a207d32d7b0e5df9fb..5783e00de5f5a5e67460a331c10eeb73342aa6c1 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -887,6 +887,24 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     public void sendEquipmentChange(@NotNull LivingEntity entity, @NotNull Map<EquipmentSlot, ItemStack> items);
 
+    // Daydream start
+    /**
+     * Send a custom query payload packet to this player
+     *
+     * @param channel The channel name
+     * @param value value
+     */
+    public void sendCustomPayload(@NotNull org.bukkit.NamespacedKey channel, @NotNull byte[] value);
+
+    /**
+     * Send a custom query payload packet to this player
+     *
+     * @param channel The channel name
+     * @param buf byte buf object
+     */
+    public void sendCustomPayload(@NotNull org.bukkit.NamespacedKey channel, @NotNull moe.caramel.daydream.network.MojangByteBuf buf);
+    // Daydream end
+
     // Paper start
     /**
      * Send a sign change. This fakes a sign change packet for a user at
