From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 8 Aug 2021 21:57:20 +0900
Subject: [PATCH] Allows cancellation of chunk unload packet


diff --git a/src/main/java/io/papermc/paper/event/packet/PlayerChunkUnloadEvent.java b/src/main/java/io/papermc/paper/event/packet/PlayerChunkUnloadEvent.java
index 12163a7b0591a7d022dc7eb9ee6608a1b6c39d9b..4577e7ed9ce0d49d069d670d53a2571cf707cc9b 100644
--- a/src/main/java/io/papermc/paper/event/packet/PlayerChunkUnloadEvent.java
+++ b/src/main/java/io/papermc/paper/event/packet/PlayerChunkUnloadEvent.java
@@ -2,6 +2,7 @@ package io.papermc.paper.event.packet;
 
 import org.bukkit.Chunk;
 import org.bukkit.entity.Player;
+import org.bukkit.event.Cancellable; // Daydream
 import org.bukkit.event.HandlerList;
 import org.bukkit.event.world.ChunkEvent;
 import org.jetbrains.annotations.NotNull;
@@ -12,16 +13,37 @@ import org.jetbrains.annotations.NotNull;
  * Should only be used for packet/clientside related stuff.
  * Not intended for modifying server side.
  */
-public class PlayerChunkUnloadEvent extends ChunkEvent {
+public class PlayerChunkUnloadEvent extends ChunkEvent implements Cancellable { // Daydream
 
     private static final HandlerList handlers = new HandlerList();
     private final Player player;
+    private boolean cancelled; // Daydream
 
     public PlayerChunkUnloadEvent(@NotNull Chunk chunk, @NotNull Player player) {
         super(chunk);
         this.player = player;
     }
 
+    // Daydream start
+    /**
+     * 청크 언로드 패킷이 차단 되었는지의 여부입니다.
+     * @return 청크 언로드 패킷이 차단 되었는지의 여부
+     */
+    @Override
+    public boolean isCancelled() {
+        return cancelled;
+    }
+
+    /**
+     * 청크 언로드 패킷의 차단 여부를 결정합니다.
+     * @param cancel 청크 언로드 패킷을 차단할 것인가?
+     */
+    @Override
+    public void setCancelled(boolean cancel) {
+        this.cancelled = cancel;
+    }
+    // Daydream end
+
     @NotNull
     @Override
     public HandlerList getHandlers() {
