From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Mon, 17 Jan 2022 02:35:40 +0900
Subject: [PATCH] Add GeoIP Support


diff --git a/build.gradle.kts b/build.gradle.kts
index b8e0965479ca59023357b896d4f968f6a686d450..08346ed0136e78b2f0f8481427988f5b4110ebd9 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -50,6 +50,7 @@ dependencies {
     api("com.mojang:authlib:3.17.30") // Daydream - Mojang Authlib (keep in sync with major MC versions)
     compileOnly("moe.caramel:acacia-api:1.0.0-SNAPSHOT") // Daydream - caramel.moe Network
     testRuntimeOnly("moe.caramel:acacia-api:1.0.0-SNAPSHOT") // Daydream - caramel.moe Network
+    api("com.neovisionaries:nv-i18n:1.29") // Daydream - add GeoIP
 
     api("org.apache.maven:maven-resolver-provider:3.8.5") // Paper, expose
     compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.3")
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 4bc2cd7498cebec7b152db0028fb5a00830bfcdf..ffc2a2ef5f8ca056226487f8cbd9da758862d1c3 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -61,6 +61,16 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     default moe.caramel.acacia.user.User getUser() {
         return moe.caramel.acacia.Acacia.getUser(this);
     }
+
+    /**
+     * Get the player's country.
+     *
+     * @return In the case below, return {@link com.neovisionaries.i18n.CountryCode#UNDEFINED}.
+     * <p><strong>1.</strong> if the server is not connected to the caramel.moe network.</p>
+     * <p><strong>2.</strong> if the player join the server with a local IP.</p>
+     * <p><strong>3.</strong> if Endless or Sunflower misinterpreted the player's IP.</p>
+     */
+    @NotNull com.neovisionaries.i18n.CountryCode getCountry();
     // Daydream end
 
     // Paper start
