From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Fri, 15 Jul 2022 19:20:09 +0900
Subject: [PATCH] Add caramel.moe Repository to LibraryLoader


diff --git a/src/main/java/org/bukkit/plugin/java/LibraryLoader.java b/src/main/java/org/bukkit/plugin/java/LibraryLoader.java
index e4b6f278a811acbb0070e311c5c3bdaff7b00474..e91a652f3c376da9f4628797360b17dff29aaede 100644
--- a/src/main/java/org/bukkit/plugin/java/LibraryLoader.java
+++ b/src/main/java/org/bukkit/plugin/java/LibraryLoader.java
@@ -70,7 +70,7 @@ public class LibraryLoader
         } );
         session.setReadOnly();
 
-        this.repositories = repository.newResolutionRepositories( session, Arrays.asList( new RemoteRepository.Builder( "central", "default", "https://repo.maven.apache.org/maven2" ).build() ) );
+        this.repositories = repository.newResolutionRepositories( session, Arrays.asList( new RemoteRepository.Builder( "central", "default", "https://repo.maven.apache.org/maven2" ).build(), new RemoteRepository.Builder( "caramel-repo", "default", "https://repo.caramel.moe/repository/maven-public" ).build() ) ); // Daydream - add caramel.moe Repository
     }
 
     @Nullable
@@ -118,6 +118,7 @@ public class LibraryLoader
             }
 
             jarFiles.add( url );
+            if (!org.bukkit.Bukkit.spigot().getDaydreamConfig().getBoolean("logging.ignore-library-loader", false)) // Daydream
             logger.log( Level.INFO, "[{0}] Loaded library {1}", new Object[]
             {
                 java.util.Objects.requireNonNullElseGet(desc.getPrefix(), desc::getName), file // Paper - use configured log prefix
