From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Mon, 17 Jan 2022 02:35:40 +0900
Subject: [PATCH] Add GeoIP Support


diff --git a/build.gradle.kts b/build.gradle.kts
index 4f2b9285a6c60cd69a3b8d668f39f6d35f7e9133..997ce2f408fee37bb38d2bff4e401878baac00cf 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -53,6 +53,7 @@ dependencies {
     // Daydream end - Add Brigadier Command API
     api("com.mojang:authlib:3.3.39") // Daydream - Mojang Authlib (keep in sync with major MC versions)
     api("moe.caramel:acacia-user:1.0.1-proto") // Daydream - caramel.moe Network
+    api("com.neovisionaries:nv-i18n:1.29") // Daydream - add GeoIP
 
     compileOnly("org.apache.maven:maven-resolver-provider:3.8.4")
     compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.2")
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index fcffe6de89ed2ec930ef0b3092bf9933ff860d2f..0ea96be7c18da2c7ab06b34ff0d13843536e71c5 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -57,6 +57,16 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @return caramel.moe Network 유저 데이터
      */
     @NotNull moe.caramel.acacia.user.User getUserData();
+
+    /**
+     * Get the player's country.
+     *
+     * @return In the case below, return {@link com.neovisionaries.i18n.CountryCode#UNDEFINED}.
+     * <p><strong>1.</strong> if the server is not connected to the caramel.moe network.</p>
+     * <p><strong>2.</strong> if the player join the server with a local IP.</p>
+     * <p><strong>3.</strong> if Endless or Sunflower misinterpreted the player's IP.</p>
+     */
+    @NotNull com.neovisionaries.i18n.CountryCode getCountry();
     // Daydream end
 
     // Paper start
