From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Mon, 25 Dec 2023 10:38:40 +0900
Subject: [PATCH] Add cancel callback to Folia scheduler


diff --git a/src/main/java/moe/caramel/daydream/scheduler/CallbackTask.java b/src/main/java/moe/caramel/daydream/scheduler/CallbackTask.java
new file mode 100644
index 0000000000000000000000000000000000000000..8578ab2e76edd757f2ba6b78c67809c8a604a282
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/scheduler/CallbackTask.java
@@ -0,0 +1,15 @@
+package moe.caramel.daydream.scheduler;
+
+import io.papermc.paper.threadedregions.scheduler.ScheduledTask;
+import java.util.function.Consumer;
+
+/**
+ * 테스크 취소 요청을 받았을 때 수행할 작업을 지정할 수 있습니다.
+ */
+public interface CallbackTask extends Consumer<ScheduledTask> {
+
+    /**
+     * 테스크 취소 요청을 받았을 때 수행할 작업
+     */
+    void cancelCallback();
+}
