From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 8 Aug 2021 22:10:59 +0900
Subject: [PATCH] Add more World Block API


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 4f673e9123145dc78564dc3eef0edf75795dafc2..1da1a24bfdf6840ebcbf56efae9fa08b7a989448 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -84,6 +84,14 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     public boolean lineOfSightExists(@NotNull Location from, @NotNull Location to);
     // Paper end
 
+    // Daydream start
+    /**
+     * @param boundingBox BoundingBox to be checked
+     * @return Returns {@code false} if a block exists within the {@code boundingBox} range.
+     */
+    boolean boundingBoxNoBlocksAround(@NotNull BoundingBox boundingBox);
+    // Daydream end
+
     /**
      * Gets the {@link Block} at the given coordinates
      *
diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index 70caaf05be813ba390412714ba0a39981edc2475..c3f8054a1bdf66a8b8005323b819ea8fd91fe56d 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -542,6 +542,19 @@ public interface Block extends Metadatable, net.kyori.adventure.translation.Tran
     boolean breakNaturally(@NotNull ItemStack tool, boolean triggerEffect);
     // Paper end
 
+    // Daydream start
+    /**
+     * Breaks the block and spawns items as if a player had digged it with a
+     * specific tool
+     *
+     * @param tool The tool or item in hand used for digging
+     * @param triggerEffect Play the block break particle effect and sound
+     * @param dropItem Item Drop
+     * @return true if the block was destroyed
+     */
+    boolean breakNaturally(@NotNull ItemStack tool, boolean triggerEffect, boolean dropItem);
+    // Daydream end
+
     /**
      * Simulate bone meal application to this block (if possible).
      *
