From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Tue, 24 Aug 2021 12:02:16 +0900
Subject: [PATCH] Add Individual Sidebar API


diff --git a/src/main/java/moe/caramel/daydream/sidebar/EntryBuilder.java b/src/main/java/moe/caramel/daydream/sidebar/EntryBuilder.java
new file mode 100644
index 0000000000000000000000000000000000000000..581e3a7f19cea2bc6d2b1832638be56be1a9b51a
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/sidebar/EntryBuilder.java
@@ -0,0 +1,166 @@
+package moe.caramel.daydream.sidebar;
+
+import io.papermc.paper.scoreboard.numbers.NumberFormat;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * SideBar API 사용을 위한 {@link Entry} 빌더
+ */
+public final class EntryBuilder {
+
+    public static final int MAX_ENTRY_SIZE = 15;
+    private final List<Entry> entries = new ArrayList<>(MAX_ENTRY_SIZE);
+    private EntryBuilder() { }
+
+    // ==============================================
+
+    /**
+     * Entry 빌더를 생성합니다.
+     *
+     * @return {@link EntryBuilder}
+     */
+    @NotNull
+    public static EntryBuilder builder() {
+        return new EntryBuilder();
+    }
+
+    /**
+     * 현재 {@link Entry} 리스트의 크기를 가져옵니다.
+     *
+     * @return 현재 {@link Entry} 리스트의 크기
+     */
+    public int size() {
+        return (this.entries.size());
+    }
+
+    /**
+     * {@link Entry} 리스트에 빈 줄을 추가합니다.
+     *
+     * @return this
+     */
+    @NotNull
+    public EntryBuilder blank() {
+        return blank(null);
+    }
+
+    /**
+     * {@link Entry} 리스트에 빈 줄을 추가합니다.
+     *
+     * @param format 점수 표기 방식
+     * @return this
+     */
+    @NotNull
+    public EntryBuilder blank(final @Nullable NumberFormat format) {
+        return next(Component.empty(), format);
+    }
+
+    /**
+     * {@link Entry} 리스트에 제공한 텍스트를 추가합니다.
+     *
+     * @param value 리스트에 추가할 텍스트
+     * @return this
+     */
+    @NotNull
+    public EntryBuilder next(final @NotNull String value) {
+        return next(value, null);
+    }
+
+    /**
+     * {@link Entry} 리스트에 제공한 텍스트를 추가합니다.
+     *
+     * @param value 리스트에 추가할 텍스트
+     * @param format 점수 표기 방식
+     * @return this
+     */
+    @NotNull
+    public EntryBuilder next(final @NotNull String value, final @Nullable NumberFormat format) {
+        return next(MiniMessage.miniMessage().deserialize(value), format);
+    }
+
+    /**
+     * {@link Entry} 리스트에 제공한 텍스트를 추가합니다.
+     *
+     * @param value 리스트에 추가할 텍스트
+     * @return this
+     */
+    @NotNull
+    public EntryBuilder next(final @NotNull Component value) {
+        return next(value, null);
+    }
+
+    /**
+     * {@link Entry} 리스트에 제공한 텍스트를 추가합니다.
+     *
+     * @param value 리스트에 추가할 텍스트
+     * @param format 점수 표기 방식
+     * @return this
+     */
+    @NotNull
+    public EntryBuilder next(final @NotNull Component value, final @Nullable NumberFormat format) {
+        if (entries.size() < MAX_ENTRY_SIZE) {
+            this.entries.add(new Entry(value, format));
+        }
+        return this;
+    }
+
+    // ==============================================
+
+    /**
+     * SideBar에 표시될 항목의 데이터
+     */
+    public static final class Entry {
+
+        private final Component value;
+        private final NumberFormat format;
+
+        Entry(final @NotNull Component value, final @Nullable NumberFormat callback) {
+            this.value = value;
+            this.format = callback;
+        }
+
+        @NotNull
+        public Component value() {
+            return value;
+        }
+
+        @Nullable
+        public NumberFormat numberFormat() {
+            return format;
+        }
+
+        @Override
+        public String toString() {
+            return "Entry{" +
+                   "value=" + value +
+                   ", format=" + format +
+                   '}';
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            final Entry entry = (Entry) o;
+            return value.equals(entry.value) && Objects.equals(format, entry.format);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(value, format);
+        }
+    }
+
+    // ==============================================
+
+    @NotNull @ApiStatus.Internal
+    public List<Entry> entries() {
+        return entries;
+    }
+}
diff --git a/src/main/java/moe/caramel/daydream/sidebar/SidebarDecorator.java b/src/main/java/moe/caramel/daydream/sidebar/SidebarDecorator.java
new file mode 100644
index 0000000000000000000000000000000000000000..42dde15ad6550ed64a0442b449d9f5d2c91469b9
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/sidebar/SidebarDecorator.java
@@ -0,0 +1,43 @@
+package moe.caramel.daydream.sidebar;
+
+import io.papermc.paper.scoreboard.numbers.NumberFormat;
+import net.kyori.adventure.text.Component;
+import org.bukkit.entity.Player;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * 사이드바 데코레이터
+ */
+public interface SidebarDecorator {
+
+    /**
+     * 사이드바 글로벌 포맷을 가져옵니다.
+     *
+     * @param player 대상 플레이어
+     * @return 사이드바 글로벌 포맷
+     */
+    @Nullable
+    default NumberFormat globalFormat(final @NotNull Player player) {
+        return null;
+    }
+
+    /**
+     * 플레이어별 사이드바에 표시할 제목을 가져옵니다.
+     *
+     * @param player 대상 플레이어
+     * @return 사이드바에 표시할 제목
+     */
+    @NotNull
+    Component title(final @NotNull Player player);
+
+    /**
+     * 플레이어별 사이드바에 표시할 항목 리스트를 가져옵니다.
+     * <p>{@code null}을 반환하는 경우 리스트를 업데이트하지 않습니다.</p>
+     *
+     * @param player 대상 플레이어
+     * @return 사이드바에 표시할 항목 리스트
+     */
+    @Nullable
+    EntryBuilder entries(final @NotNull Player player);
+}
diff --git a/src/main/java/moe/caramel/daydream/sidebar/SidebarProcessor.java b/src/main/java/moe/caramel/daydream/sidebar/SidebarProcessor.java
new file mode 100644
index 0000000000000000000000000000000000000000..545cafad82bc93cebe17e59c1f125ce579354b1a
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/sidebar/SidebarProcessor.java
@@ -0,0 +1,85 @@
+package moe.caramel.daydream.sidebar;
+
+import org.bukkit.entity.Player;
+import org.bukkit.plugin.Plugin;
+import org.jetbrains.annotations.NotNull;
+import java.util.Set;
+import java.util.UUID;
+
+public interface SidebarProcessor {
+
+    /**
+     * 특정 플레이어에게 개인 사이드바 기능을 활성화합니다.
+     *
+     * @param owner 소유 플러그인
+     * @param decorator 사용자 지정 {@link SidebarDecorator} 데코레이터
+     * @param holder 대상 플레이어
+     * @param updateInterval 사이드바 업데이트 시간(Tick), 0 이하일 수 없습니다.
+     * @return 이미 사용하고 있는 경우 {@code false}를 반환
+     */
+    boolean activate(final @NotNull Plugin owner, final @NotNull SidebarDecorator decorator, final @NotNull Player holder, final long updateInterval);
+
+    /**
+     * 특정 플레이어에게 개인 사이드바 기능을 비활성화합니다.
+     *
+     * @param holder 대상 플레이어
+     * @return 활성화되어있지 않은 경우 {@code false}를 반환
+     */
+    default boolean deactivate(final @NotNull Player holder) {
+        return deactivate(holder.getUniqueId());
+    }
+
+    /**
+     * 특정 플레이어에게 개인 사이드바 기능을 비활성화합니다.
+     *
+     * @param uuid 대상 플레이어의 UUID
+     * @return 활성화되어있지 않은 경우 {@code false}를 반환
+     */
+    boolean deactivate(final @NotNull UUID uuid);
+
+    /**
+     * 특정 플레이어의 개인 사이드바 기능 사용 여부를 가져옵니다.
+     *
+     * @param holder 대상 플레이어
+     * @return 개인 사이드바 기능 사용 여부
+     */
+    default boolean activated(final @NotNull Player holder) {
+        return activated(holder.getUniqueId());
+    }
+
+    /**
+     * 특정 플레이어의 개인 사이드바 기능 사용 여부를 가져옵니다.
+     *
+     * @param uuid 대상 플레이어의 UUID
+     * @return 개인 사이드바 기능 사용 여부
+     */
+    boolean activated(final @NotNull UUID uuid);
+
+    /**
+     * 특정 플레이어의 개인 사이드바 업데이트 시간을 가져옵니다.
+     *
+     * @param holder 대상 플레이어
+     * @return 개인 사이드바 업데이트 주기 (틱).
+     * {@code -1}인 경우 비활성화 상태
+     */
+    default long updateInterval(final @NotNull Player holder) {
+        return updateInterval(holder.getUniqueId());
+    }
+
+    /**
+     * 특정 플레이어의 개인 사이드바 업데이트 시간을 가져옵니다.
+     *
+     * @param uuid 대상 플레이어의 UUID
+     * @return 개인 사이드바 업데이트 주기 (틱).
+     * {@code -1}인 경우 비활성화 상태
+     */
+    long updateInterval(final @NotNull UUID uuid);
+
+    /**
+     * 개인 사이드바가 활성화된 플레이어의 UUID 목록을 가져옵니다.
+     *
+     * @return UUID 목록
+     */
+    @NotNull
+    Set<UUID> trackedPlayers();
+}
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 8b7e06b25813767822263a1bcc420c77e83ad298..3751b17a1e9f7c66019aadb99da72b659ea30a73 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3010,6 +3010,18 @@ public final class Bukkit {
     }
     // Daydream end - Add Packet API
 
+    // Daydream start - Add Individual Sidebar API
+    /**
+     * 개인 사이드바 처리기를 가져옵니다.
+     *
+     * @return 개인 사이드바 처리기
+     */
+    @NotNull
+    public static moe.caramel.daydream.sidebar.SidebarProcessor sidebarProcessor() {
+        return server.sidebarProcessor();
+    }
+    // Daydream end - Add Individual Sidebar API
+
     @NotNull
     public static Server.Spigot spigot() {
         return server.spigot();
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 0e0f2f5d34c63a94242a9381db3528f09dcda9cb..0ffd8037b8f29f55a24e010535edb0a0f509a524 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2637,4 +2637,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     <T extends moe.caramel.daydream.packet.Packet> T createPacket(@NotNull Class<T> packet);
     // Daydream end - Add Packet API
+
+    // Daydream start - Add Individual Sidebar API
+    /**
+     * 개인 사이드바 처리기를 가져옵니다.
+     *
+     * @return 개인 사이드바 처리기
+     */
+    @NotNull
+    moe.caramel.daydream.sidebar.SidebarProcessor sidebarProcessor();
+    // Daydream end - Add Individual Sidebar API
 }
