From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Tue, 24 Aug 2021 12:02:16 +0900
Subject: [PATCH] Add Individual Sidebar API


diff --git a/src/main/java/moe/caramel/daydream/sidebar/EntryBuilder.java b/src/main/java/moe/caramel/daydream/sidebar/EntryBuilder.java
new file mode 100644
index 0000000000000000000000000000000000000000..dedd724319d2e3c1258efc8717e136c597e5b564
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/sidebar/EntryBuilder.java
@@ -0,0 +1,118 @@
+package moe.caramel.daydream.sidebar;
+
+import io.papermc.paper.scoreboard.numbers.NumberFormat;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import org.jetbrains.annotations.ApiStatus;
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * SideBar API 사용을 위한 {@link Entry} 빌더
+ */
+@NullMarked
+public final class EntryBuilder {
+
+    public static final int MAX_ENTRY_SIZE = 15;
+    private final List<Entry> entries = new ArrayList<>(MAX_ENTRY_SIZE);
+    private EntryBuilder() { }
+
+    // ==============================================
+
+    /**
+     * Entry 빌더를 생성합니다.
+     *
+     * @return {@link EntryBuilder}
+     */
+    public static EntryBuilder builder() {
+        return new EntryBuilder();
+    }
+
+    /**
+     * 현재 {@link Entry} 리스트의 크기를 가져옵니다.
+     *
+     * @return 현재 {@link Entry} 리스트의 크기
+     */
+    public int size() {
+        return (this.entries.size());
+    }
+
+    /**
+     * {@link Entry} 리스트에 빈 줄을 추가합니다.
+     *
+     * @return this
+     */
+    public EntryBuilder blank() {
+        return blank(null);
+    }
+
+    /**
+     * {@link Entry} 리스트에 빈 줄을 추가합니다.
+     *
+     * @param format 점수 표기 방식
+     * @return this
+     */
+    public EntryBuilder blank(final @Nullable NumberFormat format) {
+        return next(Component.empty(), format);
+    }
+
+    /**
+     * {@link Entry} 리스트에 제공한 텍스트를 추가합니다.
+     *
+     * @param value 리스트에 추가할 텍스트
+     * @return this
+     */
+    public EntryBuilder next(final String value) {
+        return next(value, null);
+    }
+
+    /**
+     * {@link Entry} 리스트에 제공한 텍스트를 추가합니다.
+     *
+     * @param value 리스트에 추가할 텍스트
+     * @param format 점수 표기 방식
+     * @return this
+     */
+    public EntryBuilder next(final String value, final @Nullable NumberFormat format) {
+        return next(MiniMessage.miniMessage().deserialize(value), format);
+    }
+
+    /**
+     * {@link Entry} 리스트에 제공한 텍스트를 추가합니다.
+     *
+     * @param value 리스트에 추가할 텍스트
+     * @return this
+     */
+    public EntryBuilder next(final Component value) {
+        return next(value, null);
+    }
+
+    /**
+     * {@link Entry} 리스트에 제공한 텍스트를 추가합니다.
+     *
+     * @param value 리스트에 추가할 텍스트
+     * @param format 점수 표기 방식
+     * @return this
+     */
+    public EntryBuilder next(final Component value, final @Nullable NumberFormat format) {
+        if (entries.size() < MAX_ENTRY_SIZE) {
+            this.entries.add(new Entry(value, format));
+        }
+        return this;
+    }
+
+    // ==============================================
+
+    /**
+     * SideBar 표시될 항목의 데이터
+     */
+    public record Entry(Component value, @Nullable NumberFormat numberFormat) {
+    }
+
+    @ApiStatus.Internal
+    public List<Entry> entries() {
+        return entries;
+    }
+}
diff --git a/src/main/java/moe/caramel/daydream/sidebar/SidebarDecorator.java b/src/main/java/moe/caramel/daydream/sidebar/SidebarDecorator.java
new file mode 100644
index 0000000000000000000000000000000000000000..db9df8d1d97e82934103e1ab5efca465e814e8df
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/sidebar/SidebarDecorator.java
@@ -0,0 +1,41 @@
+package moe.caramel.daydream.sidebar;
+
+import io.papermc.paper.scoreboard.numbers.NumberFormat;
+import net.kyori.adventure.text.Component;
+import org.bukkit.entity.Player;
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+/**
+ * 사이드바 데코레이터
+ */
+@NullMarked
+public interface SidebarDecorator {
+
+    /**
+     * 사이드바 글로벌 포맷을 가져옵니다.
+     *
+     * @param player 대상 플레이어
+     * @return 사이드바 글로벌 포맷
+     */
+    default @Nullable NumberFormat globalFormat(final Player player) {
+        return null;
+    }
+
+    /**
+     * 플레이어별 사이드바에 표시할 제목을 가져옵니다.
+     *
+     * @param player 대상 플레이어
+     * @return 사이드바에 표시할 제목
+     */
+    Component title(final Player player);
+
+    /**
+     * 플레이어별 사이드바에 표시할 항목 리스트를 가져옵니다.
+     * <p>{@code null}을 반환하는 경우 리스트를 업데이트하지 않습니다.</p>
+     *
+     * @param player 대상 플레이어
+     * @return 사이드바에 표시할 항목 리스트
+     */
+    @Nullable EntryBuilder entries(final Player player);
+}
diff --git a/src/main/java/moe/caramel/daydream/sidebar/SidebarProcessor.java b/src/main/java/moe/caramel/daydream/sidebar/SidebarProcessor.java
new file mode 100644
index 0000000000000000000000000000000000000000..24918ce807ff99a4a0e287180dbcfdbb50820cab
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/sidebar/SidebarProcessor.java
@@ -0,0 +1,85 @@
+package moe.caramel.daydream.sidebar;
+
+import org.bukkit.entity.Player;
+import org.bukkit.plugin.Plugin;
+import org.jspecify.annotations.NullMarked;
+import java.util.Set;
+import java.util.UUID;
+
+@NullMarked
+public interface SidebarProcessor {
+
+    /**
+     * 특정 플레이어에게 개인 사이드바 기능을 활성화합니다.
+     *
+     * @param owner 소유 플러그인
+     * @param decorator 사용자 지정 {@link SidebarDecorator} 데코레이터
+     * @param holder 대상 플레이어
+     * @param updateInterval 사이드바 업데이트 시간(Tick), 0 이하일 수 없습니다.
+     * @return 이미 사용하고 있는 경우 {@code false}를 반환
+     */
+    boolean activate(final Plugin owner, final SidebarDecorator decorator, final Player holder, final long updateInterval);
+
+    /**
+     * 특정 플레이어에게 개인 사이드바 기능을 비활성화합니다.
+     *
+     * @param holder 대상 플레이어
+     * @return 활성화되어있지 않은 경우 {@code false}를 반환
+     */
+    default boolean deactivate(final Player holder) {
+        return deactivate(holder.getUniqueId());
+    }
+
+    /**
+     * 특정 플레이어에게 개인 사이드바 기능을 비활성화합니다.
+     *
+     * @param uuid 대상 플레이어의 UUID
+     * @return 활성화되어있지 않은 경우 {@code false}를 반환
+     */
+    boolean deactivate(final UUID uuid);
+
+    /**
+     * 특정 플레이어의 개인 사이드바 기능 사용 여부를 가져옵니다.
+     *
+     * @param holder 대상 플레이어
+     * @return 개인 사이드바 기능 사용 여부
+     */
+    default boolean activated(final Player holder) {
+        return activated(holder.getUniqueId());
+    }
+
+    /**
+     * 특정 플레이어의 개인 사이드바 기능 사용 여부를 가져옵니다.
+     *
+     * @param uuid 대상 플레이어의 UUID
+     * @return 개인 사이드바 기능 사용 여부
+     */
+    boolean activated(final UUID uuid);
+
+    /**
+     * 특정 플레이어의 개인 사이드바 업데이트 시간을 가져옵니다.
+     *
+     * @param holder 대상 플레이어
+     * @return 개인 사이드바 업데이트 주기 (틱).
+     * {@code -1}인 경우 비활성화 상태
+     */
+    default long updateInterval(final Player holder) {
+        return updateInterval(holder.getUniqueId());
+    }
+
+    /**
+     * 특정 플레이어의 개인 사이드바 업데이트 시간을 가져옵니다.
+     *
+     * @param uuid 대상 플레이어의 UUID
+     * @return 개인 사이드바 업데이트 주기 (틱).
+     * {@code -1}인 경우 비활성화 상태
+     */
+    long updateInterval(final UUID uuid);
+
+    /**
+     * 개인 사이드바가 활성화된 플레이어의 UUID 목록을 가져옵니다.
+     *
+     * @return UUID 목록
+     */
+    Set<UUID> trackedPlayers();
+}
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 12e385cee2060687c77a80896695c81ebb9ba22a..40cc7349ea6eca40bddba25c270acb158631dd4a 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2985,6 +2985,18 @@ public final class Bukkit {
     }
     // Daydream end - Add Packet API
 
+    // Daydream start - Add Individual Sidebar API
+    /**
+     * 개인 사이드바 처리기를 가져옵니다.
+     *
+     * @return 개인 사이드바 처리기
+     */
+    @NotNull
+    public static moe.caramel.daydream.sidebar.SidebarProcessor sidebarProcessor() {
+        return server.sidebarProcessor();
+    }
+    // Daydream end - Add Individual Sidebar API
+
     @NotNull
     public static Server.Spigot spigot() {
         return server.spigot();
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index d0f772b141cf33dd416e7ac78e8e796ff3afeefd..532bc15dab0c475edf79d257e67db17859d0bbf9 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2611,4 +2611,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     <T extends moe.caramel.daydream.packet.Packet> T createPacket(@NotNull Class<T> packet);
     // Daydream end - Add Packet API
+
+    // Daydream start - Add Individual Sidebar API
+    /**
+     * 개인 사이드바 처리기를 가져옵니다.
+     *
+     * @return 개인 사이드바 처리기
+     */
+    @NotNull
+    moe.caramel.daydream.sidebar.SidebarProcessor sidebarProcessor();
+    // Daydream end - Add Individual Sidebar API
 }
