From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nopjar <code.nopjar@gmail.com>
Date: Mon, 13 Jun 2022 11:29:23 +0200
Subject: [PATCH] Add sendTitleUpdate for inventories

https://github.com/PaperMC/Paper/pull/7979

diff --git a/src/main/java/org/bukkit/inventory/InventoryView.java b/src/main/java/org/bukkit/inventory/InventoryView.java
index 72f93377b60a3b6a08b8c8264ee7313e89c15da0..1bb51673595e3f1308d3e9525ea14e5efc58fde9 100644
--- a/src/main/java/org/bukkit/inventory/InventoryView.java
+++ b/src/main/java/org/bukkit/inventory/InventoryView.java
@@ -309,4 +309,16 @@ public interface InventoryView {
      * @param title The new title.
      */
     public void setTitle(@NotNull String title);
+
+    // Paper start - (Daydream) Add sendTitleUpdate for inventories
+    /**
+     * Sends a packet to the client to open a new inventory with the new title.
+     * <p>
+     * The ID of the inventory and the contents stay the same. This does not affect the title on the
+     * server. Methods like {@link #title()} will still return the old title.
+     *
+     * @param title the new title
+     */
+    public void sendTitleUpdate(@NotNull net.kyori.adventure.text.Component title);
+    // Paper end - (Daydream) Add sendTitleUpdate for inventories
 }
