From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LoaiDev <lolyakram11@gmail.com>
Date: Tue, 11 Jan 2022 18:01:33 +0200
Subject: [PATCH] Add Entity Pose API

https://github.com/PaperMC/Paper/pull/7318

diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index a8406e5260666b1ed48cd20cfb3e991631be52db..dc1d7062f1b808bb8c0d56bffbcbdd0b07ca0d3f 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -777,6 +777,39 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
     @NotNull
     Pose getPose();
 
+    // Paper start - Add Entity Pose API
+    /**
+     * Sets the entity's current pose.
+     *
+     * <b>Note that changing the pose only has a visual effect and doesn't
+     * change other attributes such as speed, etc.</b>
+     *
+     * @param pose New {@link Pose}
+     */
+    default void setPose(@NotNull Pose pose) {
+        setPose(pose, false);
+    }
+
+    /**
+     * Sets the entity's current pose.
+     *
+     * <b>Note that changing the pose only has a visual effect and doesn't
+     * change other attributes such as speed, etc.</b>
+     *
+     * @param pose New {@link Pose}
+     * @param fixed Whether or not allow the entity to change its own pose (e.g. sleeping).<br>
+     * But plugins can still change it.
+     */
+    void setPose(@NotNull Pose pose, boolean fixed);
+
+    /**
+     * Whether or not the entity can change its own pose (e.g. sleeping).
+     *
+     * @return if entity has fixed pose
+     */
+    boolean hasFixedPose();
+    // Paper end - Add Entity Pose API
+
     // Paper start
     /**
      * Returns if the entity is in sneak mode
