From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Fri, 4 Feb 2022 11:18:12 +0900
Subject: [PATCH] Add more Advancement API


diff --git a/src/main/java/moe/caramel/daydream/advancement/AdvancementBuilder.java b/src/main/java/moe/caramel/daydream/advancement/AdvancementBuilder.java
new file mode 100644
index 0000000000000000000000000000000000000000..58cad714be43b4fed2c506a226dcd65f9a891169
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/advancement/AdvancementBuilder.java
@@ -0,0 +1,179 @@
+package moe.caramel.daydream.advancement;
+
+import io.papermc.paper.advancement.AdvancementDisplay;
+import net.kyori.adventure.text.Component;
+import org.bukkit.Material;
+import org.bukkit.NamespacedKey;
+import org.bukkit.advancement.Advancement;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.function.Function;
+
+/**
+ * Advancement builder.
+ */
+public interface AdvancementBuilder {
+
+    /**
+     * Sets the Advancement parent.
+     *
+     * @param advancement the parent advancement
+     * @return the builder
+     */
+    @NotNull
+    AdvancementBuilder parent(@NotNull Advancement advancement);
+
+    /**
+     * Call this method if this advancement is root, and it is visible by default.
+     *
+     * @return the builder
+     */
+    @NotNull
+    AdvancementBuilder visibleRoot();
+
+    /**
+     * Set the display info of this advancement.
+     *
+     * @param icon the icon show in the frame in the advancements screen
+     * @param title the advancement title
+     * @param description the advancement description
+     * @param background the texture display behind the advancement tree when selected (root only)
+     * @param frame the {@link AdvancementDisplay.Frame}
+     * @param showToast {@code true} if a toast should be shown
+     * @param announceToChat {@code true} if a message should be sent
+     * @param hidden {@code true} if hidden
+     * @return the builder
+     */
+    @NotNull
+    default AdvancementBuilder display(
+            @NotNull Material icon, @NotNull Component title, @Nullable Component description,
+            @Nullable NamespacedKey background, @NotNull AdvancementDisplay.Frame frame,
+            boolean showToast, boolean announceToChat, boolean hidden
+    ) {
+        return this.display(icon, title, description, background, frame, 0, 0, showToast, announceToChat, hidden);
+    }
+
+    /**
+     * Set the display info of this advancement.
+     *
+     * @param icon the icon show in the frame in the advancements screen
+     * @param title the advancement title
+     * @param description the advancement description
+     * @param background the texture display behind the advancement tree when selected (root only)
+     * @param frame the {@link AdvancementDisplay.Frame}
+     * @param showToast {@code true} if a toast should be shown
+     * @param announceToChat {@code true} if a message should be sent
+     * @param hidden {@code true} if hidden
+     * @return the builder
+     */
+    @NotNull
+    default AdvancementBuilder display(
+            @NotNull ItemStack icon, @NotNull Component title, @Nullable Component description,
+            @Nullable NamespacedKey background, @NotNull AdvancementDisplay.Frame frame,
+            boolean showToast, boolean announceToChat, boolean hidden
+    ) {
+        return this.display(icon, title, description, background, frame, 0, 0, showToast, announceToChat, hidden);
+    }
+
+    /**
+     * Set the display info of this advancement.
+     *
+     * @param icon the icon show in the frame in the advancements screen
+     * @param title the advancement title
+     * @param description the advancement description
+     * @param background the texture display behind the advancement tree when selected (root only)
+     * @param frame the {@link AdvancementDisplay.Frame}
+     * @param x the x-axis of icon
+     * @param y the y-axis of icon
+     * @param showToast {@code true} if a toast should be shown
+     * @param announceToChat {@code true} if a message should be sent
+     * @param hidden {@code true} if hidden
+     * @return the builder
+     */
+    @NotNull
+    default AdvancementBuilder display(
+            @NotNull Material icon, @NotNull Component title, @Nullable Component description,
+            @Nullable NamespacedKey background, @NotNull AdvancementDisplay.Frame frame,
+            float x, float y, boolean showToast, boolean announceToChat, boolean hidden
+    ) {
+        return this.display(new ItemStack(icon), title, description, background, frame, x, y, showToast, announceToChat, hidden);
+    }
+
+    /**
+     * Set the display info of this advancement.
+     *
+     * @param icon the icon show in the frame in the advancements screen
+     * @param title the advancement title
+     * @param description the advancement description
+     * @param background the texture display behind the advancement tree when selected (root only)
+     * @param frame the {@link AdvancementDisplay.Frame}
+     * @param x the x-axis of icon
+     * @param y the y-axis of icon
+     * @param showToast {@code true} if a toast should be shown
+     * @param announceToChat {@code true} if a message should be sent
+     * @param hidden {@code true} if hidden
+     * @return the builder
+     */
+    @NotNull
+    AdvancementBuilder display(
+            @NotNull ItemStack icon, @NotNull Component title, @Nullable Component description,
+            @Nullable NamespacedKey background, @NotNull AdvancementDisplay.Frame frame,
+            float x, float y, boolean showToast, boolean announceToChat, boolean hidden
+    );
+
+    /**
+     * Add a criterion.
+     * <br>
+     * Trigger is currently unavailable.
+     *
+     * @param name criterion name
+     * @return the builder
+     */
+    @NotNull
+    AdvancementBuilder addImpossibleCriterion(@NotNull String name);
+
+    /**
+     * Set the requirements criteria player need to complete the advancement.
+     *
+     * @param and if {@code true}, player have to complete all the criteria. (if {@code false}, OR)
+     * @return this builder
+     */
+    @NotNull
+    AdvancementBuilder requirements(boolean and);
+
+    /**
+     * Set the requirements criteria player need to complete the advancement.
+     *
+     * @param requirements the requirements
+     * @return this builder
+     */
+    @NotNull
+    AdvancementBuilder requirements(@NotNull String[][] requirements);
+
+    /**
+     * Use counts for this advancement. You don't need to set the criteria.
+     *
+     * @param count max count
+     * @return this builder
+     */
+    @NotNull
+    AdvancementBuilder useCount(int count);
+
+    /**
+     * Returns whether you can build this advancement or not.
+     *
+     * @param parentProvider the parent provider
+     * @return this builder
+     */
+    boolean canBuild(@NotNull Function<NamespacedKey, Advancement> parentProvider);
+
+    /**
+     * Build it!
+     *
+     * @return the advancement
+     */
+    @NotNull
+    Advancement build();
+}
diff --git a/src/main/java/moe/caramel/daydream/event/player/PlayerAdvancementPreLoadEvent.java b/src/main/java/moe/caramel/daydream/event/player/PlayerAdvancementPreLoadEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..bc71837fb7837c3b0274c041a8404a1dca4ae347
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/event/player/PlayerAdvancementPreLoadEvent.java
@@ -0,0 +1,75 @@
+package moe.caramel.daydream.event.player;
+
+import com.google.gson.JsonElement;
+import org.bukkit.entity.Player;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Called when just before load the player's Advancement data.
+ * <br>
+ * It's usually called earlier than {@link org.bukkit.event.player.PlayerJoinEvent}.
+ */
+public final class PlayerAdvancementPreLoadEvent extends PlayerEvent implements Cancellable {
+
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+    public @NotNull HandlerList getHandlers() { return HANDLER_LIST; }
+    public static @NotNull HandlerList getHandlerList() { return HANDLER_LIST; }
+
+    private JsonElement jsonElement;
+    private boolean cancel, useDataFixer = true;
+
+    public PlayerAdvancementPreLoadEvent(@NotNull Player who) {
+        super(who);
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return this.cancel;
+    }
+
+    @Override
+    public void setCancelled(boolean cancel) {
+        this.cancel = cancel;
+    }
+
+    /**
+     * Returns the Advancement json to be load.
+     *
+     * @return the Advancement json
+     */
+    @Nullable
+    public JsonElement getJsonElement() {
+        return jsonElement;
+    }
+
+    /**
+     * Sets the Advancement json to be load.
+     *
+     * @param jsonElement the Advancement json
+     */
+    public void setJsonElement(@NotNull JsonElement jsonElement) {
+        this.jsonElement = jsonElement;
+    }
+
+    /**
+     * Returns whether to use Mojang's DataFixer or not.
+     *
+     * @return if {@code true}, use Mojang's DataFixer.
+     */
+    public boolean isUseDataFixer() {
+        return this.useDataFixer;
+    }
+
+    /**
+     * Sets whether to use Mojang's DataFixer or not.
+     *
+     * @param useDataFixer if {@code true}, use Mojang's DataFixer.
+     */
+    public void setUseDataFixer(boolean useDataFixer) {
+        this.useDataFixer = useDataFixer;
+    }
+}
diff --git a/src/main/java/moe/caramel/daydream/event/player/PlayerAdvancementSaveEvent.java b/src/main/java/moe/caramel/daydream/event/player/PlayerAdvancementSaveEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..42a3f72a73d258aa378f407c460ae05e704905e4
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/event/player/PlayerAdvancementSaveEvent.java
@@ -0,0 +1,76 @@
+package moe.caramel.daydream.event.player;
+
+import com.google.gson.JsonElement;
+import org.bukkit.entity.Player;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Called when player Advancement data is saved.
+ */
+public final class PlayerAdvancementSaveEvent extends PlayerEvent implements Cancellable {
+
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+    public @NotNull HandlerList getHandlers() { return HANDLER_LIST; }
+    public static @NotNull HandlerList getHandlerList() { return HANDLER_LIST; }
+
+    private final JsonElement jsonElement;
+    private final int dataVersion;
+    private boolean cancel, saveDataVersion = true;
+
+    public PlayerAdvancementSaveEvent(@NotNull Player who, @NotNull JsonElement jsonElement, int dataVersion) {
+        super(who);
+        this.jsonElement = jsonElement;
+        this.dataVersion = dataVersion;
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return this.cancel;
+    }
+
+    @Override
+    public void setCancelled(boolean cancel) {
+        this.cancel = cancel;
+    }
+
+    /**
+     * Returns the Advancement json to be saved.
+     *
+     * @return the Advancement json
+     */
+    @NotNull
+    public JsonElement getJsonElement() {
+        if (this.isSaveDataVersion()) jsonElement.getAsJsonObject().addProperty("DataVersion", dataVersion);
+        return jsonElement;
+    }
+
+    /**
+     * Returns the server's data version.
+     *
+     * @return the server's data version
+     */
+    public int getDataVersion() {
+        return dataVersion;
+    }
+
+    /**
+     * Returns whether to include the data version in json.
+     *
+     * @return if {@code true}, include the data version in json.
+     */
+    public boolean isSaveDataVersion() {
+        return this.saveDataVersion;
+    }
+
+    /**
+     * Sets whether to include the data version in json.
+     *
+     * @param saveDataVersion if {@code true}, include the data version in json.
+     */
+    public void setSaveDataVersion(boolean saveDataVersion) {
+        this.saveDataVersion = saveDataVersion;
+    }
+}
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 51102225ad101381f29eb963cd99c11616f17f17..7c5d6c2b52c051b89bca9c120ec896395cda2b2f 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2073,6 +2073,19 @@ public final class Bukkit {
         return server.getAdvancement(key);
     }
 
+    // Daydream start
+    /**
+     * Create the advancement specified by this key.
+     *
+     * @param key unique advancement key
+     * @return advancement builder
+     */
+    @NotNull
+    public static moe.caramel.daydream.advancement.AdvancementBuilder createAdvancement(@NotNull NamespacedKey key) {
+        return server.createAdvancement(key);
+    }
+    // Daydream end
+
     /**
      * Get an iterator through all advancements. Advancements cannot be removed
      * from this iterator,
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 070ea79ef65489c2cd5d493e9cea45a7d644de96..5d9a27a55e95dd84fe3d108a5b84242c40bd2b27 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1756,6 +1756,17 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @Nullable
     Advancement getAdvancement(@NotNull NamespacedKey key);
 
+    // Daydream start
+    /**
+     * Create the advancement specified by this key.
+     *
+     * @param key unique advancement key
+     * @return advancement builder
+     */
+    @NotNull
+    moe.caramel.daydream.advancement.AdvancementBuilder createAdvancement(@NotNull NamespacedKey key);
+    // Daydream end
+
     /**
      * Get an iterator through all advancements. Advancements cannot be removed
      * from this iterator,
diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index 4fcafddf3792b66c618f91e04d102f374de565a8..eb3397c45d37e5236155156cf3cdd49ea5e0a491 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -71,6 +71,20 @@ public interface UnsafeValues {
      */
     Advancement loadAdvancement(NamespacedKey key, String advancement);
 
+    // Daydream start
+    /**
+     * Load an advancement represented by the specified builder into the server.
+     * <br>
+     * Loaded advancements will be deleted upon server restart and reload.
+     * <br>
+     * Callers should be prepared for {@link Exception} to be thrown.
+     *
+     * @param builder representation of the advancement
+     * @return the loaded advancement or null if an error occurred
+     */
+    Advancement loadAdvancement(moe.caramel.daydream.advancement.AdvancementBuilder builder);
+    // Daydream end
+
     /**
      * Delete an advancement which was loaded and saved by
      * {@link #loadAdvancement(org.bukkit.NamespacedKey, java.lang.String)}.
diff --git a/src/main/java/org/bukkit/advancement/Advancement.java b/src/main/java/org/bukkit/advancement/Advancement.java
index 792d5ab90e9252b4f660434f72f004fade14a84c..00b2b9d119c5f8b2e69baefcfe594e09f2e93ed7 100644
--- a/src/main/java/org/bukkit/advancement/Advancement.java
+++ b/src/main/java/org/bukkit/advancement/Advancement.java
@@ -56,4 +56,19 @@ public interface Advancement extends Keyed {
     @NotNull
     Advancement getRoot();
     // Paper end
+    // Daydream start
+    /**
+     * Get the advancement uses the count.
+     *
+     * @return if the advancement uses the count, it returns {@code true}.
+     */
+    boolean useCount();
+
+    /**
+     * Get the advancement's max count.
+     *
+     * @return advancement's max count.
+     */
+    int maxCount();
+    // Daydream end
 }
diff --git a/src/main/java/org/bukkit/advancement/AdvancementProgress.java b/src/main/java/org/bukkit/advancement/AdvancementProgress.java
index f9bc179da071e7bd57cefc50d6763317fb643b74..532f51a1d537a342c0251d650cd503b100876201 100644
--- a/src/main/java/org/bukkit/advancement/AdvancementProgress.java
+++ b/src/main/java/org/bukkit/advancement/AdvancementProgress.java
@@ -67,4 +67,40 @@ public interface AdvancementProgress {
      */
     @NotNull
     Collection<String> getAwardedCriteria();
+
+    // Daydream start
+    /**
+     * Increase the Advancement Progress.
+     *
+     * @throws UnsupportedOperationException This AdvancementProgress
+     * doesn't use counts.
+     * @return true if increased, false if already reached the
+     * maximum count.
+     */
+    boolean increaseCount();
+
+    /**
+     * Decrease the Advancement Progress.
+     *
+     * @throws UnsupportedOperationException This AdvancementProgress
+     * doesn't use counts.
+     * @return true if decreased, false if already reached the
+     * minimum count.
+     */
+    boolean decreaseCount();
+
+    /**
+     * Get the advancement uses the count.
+     *
+     * @return if the advancement uses the count, it returns {@code true}.
+     */
+    boolean useCount();
+
+    /**
+     * Get the advancement's count.
+     *
+     * @return advancement's count.
+     */
+    int count();
+    // Daydream end
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index b1387d98cb7197ca0f665bdb0a21d7be26000d57..321d31aa6469b924d04555e3a3a2f713410d0d3e 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2219,6 +2219,16 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @NotNull
     public AdvancementProgress getAdvancementProgress(@NotNull Advancement advancement);
 
+    // Daydream start
+    /**
+     * Return the player's serialized advancement progress data.
+     *
+     * @return serialized advancement data
+     */
+    @NotNull
+    public com.google.gson.JsonElement serializedAdvancementProgress();
+    // Daydream end
+
     /**
      * Get the player's current client side view distance.
      * <br>
