From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nopjar <code.nopjar@gmail.com>
Date: Mon, 13 Jun 2022 11:29:23 +0200
Subject: [PATCH] Add sendTitleUpdate for inventories

https://github.com/PaperMC/Paper/pull/7979

diff --git a/src/main/java/org/bukkit/inventory/InventoryView.java b/src/main/java/org/bukkit/inventory/InventoryView.java
index 77a0cd901038405ed0e267f0432f13deccae3ab5..46c8ebfda8120a6f9e1322afd5e289e81573b0c2 100644
--- a/src/main/java/org/bukkit/inventory/InventoryView.java
+++ b/src/main/java/org/bukkit/inventory/InventoryView.java
@@ -489,4 +489,16 @@ public abstract class InventoryView {
      * @param title The new title.
      */
     public abstract void setTitle(@NotNull String title);
+
+    // Daydream start
+    /**
+     * Sends a packet to the client to open a new inventory with the new title.
+     * <p>
+     * The ID of the inventory and the contents stay the same. This does not affect the title on the
+     * server. Methods like {@link #title()} will still return the old title.
+     *
+     * @param title the new title
+     */
+    public abstract void sendTitleUpdate(@NotNull net.kyori.adventure.text.Component title);
+    // Daydream end
 }
