From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Mon, 4 Sep 2023 01:12:21 -0800
Subject: [PATCH] Add Legacy titleOverride to InventoryOpenEvent


diff --git a/src/main/java/org/bukkit/event/inventory/InventoryOpenEvent.java b/src/main/java/org/bukkit/event/inventory/InventoryOpenEvent.java
index 8e2afeab4c62724148e8bb0c83fb7eec569c7a0c..2dbe282a64eac2962474b9d922bde560d3decc73 100644
--- a/src/main/java/org/bukkit/event/inventory/InventoryOpenEvent.java
+++ b/src/main/java/org/bukkit/event/inventory/InventoryOpenEvent.java
@@ -84,6 +84,37 @@ public class InventoryOpenEvent extends InventoryEvent implements Cancellable {
     }
     // Paper end
 
+    // Daydream start
+    /**
+     * Gets the title override set by another event or null
+     * if not set.
+     *
+     * @return the title override or null
+     * @deprecated 1.19.4의 지원이 종료되었을 때 제거됩니다.
+     */
+    @Deprecated(forRemoval = true)
+    public @org.jetbrains.annotations.Nullable net.kyori.adventure.text.Component getTitleOverride() {
+        return titleOverride();
+    }
+
+    /**
+     * Sets the title override or clears the override.
+     * <p>
+     * This is only the title sent to the client in the open packet, this doesn't change
+     * the title returned by {@link InventoryView#title()}, hence "override".
+     * <p>
+     * <b>NOTE:</b> Horse inventories are a special case where setting this will
+     * have no effect. Horse inventory titles are set by the horse display name.
+     *
+     * @param titleOverride the title override or null
+     * @deprecated 1.19.4의 지원이 종료되었을 때 제거됩니다.
+     */
+    @Deprecated(forRemoval = true)
+    public void setTitleOverride(@org.jetbrains.annotations.Nullable net.kyori.adventure.text.Component titleOverride) {
+        this.titleOverride(titleOverride);
+    }
+    // Daydream end
+
     @NotNull
     @Override
     public HandlerList getHandlers() {
