From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Wed, 1 Mar 2023 00:44:47 +0900
Subject: [PATCH] Add more Projectile API


diff --git a/src/main/java/org/bukkit/entity/AbstractArrow.java b/src/main/java/org/bukkit/entity/AbstractArrow.java
index d173010d51d05928c35bb4bf5fbc08ce221ec474..238af6b22f35cb7a696bec1afa1ec51fbfcde01a 100644
--- a/src/main/java/org/bukkit/entity/AbstractArrow.java
+++ b/src/main/java/org/bukkit/entity/AbstractArrow.java
@@ -227,4 +227,20 @@ public interface AbstractArrow extends Projectile {
         CREATIVE_ONLY;
     }
     // Paper end
+
+    // Daydream start - Add more Projectile API
+    /**
+     * 대상 엔티티에게 화살 카운트를 추가할지의 여부를 설정합니다.
+     *
+     * @param use {@code false}인 경우 카운트를 추가하지 않음
+     */
+    void setUseEntityArrowCount(boolean use);
+
+    /**
+     * 대상 엔티티에게 화살 카운트를 추가할지의 여부를 가져옵니다.
+     *
+     * @return {@code false}인 경우 카운트를 추가하지 않음
+     */
+    boolean useEntityArrowCount();
+    // Daydream end - Add more Projectile API
 }
diff --git a/src/main/java/org/bukkit/entity/Firework.java b/src/main/java/org/bukkit/entity/Firework.java
index 217d348ad0bbef720b25d3b507a55ca8105b7731..71c8ff8b53795e20096fac9bdab67681fa1d9c8c 100644
--- a/src/main/java/org/bukkit/entity/Firework.java
+++ b/src/main/java/org/bukkit/entity/Firework.java
@@ -185,4 +185,20 @@ public interface Firework extends Projectile {
      */
     void setTicksToDetonate(int ticks);
     // Paper end
+
+    // Daydream start - Add more Projectile API
+    /**
+     * 폭발 대미지 제거 여부를 가져옵니다.
+     *
+     * @return 폭발 대미지 제거 여부
+     */
+    boolean isDisableExplosionDeal();
+
+    /**
+     * 폭발 대미지 제거 여부를 설정합니다.
+     *
+     * @param disable 폭발 대미지 제거 여부
+     */
+    void setDisableExplosionDeal(boolean disable);
+    // Daydream end - Add more Projectile API
 }
