From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 8 Aug 2021 21:49:16 +0900
Subject: [PATCH] Add Force update entity metadata API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 350384ba10435b1115e4386a78d8145f748b6b29..861c9b260926aa63da1dad7a6264a51393fd67b8 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -795,4 +795,11 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     public boolean spawnAt(@NotNull Location location, @NotNull org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason);
     // Paper end
+
+    // Daydream start
+    /**
+     * Send updated Entity Metadata packet to tracked player.
+     */
+    public void updateMetadata();
+    // Daydream end
 }
