From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Tue, 26 Jul 2022 10:27:57 +0900
Subject: [PATCH] Add Packet Invisible Entity API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index c53fbc05de52137922fa1e21901b304033a1f9f3..79a75cbb9cc86dbe5d594f4c501dbb906ad8e943 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1078,4 +1078,28 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     @NotNull String getScoreboardEntryName();
     // Paper end - entity scoreboard name
+
+    // Daydream start - Add Packet Invisible API
+    /**
+     * 엔티티가 생성되었는지 확인합니다.
+     *
+     * @return 만약 엔티티가 생성되고 패킷을 한 번이라도
+     * 전송하려고 한 경우 {@code true}를 반환합니다.
+     */
+    boolean isInitialized();
+
+    /**
+     * 이 엔티티가 패킷 숨기기를 사용하는지 확인합니다.
+     *
+     * @return 만약 이 엔티티가 패킷 숨기기를 사용하는 경우 {@code true}를 반환합니다.
+     */
+    boolean usePacketInvisible();
+
+    /**
+     * 패킷 숨기기 사용 여부를 정합니다.
+     *
+     * @param invisible {@code true}인 경우, 패킷 숨기기를 사용하며 이 엔티티와 관련된 모든 패킷이 전송되지 않습니다.
+     */
+    void setPacketInvisible(boolean invisible);
+    // Daydream end - Add Packet Invisible API
 }
