From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 8 Aug 2021 21:49:16 +0900
Subject: [PATCH] Add Force update entity metadata API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index a621ce84ae379352a8e11a60239e366915a3136c..459e2524d9621e07d973d02b052113ab793fc92a 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -712,4 +712,11 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     public boolean isTicking();
     // Paper end
+
+    // Daydream start
+    /**
+     * Send updated Entity Metadata packet to tracked player.
+     */
+    public void updateMetadata();
+    // Daydream end
 }
