From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 3 Oct 2021 17:47:34 +0900
Subject: [PATCH] Add Illegal Movement Packet API


diff --git a/src/main/java/moe/caramel/daydream/event/packet/PlayerIllegalPacketEvent.java b/src/main/java/moe/caramel/daydream/event/packet/PlayerIllegalPacketEvent.java
index 9eb89c99490b7ed9196e4d84e126f0d10cb938ed..f79054f0019aee3e93e9d6b6186c84a392a1e02b 100644
--- a/src/main/java/moe/caramel/daydream/event/packet/PlayerIllegalPacketEvent.java
+++ b/src/main/java/moe/caramel/daydream/event/packet/PlayerIllegalPacketEvent.java
@@ -42,6 +42,7 @@ public class PlayerIllegalPacketEvent extends PlayerEvent {
         BLOCK_DIG,
         DIFFICULTY_CHANGE,
         DIFFICULTY_LOCK,
-        TAB_COMPLETE
+        TAB_COMPLETE,
+        MOVE
     }
 }
