From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Wed, 5 Apr 2023 00:53:49 +0900
Subject: [PATCH] Add ItemStack Utilities


diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index 13b1902c7d289dfae60a453aab248d308459aded..ba16e4ab7e5bef33e5fbd79106d2dabdf7e566ea 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -304,4 +304,18 @@ public interface UnsafeValues {
     @NotNull java.util.List<net.kyori.adventure.text.Component> computeTooltipLines(@NotNull ItemStack itemStack, @NotNull io.papermc.paper.inventory.tooltip.TooltipContext tooltipContext, @Nullable org.bukkit.entity.Player player); // Paper - expose itemstack tooltip lines
 
     @NotNull moe.caramel.daydream.world.DamageMata getDamageMeta(@NotNull Object damageSource); // Daydream - Add more Damage API
+
+    // Daydream start - Add ItemStack Utilities
+    /**
+     * Create a new ItemStack
+     * (for Paper Compatibility)
+     *
+     * @param material the material
+     * @param count item count
+     * @return ItemStack
+     * @see <a href=https://github.com/PaperMC/Paper/pull/8711>here</a>
+     */
+    @org.jetbrains.annotations.NotNull
+    ItemStack newItem(@org.jetbrains.annotations.NotNull org.bukkit.Material material, int count);
+    // Daydream end - Add ItemStack Utilities
 }
diff --git a/src/main/java/org/bukkit/inventory/ItemStack.java b/src/main/java/org/bukkit/inventory/ItemStack.java
index 7414b4fa690d393a8e9557cc1fd1ce12fa426940..e8413bf04ff118b918d9c9f3421cdd017e8d3c3c 100644
--- a/src/main/java/org/bukkit/inventory/ItemStack.java
+++ b/src/main/java/org/bukkit/inventory/ItemStack.java
@@ -34,6 +34,99 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, Translat
     @Utility
     protected ItemStack() {}
 
+    // Daydream start - Add ItemStack Utilities
+    /**
+     * Defaults stack size to 1, with no extra data.
+     * <p>
+     * <b>IMPORTANT: An <i>Item</i>Stack is only designed to contain
+     * <i>items</i>. Do not use this class to encapsulate Materials for which
+     * {@link Material#isItem()} returns false.</b>
+     *
+     * @param type item material
+     */
+    @NotNull
+    public static ItemStack create(final @NotNull Material type) {
+        return Bukkit.getUnsafe().newItem(type, 1);
+    }
+
+    /**
+     * Create Item stack.
+     *
+     * @param type item material
+     * @param consumer the meta consumer
+     * @return the item stack
+     */
+    @NotNull
+    public static ItemStack create(final @NotNull Material type, final @NotNull java.util.function.Consumer<? super ItemMeta> consumer) {
+        final ItemStack item = Bukkit.getUnsafe().newItem(type, 1);
+        item.editMeta(consumer);
+        return item;
+    }
+
+    /**
+     * Create Item stack.
+     *
+     * @param type item material
+     * @param metaClass the type of meta to edit
+     * @param consumer the meta consumer
+     * @param <M> the meta type
+     * @return the item stack
+     */
+    @NotNull
+    public static <M extends ItemMeta> ItemStack create(final @NotNull Material type, final @NotNull Class<M> metaClass, final @NotNull java.util.function.Consumer<@NotNull ? super M> consumer) {
+        final ItemStack item = Bukkit.getUnsafe().newItem(type, 1);
+        item.editMeta(metaClass, consumer);
+        return item;
+    }
+
+    /**
+     * An item stack with no extra data.
+     * <p>
+     * <b>IMPORTANT: An <i>Item</i>Stack is only designed to contain
+     * <i>items</i>. Do not use this class to encapsulate Materials for which
+     * {@link Material#isItem()} returns false.</b>
+     *
+     * @param type item material
+     * @param amount stack size
+     */
+    @NotNull
+    public static ItemStack create(final @NotNull Material type, final int amount) {
+        return Bukkit.getUnsafe().newItem(type, amount);
+    }
+
+    /**
+     * Create Item stack.
+     *
+     * @param type item material
+     * @param amount stack size
+     * @param consumer the meta consumer
+     * @return the item stack
+     */
+    @NotNull
+    public static ItemStack create(final @NotNull Material type, final int amount, final @NotNull java.util.function.Consumer<? super ItemMeta> consumer) {
+        final ItemStack item = Bukkit.getUnsafe().newItem(type, amount);
+        item.editMeta(consumer);
+        return item;
+    }
+
+    /**
+     * Create Item stack.
+     *
+     * @param type item material
+     * @param amount stack size
+     * @param metaClass the type of meta to edit
+     * @param consumer the meta consumer
+     * @param <M> the meta type
+     * @return the item stack
+     */
+    @NotNull
+    public static <M extends ItemMeta> ItemStack create(final @NotNull Material type, final int amount, final @NotNull Class<M> metaClass, final @NotNull java.util.function.Consumer<@NotNull ? super M> consumer) {
+        final ItemStack item = Bukkit.getUnsafe().newItem(type, amount);
+        item.editMeta(metaClass, consumer);
+        return item;
+    }
+    // Daydream end - Add ItemStack Utilities
+
     /**
      * Defaults stack size to 1, with no extra data.
      * <p>
diff --git a/src/main/java/org/bukkit/inventory/meta/ItemMeta.java b/src/main/java/org/bukkit/inventory/meta/ItemMeta.java
index 478b6d2acbc5788f72db81f387daa08df7837a5d..9826eb48ec6f421ce4ae89a369b6415bac06e8d3 100644
--- a/src/main/java/org/bukkit/inventory/meta/ItemMeta.java
+++ b/src/main/java/org/bukkit/inventory/meta/ItemMeta.java
@@ -97,6 +97,47 @@ public interface ItemMeta extends Cloneable, ConfigurationSerializable, Persiste
     @Deprecated
     void setDisplayNameComponent(@Nullable net.md_5.bungee.api.chat.BaseComponent[] component);
     // Paper end
+
+    // Daydream start - Add ItemStack Utilities (from 1.20.5)
+    /**
+     * Checks for existence of an item name.
+     * <br>
+     * Item name differs from display name in that it is cannot be edited by an
+     * anvil, is not styled with italics, and does not show labels.
+     *
+     * @return true if this has an item name
+     */
+    default boolean hasItemName() {
+        return hasDisplayName();
+    }
+
+    /**
+     * Gets the item name component that is set.
+     * <br>
+     * Item name differs from display name in that it is cannot be edited by an
+     * anvil, is not styled with italics, and does not show labels.
+     * <p>
+     * Plugins should check that {@link #hasItemName()} returns <code>true</code> before
+     * calling this method.
+     *
+     * @return the item name that is set
+     * @see #hasItemName()
+     */
+    default @NotNull Component itemName() {
+        return java.util.Objects.requireNonNullElseGet(this.displayName(), Component::empty);
+    }
+
+    /**
+     * Sets the item name.
+     * <br>
+     * Item name differs from display name in that it is cannot be edited by an
+     * anvil, is not styled with italics, and does not show labels.
+     *
+     * @param name the name to set, null to remove it
+     */
+    void itemName(@Nullable final Component name);
+    // Daydream end - Add ItemStack Utilities (from 1.20.5)
+
     /**
      * Checks for existence of a localized name.
      *
@@ -153,6 +194,170 @@ public interface ItemMeta extends Cloneable, ConfigurationSerializable, Persiste
     void lore(final @Nullable List<? extends net.kyori.adventure.text.Component> lore);
     // Paper end
 
+    // Daydream start - Add ItemStack Utilities
+    /**
+     * 현재 로어의 크기를 가져옵니다.
+     *
+     * @return 로어의 크기
+     */
+    int loreSize();
+
+    /**
+     * Gets the lore.
+     *
+     * @return the lore
+     */
+    @NotNull List<net.kyori.adventure.text.Component> loreNonNull();
+
+    /**
+     * Sets the lore.
+     *
+     * @param style style for all lore
+     * @param lore the lore to set
+     */
+    void lore(final @NotNull net.kyori.adventure.text.format.Style style, final @Nullable List<? extends net.kyori.adventure.text.Component> lore);
+
+    /**
+     * Sets the lore excluding italic formatting.
+     *
+     * @param lore the lore to set
+     */
+    void loreNoItalic(final @Nullable List<? extends net.kyori.adventure.text.Component> lore);
+
+    /**
+     * Sets the lore excluding italic formatting.
+     *
+     * @param color color for all lore
+     * @param lore the lore to set
+     */
+    void loreNoItalic(final @Nullable net.kyori.adventure.text.format.TextColor color, final @Nullable List<? extends net.kyori.adventure.text.Component> lore);
+
+    /**
+     * Sets the lore excluding client style formatting.
+     *
+     * @param lore the lore to set
+     */
+    void loreDefault(final @Nullable List<? extends net.kyori.adventure.text.Component> lore);
+
+    /**
+     * Sets the lore.
+     *
+     * @param lore the lore to set
+     */
+    void withLore(final @Nullable net.kyori.adventure.text.ComponentLike... lore);
+
+    /**
+     * Sets the lore.
+     *
+     * @param style style for all lore
+     * @param lore the lore to set
+     */
+    void withLore(final @NotNull net.kyori.adventure.text.format.Style style, final @Nullable net.kyori.adventure.text.ComponentLike... lore);
+
+    /**
+     * Sets the lore excluding italic formatting.
+     *
+     * @param lore the lore to set
+     */
+    void withLoreNoItalic(final @Nullable net.kyori.adventure.text.ComponentLike... lore);
+
+    /**
+     * Sets the lore excluding italic formatting.
+     *
+     * @param color color for all lore
+     * @param lore the lore to set
+     */
+    void withLoreNoItalic(final @Nullable net.kyori.adventure.text.format.TextColor color, final @Nullable net.kyori.adventure.text.ComponentLike... lore);
+
+    /**
+     * Sets the lore excluding client style formatting.
+     *
+     * @param lore the lore to set
+     */
+    void withLoreDefault(final @Nullable net.kyori.adventure.text.ComponentLike... lore);
+
+    /**
+     * 로어를 추가합니다.
+     *
+     * @param lore 추가할 로어 리스트
+     */
+    void appendLore(final @NotNull List<? extends net.kyori.adventure.text.ComponentLike> lore);
+
+    /**
+     * 로어를 추가합니다.
+     *
+     * @param style 모든 줄에 적용할 스타일
+     * @param lore 추가할 로어 리스트
+     */
+    void appendLore(final @NotNull net.kyori.adventure.text.format.Style style, final @NotNull List<? extends net.kyori.adventure.text.ComponentLike> lore);
+
+    /**
+     * 로어를 이탤릭체를 비활성화하고 추가합니다.
+     *
+     * @param lore 추가할 로어 리스트
+     */
+    void appendLoreNoItalic(final @NotNull List<? extends net.kyori.adventure.text.ComponentLike> lore);
+
+    /**
+     * 로어를 이탤릭체를 비활성화하고 추가합니다.
+     *
+     * @param color 모든 줄에 적용할 문자 색
+     * @param lore 추가할 로어 리스트
+     */
+    void appendLoreNoItalic(final @Nullable net.kyori.adventure.text.format.TextColor color, final @NotNull List<? extends net.kyori.adventure.text.ComponentLike> lore);
+
+    /**
+     * 로어를 추가하며 클라이언트의 로어 수정을 비활성화합니다.
+     *
+     * @param lore 추가할 로어 리스트
+     */
+    void appendLoreDefault(final @NotNull List<? extends net.kyori.adventure.text.ComponentLike> lore);
+
+    /**
+     * 로어를 추가합니다.
+     *
+     * @param lore 추가할 로어
+     */
+    void appendLore(final @NotNull net.kyori.adventure.text.ComponentLike... lore);
+
+    /**
+     * 로어를 추가합니다.
+     *
+     * @param style 모든 줄에 적용할 스타일
+     * @param lore 추가할 로어
+     */
+    void appendLore(final @NotNull net.kyori.adventure.text.format.Style style, final @NotNull net.kyori.adventure.text.ComponentLike... lore);
+
+    /**
+     * 로어를 이탤릭체를 비활성화하고 추가합니다.
+     *
+     * @param lore 추가할 로어
+     */
+    void appendLoreNoItalic(final @NotNull net.kyori.adventure.text.ComponentLike... lore);
+
+    /**
+     * 로어를 이탤릭체를 비활성화하고 추가합니다.
+     *
+     * @param color 모든 줄에 적용할 문자 색
+     * @param lore 추가할 로어
+     */
+    void appendLoreNoItalic(final @Nullable net.kyori.adventure.text.format.TextColor color, final @NotNull net.kyori.adventure.text.ComponentLike... lore);
+
+    /**
+     * 로어를 추가하며 클라이언트의 로어 수정을 비활성화합니다.
+     *
+     * @param lore 추가할 로어
+     */
+    void appendLoreDefault(final @NotNull net.kyori.adventure.text.ComponentLike... lore);
+
+    /**
+     * 특정 위치의 로어를 제거합니다.
+     *
+     * @param index 로어 인덱스
+     */
+    void removeLore(final int index);
+    // Daydream end - Add ItemStack Utilities
+
     /**
      * Gets the lore that is set.
      * <p>
@@ -222,6 +427,18 @@ public interface ItemMeta extends Cloneable, ConfigurationSerializable, Persiste
      */
     int getCustomModelData();
 
+    // Daydream start - Add ItemStack Utilities
+    /**
+     * Gets the custom model data that is set.
+     * <p>
+     * CustomModelData is an integer that may be associated client side with a
+     * custom item model.
+     *
+     * @return the custom model data that is set
+     */
+    @Nullable Integer customModelData();
+    // Daydream end - Add ItemStack Utilities
+
     /**
      * Sets the custom model data.
      * <p>
@@ -344,6 +561,23 @@ public interface ItemMeta extends Cloneable, ConfigurationSerializable, Persiste
      */
     void setUnbreakable(boolean unbreakable);
 
+    // Daydream start - Add ItemStack Utilities
+    /**
+     * 아이템의 빛남 여부를 가져옵니다.
+     *
+     * @return 만약 {@code true}라면 아이템이 빛남
+     */
+    boolean isGlowing();
+
+    /**
+     * 아이템의 빛남 여부를 설정합니다. 일반적으로 허용하지 않는 방법을 통해 아이템에 글로우 효과를 추가하므로,
+     * 여러 문제가 발생할 수 있습니다. 이 기능을 사용하는 경우 모든 인첸트가 해제되며, 인첸트가 추가되는 경우 해제됩니다.
+     *
+     * @param glowing 만약 {@code true}라면 빛나게 설정
+     */
+    void setGlowing(boolean glowing);
+    // Daydream end - Add ItemStack Utilities
+
     /**
      * Checks for the existence of any AttributeModifiers.
      *
