From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Thu, 9 Mar 2023 19:46:01 +0900
Subject: [PATCH] Add more Team API


diff --git a/src/main/java/moe/caramel/daydream/team/TeamListener.java b/src/main/java/moe/caramel/daydream/team/TeamListener.java
new file mode 100644
index 0000000000000000000000000000000000000000..6f695d6f2c82d7628ab2d0b64cf5acaacd82b1c8
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/team/TeamListener.java
@@ -0,0 +1,53 @@
+package moe.caramel.daydream.team;
+
+import org.bukkit.OfflinePlayer;
+import org.jspecify.annotations.NullMarked;
+
+/**
+ * {@link org.bukkit.scoreboard.Team}의 변경 사항을 리스닝합니다.
+ */
+@NullMarked
+public interface TeamListener {
+
+    /**
+     * 팀이 제거되었을 때
+     */
+    default void unregister() {
+    }
+
+    /**
+     * 엔트리가 팀에 추가되었을 때
+     *
+     * @param entry 엔트리
+     */
+    default void addEntry(final String entry) {
+    }
+
+    /**
+     * 플레이어가 팀에 추가되었을 때
+     * <br>
+     * {@link #addEntry(String)}와 동시에 호출될 수 있습니다.
+     *
+     * @param player 플레이어
+     */
+    default void addPlayer(final OfflinePlayer player) {
+    }
+
+    /**
+     * 엔트리가 팀에서 제거되었을 때
+     *
+     * @param entry 엔트리
+     */
+    default void removeEntry(final String entry) {
+    }
+
+    /**
+     * 플레이어가 팀에서 제거되었을 때
+     * <br>
+     * {@link #addEntry(String)}와 동시에 호출될 수 있습니다.
+     *
+     * @param player 플레이어
+     */
+    default void removePlayer(final OfflinePlayer player) {
+    }
+}
diff --git a/src/main/java/org/bukkit/scoreboard/Score.java b/src/main/java/org/bukkit/scoreboard/Score.java
index fba8e475c1f1a410c44a95fcc474cce19e0f515c..11519711bb99531f2ddca4f3533b10324e32d8cb 100644
--- a/src/main/java/org/bukkit/scoreboard/Score.java
+++ b/src/main/java/org/bukkit/scoreboard/Score.java
@@ -151,4 +151,24 @@ public interface Score {
     void numberFormat(@Nullable io.papermc.paper.scoreboard.numbers.NumberFormat format);
     // Paper end - number format api
 
+    // Daydream start - Add more Team API (Score)
+    /**
+     * 점수 업데이트 패킷 비활성화 여부를 가져옵니다.
+     *
+     * @return 패킷 비활성화 여부
+     */
+    boolean updateDisabled();
+
+    /**
+     * 점수 업데이트 패킷 비활성화 여부를 설정합니다.
+     *
+     * @param disable {@code true}인 경우 점수 업데이트 패킷을 전송하지 않습니다.
+     */
+    void updateDisable(final boolean disable);
+
+    /**
+     * 강제로 점수 업데이트 패킷을 전송합니다.
+     */
+    void forceSendScoreToPlayers();
+    // Daydream end - Add more Team API (Score)
 }
diff --git a/src/main/java/org/bukkit/scoreboard/Team.java b/src/main/java/org/bukkit/scoreboard/Team.java
index cacb58d25c249e2ecd6083ed0f30d5ffb345220a..5e7266df82dfe794342ca855fd3203c096e92ab4 100644
--- a/src/main/java/org/bukkit/scoreboard/Team.java
+++ b/src/main/java/org/bukkit/scoreboard/Team.java
@@ -514,6 +514,46 @@ public interface Team extends net.kyori.adventure.audience.ForwardingAudience {
     boolean hasEntity(@NotNull org.bukkit.entity.Entity entity) throws IllegalStateException, IllegalArgumentException;
     // Paper end - improve scoreboard entries
 
+    // Daydream start - Add more Team API
+    /**
+     * 팀 리스너를 가져옵니다.
+     *
+     * @return 팀 리스너
+     */
+    @Nullable
+    moe.caramel.daydream.team.TeamListener getListener();
+
+    /**
+     * 팀 리스너를 설정합니다.
+     *
+     * @param listener 팀 리스너
+     */
+    void setListener(@Nullable moe.caramel.daydream.team.TeamListener listener);
+
+    /**
+     * This puts a collection of players onto this team for the scoreboard which results in one
+     * packet for the updates rather than a packet-per-entity.
+     * <p>
+     * Players on other teams will be removed from their respective teams.
+     *
+     * @param players the players to add
+     * @throws IllegalArgumentException if players are null
+     * @throws IllegalStateException if this team has been unregistered
+     */
+    void addPlayers(@NotNull java.util.Collection<? extends OfflinePlayer> players);
+
+    /**
+     * Removes a collection of players from this team  which results in one
+     * packet for the updates rather than a packet-per-entity.
+     *
+     * @param players the entries to remove
+     * @return if any of the players were a part of this team
+     * @throws IllegalArgumentException if players is null
+     * @throws IllegalStateException if this team has been unregistered
+     */
+    boolean removePlayers(@NotNull java.util.Collection<? extends OfflinePlayer> players);
+    // Daydream end - Add more Team API
+
     /**
      * Represents an option which may be applied to this team.
      */
