From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 9 Apr 2023 14:33:20 +0900
Subject: [PATCH] Add Event Metadata API


diff --git a/src/main/java/org/bukkit/event/Event.java b/src/main/java/org/bukkit/event/Event.java
index 8ec56cd6b8e0f5c5dd8c7c88b4671e18dcf109d0..7791aa0ca77acbfc9486c7ca04078f8613465d5f 100644
--- a/src/main/java/org/bukkit/event/Event.java
+++ b/src/main/java/org/bukkit/event/Event.java
@@ -51,6 +51,20 @@ public abstract class Event {
     }
     // Paper end
 
+    // Daydream start - Add Event Metadata API
+    private final java.util.Map<String, Object> metadata = new java.util.HashMap<>();
+
+    /**
+     * Gets the metadata of this event.
+     *
+     * @return The event metadata map.
+    */
+    @NotNull
+    public java.util.Map<String, Object> getEventMetadata() {
+        return metadata;
+    }
+    // Daydream end - Add Event Metadata API
+
     /**
      * Convenience method for providing a user-friendly identifier. By
      * default, it is the event's class's {@linkplain Class#getSimpleName()
