From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Tue, 23 Nov 2021 04:12:52 +0900
Subject: [PATCH] Backport "Add Player#sendEquipmentChange"


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 0c4425ce42c8960b24bafa37200fd9cafd30a25e..99cdbe8164d3f04ef0c7ea9f9313ac1bbe8b7662 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -536,6 +536,17 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @param packet The packet instance
      */
     public void sendPacket(@NotNull Object packet);
+
+    /**
+     * Send the equipment change of an entity. This fakes the equipment change
+     * of an entity for a user. This will not actually change the inventory of
+     * the specified entity in any way.
+     *
+     * @param entity The entity that the player will see the change for
+     * @param slot The slot of the spoofed equipment change
+     * @param item The ItemStack to display for the player
+     */
+    public void sendEquipmentChange(@NotNull LivingEntity entity, @NotNull org.bukkit.inventory.EquipmentSlot slot, @NotNull ItemStack item);
     // Daydream end
 
     /**
