From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Tue, 23 Nov 2021 04:12:52 +0900
Subject: [PATCH] Backport "Add Player#sendEquipmentChange"


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index cbb4395d6861afec07a1126b450b76a556b3b302..5f9a8dc4e2518c050aec7f23a16fd9e650558711 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -561,6 +561,19 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     public void sendBlockDamage(@NotNull Location loc, float progress);
 
+    // Daydream start
+    /**
+     * Send the equipment change of an entity. This fakes the equipment change
+     * of an entity for a user. This will not actually change the inventory of
+     * the specified entity in any way.
+     *
+     * @param entity The entity that the player will see the change for
+     * @param slot The slot of the spoofed equipment change
+     * @param item The ItemStack to display for the player
+     */
+    public void sendEquipmentChange(@NotNull LivingEntity entity, @NotNull org.bukkit.inventory.EquipmentSlot slot, @NotNull ItemStack item);
+    // Daydream end
+
     /**
      * Send a chunk change. This fakes a chunk change packet for a user at a
      * certain location. The updated cuboid must be entirely within a single
