From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 8 Aug 2021 23:11:09 +0900
Subject: [PATCH] Add check range BoundingBox in block method


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 4f673e9123145dc78564dc3eef0edf75795dafc2..2b087b5ed25d0bd2135e63ddb5e7942df94aa109 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -84,6 +84,14 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     public boolean lineOfSightExists(@NotNull Location from, @NotNull Location to);
     // Paper end
 
+    // Daydream start
+    /**
+     * @param boundingBox 체크할 범위의 BoundingBox
+     * @return {@code boundingBox} 범위 내에 블록이 존재한다면 {@code true}를 반환합니다.
+     */
+    boolean checkBoundingBoxRangeInBlock(@NotNull BoundingBox boundingBox);
+    // Daydream end
+
     /**
      * Gets the {@link Block} at the given coordinates
      *
