From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Fri, 15 Jul 2022 19:24:56 +0900
Subject: [PATCH] Add sendRichMessage and sendPlainMessage methods

Backport PaperMC/Paper@09c9215

diff --git a/src/main/java/org/bukkit/command/CommandSender.java b/src/main/java/org/bukkit/command/CommandSender.java
index daab4d13556150d4a4875f0273795e2c4cb82d63..9a439d5c546b3faee15e418560444d7622099830 100644
--- a/src/main/java/org/bukkit/command/CommandSender.java
+++ b/src/main/java/org/bukkit/command/CommandSender.java
@@ -156,4 +156,27 @@ public interface CommandSender extends net.kyori.adventure.audience.Audience, Pe
         this.sendMessage(new net.md_5.bungee.api.chat.TextComponent(components).toLegacyText());
     }
     // Paper end
+
+    // Daydream start
+    /**
+     * Sends a message with the MiniMessage format to the command sender.
+     * <p>
+     * See <a href="https://docs.adventure.kyori.net/minimessage/">MiniMessage docs</a>
+     * for more information on the format.
+     *
+     * @param message MiniMessage content
+     */
+    default void sendRichMessage(final @NotNull String message) {
+        this.sendMessage(net.kyori.adventure.text.minimessage.MiniMessage.miniMessage().deserialize(message));
+    }
+
+    /**
+     * Sends a plain message to the command sender.
+     *
+     * @param message plain message
+     */
+    default void sendPlainMessage(final @NotNull String message) {
+        this.sendMessage(net.kyori.adventure.text.Component.text(message));
+    }
+    // Daydream start
 }
