From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Mon, 17 Jan 2022 02:35:40 +0900
Subject: [PATCH] Add GeoIP Support


diff --git a/build.gradle.kts b/build.gradle.kts
index 1b25de66742066c9ce3b7767c460e7ce0e6828c8..45193307da7a02f297f23bfd87c0b486355843ff 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -47,6 +47,7 @@ dependencies {
     // Daydream end
     api("com.mojang:authlib:2.3.31") // Daydream - Mojang Authlib (keep in sync with major MC versions)
     api("moe.caramel:acacia-user:1.0.1-proto") // Daydream - caramel.moe Network
+    api("com.neovisionaries:nv-i18n:1.29") // Daydream - add GeoIP
 
     implementation("org.ow2.asm:asm:9.1")
     implementation("org.ow2.asm:asm-commons:9.1")
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index dbfc395d7c5e76eb42c28cc7712abb35302c81d9..24d7603a1d93b89649b8934f225db11a33220d8e 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -45,6 +45,16 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
 
     // Daydream start
     @NotNull moe.caramel.acacia.user.User getUserData();
+
+    /**
+     * Get the player's country.
+     *
+     * @return In the case below, return {@link com.neovisionaries.i18n.CountryCode#UNDEFINED}.
+     * <p><strong>1.</strong> if the server is not connected to the caramel.moe network.</p>
+     * <p><strong>2.</strong> if the player join the server with a local IP.</p>
+     * <p><strong>3.</strong> if Endless or Sunflower misinterpreted the player's IP.</p>
+     */
+    @NotNull com.neovisionaries.i18n.CountryCode getCountry();
     // Daydream end
 
     // Paper start
