From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Fri, 9 Dec 2022 20:25:16 +0900
Subject: [PATCH] Improve Chat System


diff --git a/src/main/java/io/papermc/paper/event/player/AbstractChatEvent.java b/src/main/java/io/papermc/paper/event/player/AbstractChatEvent.java
index ed8e885f226b02b9875b23ae2294a9056d2e8b29..87c7da30eb074b5a668211ba1d3cab065826452c 100644
--- a/src/main/java/io/papermc/paper/event/player/AbstractChatEvent.java
+++ b/src/main/java/io/papermc/paper/event/player/AbstractChatEvent.java
@@ -25,6 +25,38 @@ public abstract class AbstractChatEvent extends PlayerEvent implements Cancellab
     private ChatRenderer renderer;
     private Component message;
     private boolean cancelled = false;
+    // Daydream start - Improve Chat System
+    private net.kyori.adventure.chat.ChatType chatType = null;
+
+    /**
+     * Get the player message wrapper.
+     *
+     * @return player message wrapper
+     */
+    @NotNull
+    public moe.caramel.daydream.network.chat.PlayerMessage playerMessage() {
+        return (moe.caramel.daydream.network.chat.PlayerMessage) signedMessage;
+    }
+
+    /**
+     * 채팅 포맷을 가져옵니다.
+     *
+     * @return 채팅 포맷
+     */
+    @org.jetbrains.annotations.Nullable
+    public net.kyori.adventure.chat.ChatType chatType() {
+        return chatType;
+    }
+
+    /**
+     * 채팅 포맷을 설정합니다.
+     *
+     * @param chatType 채팅 포맷
+     */
+    public void chatType(@NotNull net.kyori.adventure.chat.ChatType chatType) {
+        this.chatType = chatType;
+    }
+    // Daydream end - Improve Chat System
 
     AbstractChatEvent(final boolean async, final @NotNull Player player, final @NotNull Set<Audience> viewers, final @NotNull ChatRenderer renderer, final @NotNull Component message, final @NotNull Component originalMessage, final @NotNull SignedMessage signedMessage) {
         super(player, async);
diff --git a/src/main/java/io/papermc/paper/event/player/ChatEvent.java b/src/main/java/io/papermc/paper/event/player/ChatEvent.java
index af025704e978dc0b11be277ab4646da77eb3a60b..29b8c1f09074ee8005d3598458e0a43c3e529133 100644
--- a/src/main/java/io/papermc/paper/event/player/ChatEvent.java
+++ b/src/main/java/io/papermc/paper/event/player/ChatEvent.java
@@ -29,6 +29,13 @@ public final class ChatEvent extends AbstractChatEvent {
         super(false, player, viewers, renderer, message, originalMessage, signedMessage);
     }
 
+    // Daydream start - Improve Chat System
+    @NotNull @Override
+    public moe.caramel.daydream.network.chat.PlayerMessage playerMessage() {
+        throw new UnsupportedOperationException("Async Only.");
+    }
+    // Daydream end - Improve Chat System
+
     @NotNull
     @Override
     public HandlerList getHandlers() {
