From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 8 Aug 2021 22:10:59 +0900
Subject: [PATCH] Add more Block API


diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index 66d43ee8e0a7fe032a0b4adb283d7e35e4adf51a..b032151aaf3dd055c495ba08b23d69c45940ee09 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -526,6 +526,19 @@ public interface Block extends Metadatable, net.kyori.adventure.translation.Tran
     boolean breakNaturally(@NotNull ItemStack tool, boolean triggerEffect);
     // Paper end
 
+    // Daydream start
+    /**
+     * Breaks the block and spawns items as if a player had digged it with a
+     * specific tool
+     *
+     * @param tool The tool or item in hand used for digging
+     * @param triggerEffect Play the block break particle effect and sound
+     * @param dropItem Item Drop
+     * @return true if the block was destroyed
+     */
+    boolean breakNaturally(@NotNull ItemStack tool, boolean triggerEffect, boolean dropItem);
+    // Daydream end
+
     /**
      * Simulate bone meal application to this block (if possible).
      *
