From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 2 Jul 2023 14:40:50 +0900
Subject: [PATCH] Add more Entity API


diff --git a/src/main/java/org/bukkit/entity/Display.java b/src/main/java/org/bukkit/entity/Display.java
index 2b6f0c627a848f37bc5b3f1732d0e0b82d9f89ad..c9748731f27b8caf9533a8c236c49ef5d43f8de1 100644
--- a/src/main/java/org/bukkit/entity/Display.java
+++ b/src/main/java/org/bukkit/entity/Display.java
@@ -63,8 +63,7 @@ public interface Display extends Entity {
     /**
      * Sets the teleport duration of this display.
      *
-     * @param duration new duration
-     * @throws IllegalArgumentException if duration is not between 0 and 59
+     * @param duration new duration (60 이상의 값을 설정하는 경우 {@link Entity#disableEntityUpdate()}를 사용하세요.)
      * @see #getTeleportDuration()
      */
     public void setTeleportDuration(int duration);
diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 9727155852d73852330ad216a70ea9f9c2a3763d..2b8c2770d4ab2c4ea92677b28a623395d4affc48 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -111,6 +111,17 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     public boolean isInWater();
 
+    // Daydream start - Add more Entity API (Fluid Height)
+    /**
+     * 플레이어 위치의 유체 높이를 가져옵니다.
+     *
+     * @param fluid 유체 타입
+     * @return 유체의 높이
+     * @throws IllegalArgumentException 인자로 {@link org.bukkit.Fluid#EMPTY}를 전달할 경우 던져집니다.
+     */
+    public double getFluidHeight(@NotNull org.bukkit.Fluid fluid);
+    // Daydream end - Add more Entity API (Fluid Height)
+
     /**
      * Gets the current world this entity resides in
      *
@@ -369,6 +380,30 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
     void lockFreezeTicks(boolean locked);
     // Paper end - Freeze Tick Lock API
 
+    // Daydream start - Add more Entity API (Disable Entity Update)
+    /**
+     * 엔티티 업데이트 비활성화를 예약합니다.
+     * <p>엔티티 업데이트가 완료된 후에 다음 틱 시간부터 업데이트를 하지 않습니다.</p>
+     */
+    void scheduleDisableEntityUpdate();
+
+    /**
+     * 엔티티 업데이트 중지 여부를 가져옵니다.
+     * <p>활성화 시, 위치 보정 및 엔티티 데이터 업데이트가 중지됩니다.</p>
+     *
+     * @return 엔티티 업데이트 중지 여부
+     */
+    boolean disableEntityUpdate();
+
+    /**
+     * 엔티티 업데이트 중지 여부를 설정합니다.
+     * <p>활성화 시, 위치 보정 및 엔티티 데이터 업데이트가 중지됩니다.</p>
+     *
+     * @param disable 엔티티 업데이트 중지 여부
+     */
+    void disableEntityUpdate(final boolean disable);
+    // Daydream end - Add more Entity API (Disable Entity Update)
+
     /**
      * Mark the entity's removal.
      *
@@ -532,6 +567,16 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
     @NotNull
     public UUID getUniqueId();
 
+    // Daydream start - Add more Entity API (getStringUniqueId)
+    /**
+     * Returns a unique and persistent string id for this entity
+     *
+     * @return string unique id
+     */
+    @NotNull
+    public String getStringUniqueId();
+    // Daydream end - Add more Entity API (getStringUniqueId)
+
     // Daydream start - Add Packet API
     /**
      * Gets the metadata of the entity.
diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 68c08e7a212bc3e3885f9b5a4d9aef85fcb3b029..4e52e6f754a6ccf9663cd296b704dfbe975c1403 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -696,6 +696,32 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
     public boolean hasLineOfSight(@NotNull Location location);
     // Paper end
 
+    // Daydream start - Add more Entity API (getSeenPercent)
+    /**
+     * 특정 위치에서 이 엔티티를 어느 정도 보고 있는지 가져옵니다.
+     *
+     * @param location 위치
+     * @param expand 확장할 히트박스의 크기
+     * @param instantReturn 즉시 반환 여부, 이 값이 {@code true}라면 이 엔티티가 조금이라도 보일 때 {@code 1.0f}를 즉시 반환합니다.
+     * @return 엔티티가 보이는 정도
+     */
+    default float getSeenPercent(@NotNull Location location, double expand, boolean instantReturn) {
+        return getSeenPercent(location.x(), location.y(), location.z(), expand, instantReturn);
+    }
+
+    /**
+     * 특정 위치에서 이 엔티티를 어느 정도 보고 있는지 가져옵니다.
+     *
+     * @param x 이 엔티티를 바라볼 x 위치
+     * @param y 이 엔티티를 바라볼 y 위치
+     * @param z 이 엔티티를 바라볼 z 위치
+     * @param expand 확장할 히트박스의 크기
+     * @param instantReturn 즉시 반환 여부, 이 값이 {@code true}라면 이 엔티티가 조금이라도 보일 때 {@code 1.0f}를 즉시 반환합니다.
+     * @return 엔티티가 보이는 정도
+     */
+    public float getSeenPercent(double x, double y, double z, double expand, boolean instantReturn);
+    // Daydream end - Add more Entity API (getSeenPercent)
+
     /**
      * Returns if the living entity despawns when away from players or not.
      * <p>
