From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 2 Jul 2023 14:40:50 +0900
Subject: [PATCH] Add more Entity API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 72324b823027edad59c9b916db238394198e5762..be19584591f30ed9780bf05e5dfc91b2cd57f3fd 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -325,6 +325,33 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
     void lockFreezeTicks(boolean locked);
     // Paper end - Freeze Tick Lock API
 
+    // Daydream start - Add more Entity API (Disable Entity Update)
+    /**
+     * 엔티티 업데이트 비활성화를 예약합니다.
+     * <p>엔티티 업데이트가 완료된 후에 다음 틱 시간부터 업데이트를 하지 않습니다.</p>
+     */
+    @ApiStatus.Experimental
+    void scheduleDisableEntityUpdate();
+
+    /**
+     * 엔티티 업데이트 중지 여부를 가져옵니다.
+     * <p>활성화 시, 위치 보정 및 엔티티 데이터 업데이트가 중지됩니다.</p>
+     *
+     * @return 엔티티 업데이트 중지 여부
+     */
+    @ApiStatus.Experimental
+    boolean disableEntityUpdate();
+
+    /**
+     * 엔티티 업데이트 중지 여부를 설정합니다.
+     * <p>활성화 시, 위치 보정 및 엔티티 데이터 업데이트가 중지됩니다.</p>
+     *
+     * @param disable 엔티티 업데이트 중지 여부
+     */
+    @ApiStatus.Experimental
+    void disableEntityUpdate(final boolean disable);
+    // Daydream end - Add more Entity API (Disable Entity Update)
+
     /**
      * Mark the entity's removal.
      *
@@ -486,6 +513,16 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
     @NotNull
     public UUID getUniqueId();
 
+    // Daydream start - Add more Entity API (getStringUniqueId)
+    /**
+     * Returns a unique and persistent string id for this entity
+     *
+     * @return string unique id
+     */
+    @NotNull
+    public String getStringUniqueId();
+    // Daydream end - Add more Entity API (getStringUniqueId)
+
     // Daydream start - Add Packet API
     /**
      * Gets the metadata of the entity.
diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 0f0b965ce48d13a64b5546a0abcfb45c4f5f4722..e911fed0df308e65c11efd4fb617571def7126d8 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -630,6 +630,32 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
     public boolean hasLineOfSight(@NotNull Location location);
     // Paper end
 
+    // Daydream start - Add more Entity API (getSeenPercent)
+    /**
+     * 특정 위치에서 이 엔티티를 어느 정도 보고 있는지 가져옵니다.
+     *
+     * @param location 위치
+     * @param expand 확장할 히트박스의 크기
+     * @param instantReturn 즉시 반환 여부, 이 값이 {@code true}라면 이 엔티티가 조금이라도 보일 때 {@code 1.0f}를 즉시 반환합니다.
+     * @return 엔티티가 보이는 정도
+     */
+    default float getSeenPercent(@NotNull Location location, double expand, boolean instantReturn) {
+        return getSeenPercent(location.x(), location.y(), location.z(), expand, instantReturn);
+    }
+
+    /**
+     * 특정 위치에서 이 엔티티를 어느 정도 보고 있는지 가져옵니다.
+     *
+     * @param x 이 엔티티를 바라볼 x 위치
+     * @param y 이 엔티티를 바라볼 y 위치
+     * @param z 이 엔티티를 바라볼 z 위치
+     * @param expand 확장할 히트박스의 크기
+     * @param instantReturn 즉시 반환 여부, 이 값이 {@code true}라면 이 엔티티가 조금이라도 보일 때 {@code 1.0f}를 즉시 반환합니다.
+     * @return 엔티티가 보이는 정도
+     */
+    public float getSeenPercent(double x, double y, double z, double expand, boolean instantReturn);
+    // Daydream end - Add more Entity API (getSeenPercent)
+
     /**
      * Returns if the living entity despawns when away from players or not.
      * <p>
