From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 2 Jul 2023 14:40:50 +0900
Subject: [PATCH] Add more Entity API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index d1fb056d09cff7cbe5857dab9e15f861a74b1db5..79913b95829d5cf7c0ef7cbe0b0e58508232775c 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -486,6 +486,16 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
     @NotNull
     public UUID getUniqueId();
 
+    // Daydream start - Add more Entity API (getStringUniqueId)
+    /**
+     * Returns a unique and persistent string id for this entity
+     *
+     * @return string unique id
+     */
+    @NotNull
+    public String getStringUniqueId();
+    // Daydream end - Add more Entity API (getStringUniqueId)
+
     // Daydream start - Add Packet API
     /**
      * Gets the metadata of the entity.
diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 47b0154928b3d36e2602da202df07defbcf82108..7df3b6ae4c8bf3298dcf1ddedc95616dd141580a 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -630,6 +630,32 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
     public boolean hasLineOfSight(@NotNull Location location);
     // Paper end
 
+    // Daydream start - Add more Entity API (getSeenPercent)
+    /**
+     * 특정 위치에서 이 엔티티를 어느 정도 보고 있는지 가져옵니다.
+     *
+     * @param location 위치
+     * @param expand 확장할 히트박스의 크기
+     * @param instantReturn 즉시 반환 여부, 이 값이 {@code true}라면 이 엔티티가 조금이라도 보일 때 {@code 1.0f}를 즉시 반환합니다.
+     * @return 엔티티가 보이는 정도
+     */
+    default float getSeenPercent(@NotNull Location location, double expand, boolean instantReturn) {
+        return getSeenPercent(location.x(), location.y(), location.z(), expand, instantReturn);
+    }
+
+    /**
+     * 특정 위치에서 이 엔티티를 어느 정도 보고 있는지 가져옵니다.
+     *
+     * @param x 이 엔티티를 바라볼 x 위치
+     * @param y 이 엔티티를 바라볼 y 위치
+     * @param z 이 엔티티를 바라볼 z 위치
+     * @param expand 확장할 히트박스의 크기
+     * @param instantReturn 즉시 반환 여부, 이 값이 {@code true}라면 이 엔티티가 조금이라도 보일 때 {@code 1.0f}를 즉시 반환합니다.
+     * @return 엔티티가 보이는 정도
+     */
+    public float getSeenPercent(double x, double y, double z, double expand, boolean instantReturn);
+    // Daydream end - Add more Entity API (getSeenPercent)
+
     /**
      * Returns if the living entity despawns when away from players or not.
      * <p>
