From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Tue, 26 Jul 2022 10:27:57 +0900
Subject: [PATCH] Add Packet Invisible Entity API


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index bd35cb298972ec1bc586ba25d155ee24c5b27327..397d847dcf7f77fa907045d1600008eade7ad82c 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -2284,6 +2284,26 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     @NotNull
     public FallingBlock spawnFallingBlock(@NotNull Location location, @NotNull BlockData data) throws IllegalArgumentException;
 
+    // Daydream start
+    /**
+     * Spawn a {@link FallingBlock} entity at the given {@link Location} of
+     * the specified {@link Material}. The material dictates what is falling.
+     * When the FallingBlock hits the ground, it will place that block.
+     * <p>
+     * The Material must be a block type, check with {@link Material#isBlock()
+     * material.isBlock()}. The Material may not be air.
+     *
+     * @param location The {@link Location} to spawn the FallingBlock
+     * @param data The block data
+     * @param function The function to be run before the entity is spawned.
+     * @return The spawned {@link FallingBlock} instance
+     * @throws IllegalArgumentException if {@link Location} or {@link
+     *     BlockData} are null
+     */
+    @NotNull
+    public FallingBlock spawnFallingBlock(@NotNull Location location, @NotNull BlockData data, @NotNull Consumer<FallingBlock> function) throws IllegalArgumentException;
+    // Daydream end
+
     /**
      * Spawn a {@link FallingBlock} entity at the given {@link Location} of the
      * specified {@link Material}. The material dictates what is falling.
diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 2b4ca69037aa1ff66e2fe351cfcb99c595b58671..20091ade52b75688f42199e01194be38f9f9fbb0 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -863,4 +863,29 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     boolean isInPowderedSnow();
     // Paper end
+
+    // Daydream start - Add Packet Invisible API
+
+    /**
+     * 엔티티가 생성되었는지 확인합니다.
+     *
+     * @return 만약 엔티티가 생성되고 패킷을 한 번이라도
+     * 전송하려고 한 경우 {@code true}를 반환합니다.
+     */
+    boolean isInitialized();
+
+    /**
+     * 이 엔티티가 패킷 숨기기를 사용하는지 확인합니다.
+     *
+     * @return 만약 이 엔티티가 패킷 숨기기를 사용하는 경우 {@code true}를 반환합니다.
+     */
+    boolean usePacketInvisible();
+
+    /**
+     * 패킷 숨기기 사용 여부를 정합니다.
+     *
+     * @param invisible {@code true}인 경우, 패킷 숨기기를 사용하며 이 엔티티와 관련된 모든 패킷이 전송되지 않습니다.
+     */
+    void setPacketInvisible(boolean invisible);
+    // Daydream start - Add Packet Invisible API
 }
