From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Mon, 23 Aug 2021 19:05:20 +0900
Subject: [PATCH] Add DaydreamValue to EntityEvent & EntityPotion method add
 daydream value


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 5063b1c1b6710f36345b3172ac741598e59e4c7f..f6d857c90417072d36d0d10326271e3e1b8549ae 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -437,6 +437,24 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
     @Deprecated
     public boolean addPotionEffect(@NotNull PotionEffect effect, boolean force);
 
+    // Daydream start
+    /**
+     * Adds the given {@link PotionEffect} to the living entity.
+     * <p>
+     * Only one potion effect can be present for a given {@link
+     * PotionEffectType}.
+     *
+     * @param effect PotionEffect to be added
+     * @param force whether conflicting effects should be removed
+     * @param daydream daydream value
+     * @return whether the effect could be added
+     * @deprecated no need to force since multiple effects of the same type are
+     * now supported.
+     */
+    @Deprecated
+    public boolean addPotionEffect(@NotNull PotionEffect effect, boolean force, boolean daydream);
+    // Daydream end
+
     /**
      * Attempts to add all of the given {@link PotionEffect} to the living
      * entity.
@@ -473,6 +491,16 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      */
     public void removePotionEffect(@NotNull PotionEffectType type);
 
+    // Daydream start
+    /**
+      * Removes any effects present of the given {@link PotionEffectType}.
+      *
+      * @param type the potion type to remove
+      * @param daydream daydream value
+      */
+    public void removePotionEffect(@NotNull PotionEffectType type, boolean daydream);
+    // Daydream end
+
     /**
      * Returns all currently active {@link PotionEffect}s on the living
      * entity.
diff --git a/src/main/java/org/bukkit/event/entity/EntityEvent.java b/src/main/java/org/bukkit/event/entity/EntityEvent.java
index e3a9e32933fe84644f7458d2e0b8b09d3eacb21c..7ff0e1fb6b9cb99bb8eb262cab740b1b2e946a93 100644
--- a/src/main/java/org/bukkit/event/entity/EntityEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityEvent.java
@@ -10,6 +10,7 @@ import org.jetbrains.annotations.NotNull;
  */
 public abstract class EntityEvent extends Event {
     protected Entity entity;
+    protected boolean daydream = false; // Daydream
 
     public EntityEvent(@NotNull final Entity what) {
         entity = what;
@@ -34,4 +35,14 @@ public abstract class EntityEvent extends Event {
     public EntityType getEntityType() {
         return entity.getType();
     }
+
+    // Daydream start
+    public void setDaydreamValue(boolean daydream) {
+        this.daydream = daydream;
+    }
+
+    public boolean hasDaydreamValue() {
+        return daydream;
+    }
+    // Daydream end
 }
