From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Thu, 9 Mar 2023 19:46:01 +0900
Subject: [PATCH] Add More Team API


diff --git a/src/main/java/moe/caramel/daydream/team/TeamListener.java b/src/main/java/moe/caramel/daydream/team/TeamListener.java
new file mode 100644
index 0000000000000000000000000000000000000000..8d2ef759e1f132053162442baee824825be1ae83
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/team/TeamListener.java
@@ -0,0 +1,52 @@
+package moe.caramel.daydream.team;
+
+import org.bukkit.OfflinePlayer;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * {@link org.bukkit.scoreboard.Team}의 변경 사항을 리스닝합니다.
+ */
+public interface TeamListener {
+
+    /**
+     * 팀이 제거되었을 때
+     */
+    default void unregister() {
+    }
+
+    /**
+     * 엔트리가 팀에 추가되었을 때
+     *
+     * @param entry 엔트리
+     */
+    default void addEntry(final @NotNull String entry) {
+    }
+
+    /**
+     * 플레이어가 팀에 추가되었을 때
+     * <br>
+     * {@link #addEntry(String)}와 동시에 호출될 수 있습니다.
+     *
+     * @param player 플레이어
+     */
+    default void addPlayer(final @NotNull OfflinePlayer player) {
+    }
+
+    /**
+     * 엔트리가 팀에서 제거되었을 때
+     *
+     * @param entry 엔트리
+     */
+    default void removeEntry(final @NotNull String entry) {
+    }
+
+    /**
+     * 플레이어가 팀에서 제거되었을 때
+     * <br>
+     * {@link #addEntry(String)}와 동시에 호출될 수 있습니다.
+     *
+     * @param player 플레이어
+     */
+    default void removePlayer(final @NotNull OfflinePlayer player) {
+    }
+}
diff --git a/src/main/java/org/bukkit/scoreboard/Team.java b/src/main/java/org/bukkit/scoreboard/Team.java
index 06a5d3177ca7ab90c3fd9d2053b2ec5e887c7c62..28bc5a26abd41a02e1014eab604eb67532270eef 100644
--- a/src/main/java/org/bukkit/scoreboard/Team.java
+++ b/src/main/java/org/bukkit/scoreboard/Team.java
@@ -532,6 +532,46 @@ public interface Team {
     boolean hasEntity(@NotNull org.bukkit.entity.Entity entity) throws IllegalStateException, IllegalArgumentException;
     // Paper end
 
+    // Daydream start - Add More Team API
+    /**
+     * 팀 리스너를 가져옵니다.
+     *
+     * @return 팀 리스너
+     */
+    @Nullable
+    moe.caramel.daydream.team.TeamListener getListener();
+
+    /**
+     * 팀 리스너를 설정합니다.
+     *
+     * @param listener 팀 리스너
+     */
+    void setListener(@Nullable moe.caramel.daydream.team.TeamListener listener);
+
+    /**
+     * This puts a collection of players onto this team for the scoreboard which results in one
+     * packet for the updates rather than a packet-per-entity.
+     * <p>
+     * Players on other teams will be removed from their respective teams.
+     *
+     * @param players the players to add
+     * @throws IllegalArgumentException if players are null
+     * @throws IllegalStateException if this team has been unregistered
+     */
+    void addPlayers(@NotNull java.util.Collection<? extends OfflinePlayer> players) throws IllegalStateException, IllegalArgumentException;
+
+    /**
+     * Removes a collection of players from this team  which results in one
+     * packet for the updates rather than a packet-per-entity.
+     *
+     * @param players the entries to remove
+     * @return if any of the players were a part of this team
+     * @throws IllegalArgumentException if players is null
+     * @throws IllegalStateException if this team has been unregistered
+     */
+    boolean removePlayers(@NotNull java.util.Collection<? extends OfflinePlayer> players) throws IllegalStateException, IllegalArgumentException;
+    // Daydream end - Add More Team API
+
     /**
      * Represents an option which may be applied to this team.
      */
