From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Wed, 3 Aug 2022 12:47:42 +0900
Subject: [PATCH] Add More Explode Source API


diff --git a/src/main/java/io/papermc/paper/entity/damage/BadRespawnPointDamageOrigin.java b/src/main/java/io/papermc/paper/entity/damage/BadRespawnPointDamageOrigin.java
new file mode 100644
index 0000000000000000000000000000000000000000..436ff035138a82cfb79602804604f093403ded7a
--- /dev/null
+++ b/src/main/java/io/papermc/paper/entity/damage/BadRespawnPointDamageOrigin.java
@@ -0,0 +1,15 @@
+package io.papermc.paper.entity.damage;
+
+import org.bukkit.entity.LivingEntity;
+import org.jetbrains.annotations.Nullable;
+
+public interface BadRespawnPointDamageOrigin extends DamageOrigin {
+
+    /**
+     * 폭발에 대한 책임이 있는 엔티티를 가져옵니다.
+     *
+     * @return 엔티티를 알 수 없는 경우 {@code null}을 반환합니다.
+     */
+    @Nullable
+    LivingEntity getSource();
+}
diff --git a/src/main/java/io/papermc/paper/entity/damage/builder/BadRespawnPointDamageOriginBuilder.java b/src/main/java/io/papermc/paper/entity/damage/builder/BadRespawnPointDamageOriginBuilder.java
new file mode 100644
index 0000000000000000000000000000000000000000..91ff01ff6817d881285d8b14e34d018491c8793c
--- /dev/null
+++ b/src/main/java/io/papermc/paper/entity/damage/builder/BadRespawnPointDamageOriginBuilder.java
@@ -0,0 +1,15 @@
+package io.papermc.paper.entity.damage.builder;
+
+import io.papermc.paper.entity.damage.BadRespawnPointDamageOrigin;
+import org.bukkit.entity.LivingEntity;
+import org.jetbrains.annotations.Nullable;
+
+public interface BadRespawnPointDamageOriginBuilder extends AbstractDamageOriginBuilder<BadRespawnPointDamageOriginBuilder, BadRespawnPointDamageOrigin> {
+
+    /**
+     * 폭발에 대한 책임이 있는 엔티티를 설정합니다.
+     *
+     * @param source 폭발에 대한 책임이 있는 엔티티
+     */
+    void setSource(@Nullable LivingEntity source);
+}
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index b6992a1cd65777123de6254ee368e1b99c7cfeae..14b1d2b98ff5a8a15cf91d5330b31794b4c6962d 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2426,6 +2426,11 @@ public final class Bukkit {
     public static io.papermc.paper.entity.damage.builder.IndirectEntityDamageOriginBuilder createDamageIndirectEntitySource(@NotNull String name, @NotNull Entity directSource, @Nullable Entity source) {
         return server.createDamageIndirectEntitySource(name, directSource, source);
     }
+
+    @NotNull
+    public static io.papermc.paper.entity.damage.builder.BadRespawnPointDamageOriginBuilder createDamageBadRespawnPointSource(@NotNull String name, @Nullable org.bukkit.entity.LivingEntity source) {
+        return server.createDamageBadRespawnPointSource(name, source);
+    }
     // Paper end
 
     // Daydream start - Add Packet API
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index be4a489b46e56f4dd8d48076e4dfdd72a9479e01..52a61ede5fccb83a9e5674009fc0aa874726501f 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2112,6 +2112,9 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
 
     @NotNull
     io.papermc.paper.entity.damage.builder.IndirectEntityDamageOriginBuilder createDamageIndirectEntitySource(@NotNull String name, @NotNull Entity directSource, @Nullable Entity source);
+
+    @NotNull
+    io.papermc.paper.entity.damage.builder.BadRespawnPointDamageOriginBuilder createDamageBadRespawnPointSource(@NotNull String name, @Nullable org.bukkit.entity.LivingEntity source);
     // Paper end
 
     // Daydream start - Add Packet API
diff --git a/src/main/java/org/bukkit/entity/EnderCrystal.java b/src/main/java/org/bukkit/entity/EnderCrystal.java
index febc7b3f4c9a402d74d2bcf27e89578fc6d7d6a3..5fc135b1f32e2e5e7dccadbd347ef8a80b8f8d74 100644
--- a/src/main/java/org/bukkit/entity/EnderCrystal.java
+++ b/src/main/java/org/bukkit/entity/EnderCrystal.java
@@ -40,4 +40,15 @@ public interface EnderCrystal extends Entity {
      * @throws IllegalArgumentException for differing worlds
      */
     void setBeamTarget(@Nullable Location location);
+
+    // Daydream start - Add More Explode Source API
+    /**
+     * 엔더 수정 폭발에 대한 책임이 있는 엔티티를 가져옵니다.
+     *
+     * @return 아직 폭발되지 않았거나 엔티티가 공격하여 폭발된 것이 아닌 경우
+     * {@code null}을 반환하며, 엔티티가 공격해 폭발한 경우 해당 엔티티를 반환합니다.
+     */
+    @Nullable
+    Entity getSource();
+    // Daydream end - Add More Explode Source API
 }
