From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Wed, 25 Aug 2021 00:49:47 +0900
Subject: [PATCH] Add Protocol Constant


diff --git a/src/main/java/moe/caramel/daydream/constants/ProtocolConstants.java b/src/main/java/moe/caramel/daydream/constants/ProtocolConstants.java
new file mode 100644
index 0000000000000000000000000000000000000000..aae47d23bfd7fbf8e5f0dd2db61d78d2367289d4
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/constants/ProtocolConstants.java
@@ -0,0 +1,10 @@
+package moe.caramel.daydream.constants;
+
+public final class ProtocolConstants {
+
+    private ProtocolConstants() { throw new UnsupportedOperationException(); }
+
+    public static final int MINECRAFT_1_16_5 = 754;
+    public static final int MINECRAFT_1_17 = 755;
+    public static final int MINECRAFT_1_17_1 = 756;
+}
