From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 11 Feb 2024 15:17:38 +0900
Subject: [PATCH] Add Plugin Embedded Datapack System


diff --git a/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java b/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java
index ef393f1f93ca48264fc1b6e3a27787f6a9152e1b..b470c481c4c848a59d9c523d4a25cecc9bddb5ba 100644
--- a/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java
+++ b/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java
@@ -200,4 +200,14 @@ public interface PluginMeta {
     @Nullable
     String getAPIVersion();
 
+    // Daydream start - Add Plugin Embedded Datapack System
+    /**
+     * 플러그인 내장 데이터팩을 가지고 있는지 확인합니다.
+     *
+     * @return 내장 데이터팩 여부
+     */
+    default boolean hasEmbeddedDatapack() {
+        return false;
+    }
+    // Daydream end - Add Plugin Embedded Datapack System
 }
diff --git a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
index c0691a849831f99268fdcb7b0f471f80a1a2a70e..3eb20e8bd3805c6e5b4cb0be50d8623084fc1836 100644
--- a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
+++ b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
@@ -265,6 +265,21 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
      */
     @org.jetbrains.annotations.ApiStatus.Internal
     public PluginDescriptionFile(String rawName, String name, List<String> provides, String main, String classLoaderOf, List<String> depend, List<String> softDepend, List<String> loadBefore, String version, Map<String, Map<String, Object>> commands, String description, List<String> authors, List<String> contributors, String website, String prefix, PluginLoadOrder order, List<Permission> permissions, PermissionDefault defaultPerm, Set<PluginAwareness> awareness, String apiVersion, List<String> libraries) {
+        // Daydream start - Add Plugin Embedded Datapack System
+        this(rawName, name, provides, main, classLoaderOf, depend, softDepend, loadBefore, version, commands, description, authors, contributors, website, prefix, order, permissions, defaultPerm, awareness, apiVersion, libraries, false);
+    }
+
+    private boolean hasEmbeddedDatapack;
+
+    @Override
+    public boolean hasEmbeddedDatapack() {
+        return hasEmbeddedDatapack;
+    }
+
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public PluginDescriptionFile(String rawName, String name, List<String> provides, String main, String classLoaderOf, List<String> depend, List<String> softDepend, List<String> loadBefore, String version, Map<String, Map<String, Object>> commands, String description, List<String> authors, List<String> contributors, String website, String prefix, PluginLoadOrder order, List<Permission> permissions, PermissionDefault defaultPerm, Set<PluginAwareness> awareness, String apiVersion, List<String> libraries, boolean hasDatapack) {
+        this.hasEmbeddedDatapack = hasDatapack;
+        // Daydream end - Add Plugin Embedded Datapack System
         this.rawName = rawName;
         this.name = name;
         this.provides = provides;
@@ -1242,6 +1257,12 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
         if (map.get("prefix") != null) {
             prefix = map.get("prefix").toString();
         }
+
+        // Daydream start - Add Plugin Embedded Datapack System
+        if (map.get("has-embedded-datapack") != null) {
+            hasEmbeddedDatapack = Boolean.parseBoolean(map.get("has-embedded-datapack").toString());
+        }
+        // Daydream end - Add Plugin Embedded Datapack System
     }
 
     @NotNull
@@ -1319,6 +1340,12 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
             map.put("prefix", prefix);
         }
 
+        // Daydream start - Add Plugin Embedded Datapack System
+        if (hasEmbeddedDatapack) {
+            map.put("has-embedded-datapack", true);
+        }
+        // Daydream end - Add Plugin Embedded Datapack System
+
         return map;
     }
 
