From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Thu, 1 Jun 2023 06:32:36 +0900
Subject: [PATCH] Add More Firework API


diff --git a/src/main/java/org/bukkit/entity/Firework.java b/src/main/java/org/bukkit/entity/Firework.java
index 217d348ad0bbef720b25d3b507a55ca8105b7731..d48dcce06b6079afe9eac81ee3dc702dd950597d 100644
--- a/src/main/java/org/bukkit/entity/Firework.java
+++ b/src/main/java/org/bukkit/entity/Firework.java
@@ -185,4 +185,20 @@ public interface Firework extends Projectile {
      */
     void setTicksToDetonate(int ticks);
     // Paper end
+
+    // Daydream start - Add More Firework API
+    /**
+     * 폭발 대미지 제거 여부를 가져옵니다.
+     *
+     * @return 폭발 대미지 제거 여부
+     */
+    boolean isDisableExplosionDeal();
+
+    /**
+     * 폭발 대미지 제거 여부를 설정합니다.
+     *
+     * @param disable 폭발 대미지 제거 여부
+     */
+    void setDisableExplosionDeal(boolean disable);
+    // Daydream end - Add More Firework API
 }
