From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sat, 30 Dec 2023 18:21:22 +0900
Subject: [PATCH] Add Player Initial Locale API


diff --git a/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java b/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java
index f8b69b52ec8efa103e4e78e1b5c6a015e73d2a75..bbf1f99ad0c13c5c997a148c8c3e8726f0c935b7 100644
--- a/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java
+++ b/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java
@@ -283,6 +283,29 @@ public class AsyncPlayerPreLoginEvent extends Event {
     }
     // Paper end
 
+    // Daydream start - Add Player Initial Locale API
+    private java.util.Locale initialLocale;
+
+    /**
+     * 플레이어의 초기 언어를 가져옵니다.
+     *
+     * @return 플레이어의 초기 언어
+     */
+    @org.jetbrains.annotations.Nullable
+    public java.util.Locale initialLocale() {
+        return initialLocale;
+    }
+
+    /**
+     * 플레이어의 초기 언어를 설정합니다.
+     *
+     * @param initialLocale 플레이어의 초기 언어
+     */
+    public void initialLocale(final @org.jetbrains.annotations.Nullable java.util.Locale initialLocale) {
+        this.initialLocale = initialLocale;
+    }
+    // Daydream end - Add Player Initial Locale API
+
     @NotNull
     @Override
     public HandlerList getHandlers() {
