From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 22 Oct 2023 00:45:03 +0900
Subject: [PATCH] Make PlayerTrackEntityEvent cancellable


diff --git a/src/main/java/io/papermc/paper/event/player/PlayerTrackEntityEvent.java b/src/main/java/io/papermc/paper/event/player/PlayerTrackEntityEvent.java
index 167be75dc616578db2df1a88f1a54b17707b3a88..47101b2543c47658265c0f20c0426d8eb2a1e769 100644
--- a/src/main/java/io/papermc/paper/event/player/PlayerTrackEntityEvent.java
+++ b/src/main/java/io/papermc/paper/event/player/PlayerTrackEntityEvent.java
@@ -9,7 +9,7 @@ import org.jetbrains.annotations.NotNull;
 /**
  * Is called when a {@link Player} tracks an {@link Entity}.
  */
-public class PlayerTrackEntityEvent extends PlayerEvent {
+public class PlayerTrackEntityEvent extends PlayerEvent implements org.bukkit.event.Cancellable { // Daydream - Make PlayerTrackEntityEvent cancellable
 
     private static final HandlerList handlers = new HandlerList();
     private final Entity entity;
@@ -39,4 +39,18 @@ public class PlayerTrackEntityEvent extends PlayerEvent {
     public Entity getEntity() {
         return entity;
     }
+
+    // Daydream start - Make PlayerTrackEntityEvent cancellable
+    private boolean cancelled;
+
+    @Override
+    public boolean isCancelled() {
+        return cancelled;
+    }
+
+    @Override
+    public void setCancelled(boolean cancel) {
+        this.cancelled = cancel;
+    }
+    // Daydream end - Make PlayerTrackEntityEvent cancellable
 }
