From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 10 Dec 2023 16:28:01 +0900
Subject: [PATCH] Add Client Detector


diff --git a/src/main/java/moe/caramel/daydream/event/configure/PlayerClientDetectedEvent.java b/src/main/java/moe/caramel/daydream/event/configure/PlayerClientDetectedEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..807f72743db0952c8dcce778a96e94f7a694a0b0
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/event/configure/PlayerClientDetectedEvent.java
@@ -0,0 +1,63 @@
+package moe.caramel.daydream.event.configure;
+
+import moe.caramel.daydream.player.ClientType;
+import net.kyori.adventure.text.Component;
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Configuration 페이즈에서 플레이어의 클라이언트 종류를 감지 완료하였을 때 호출됩니다.
+ */
+public final class PlayerClientDetectedEvent extends ConfigurationPhaseEvent {
+
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+    public @NotNull HandlerList getHandlers() { return HANDLER_LIST; }
+    public static @NotNull HandlerList getHandlerList() { return HANDLER_LIST; }
+
+    private Component disconnectReason;
+
+    public PlayerClientDetectedEvent(final @NotNull Player who) {
+        super(who, true);
+    }
+
+    /**
+     * 감지된 클라이언트의 타입을 가져옵니다.
+     *
+     * @return 클라이언트의 타입
+     */
+    @NotNull
+    public ClientType clientType() {
+        return player.getClientType();
+    }
+
+    /**
+     * 클라이언트의 브랜드 이름을 가져옵니다.
+     *
+     * @return 브랜드 이름 (수정된 클라이언트가 서버로 정보를 전송하지 않을 수도 있음)
+     */
+    @Nullable
+    public String clientBrandName() {
+        return player.getClientBrandName();
+    }
+
+    /**
+     * 플레이어 추방 사유를 가져옵니다.
+     *
+     * @return 플레이어 추방 사유
+     */
+    @Nullable
+    public Component disconnectReason() {
+        return disconnectReason;
+    }
+
+    /**
+     * 플레이어를 서버에서 추방하기로 큰 결심을 한 경우, 추방 사유를 설정합니다.
+     *
+     * @param disconnectReason 플레이어 추방 사유 ({@code null}인 경우 추방하지 않습니다.)
+     */
+    public void disconnectReason(final @Nullable Component disconnectReason) {
+        this.disconnectReason = disconnectReason;
+    }
+}
diff --git a/src/main/java/moe/caramel/daydream/player/ClientType.java b/src/main/java/moe/caramel/daydream/player/ClientType.java
new file mode 100644
index 0000000000000000000000000000000000000000..963d3aa970ef24bafd2aaa17bde4b8c2bed88c57
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/player/ClientType.java
@@ -0,0 +1,34 @@
+package moe.caramel.daydream.player;
+
+/**
+ * 클라이언트 타입
+ */
+public enum ClientType {
+
+    /**
+     * 바닐라 클라이언트 또는 알 수 없는 클라이언트
+     */
+    VANILLA,
+    /**
+     * 모드 로더이지만 자세히는 알 수 없음
+     */
+    UNKNOWN_CLIENT,
+
+    /**
+     * Fabric 모드 로더
+     */
+    FABRIC,
+    /**
+     * Quilt 모드 로더
+     */
+    QUILT,
+
+    /**
+     * Forge 모드 로더
+     */
+    FORGE,
+    /**
+     * NeoForge 모드 로더
+     */
+    NEOFORGE,
+}
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 4d759fb0dd2cd59242e855eacbc6826ec6d8c3f9..7622a5df12f729f436693d50003282b53fae3d1d 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3574,6 +3574,19 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     String getClientBrandName();
     // Paper end
 
+    // Daydream start - Add Client Detector
+    /**
+     * 서버에서 감지한 클라이언트의 종류를 가져옵니다.
+     * <p>
+     * Daydream 설정에서 감지가 비활성화 되었거나, 아직 프로세스를
+     * 시작하지 않은 경우 {@link moe.caramel.daydream.player.ClientType#VANILLA}를 반환합니다.
+     *
+     * @return 클라이언트의 종류
+     */
+    @NotNull
+    moe.caramel.daydream.player.ClientType getClientType();
+    // Daydream end - Add Client Detector
+
     // Paper start - Teleport API
     /**
      * Sets the player's rotation.
