--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -65,6 +_,15 @@
      */
     public MainHand getMainHand();
 
+    // Daydream start - Settable Player Arm
+    /**
+     * 플레이어의 주로 사용하는 손을 변경합니다.
+     *
+     * @param mainHand 주로 사용하는 손
+     */
+    public void setMainHand(MainHand mainHand);
+    // Daydream end - Settable Player Arm
+
     /**
      * If the player currently has an inventory window open, this method will
      * set a property of that window, such as the state of a progress bar.
@@ -429,6 +_,16 @@
      * @param ticks the amount of ticks to set or 0 to remove
      */
     public void setCooldown(ItemStack item, int ticks);
+
+    // Daydream start - Add Group cooldown methods
+    /**
+     * 이 그룹에 대한 쿨다운이 존재하는지 확인합니다.
+     *
+     * @param cooldownGroup 쿨다운 그룹
+     * @return 쿨다운을 가지고 있다면 {@code true}
+     */
+    public boolean hasCooldown(Key cooldownGroup);
+    // Daydream end - Add Group cooldown methods
 
     /**
      * Get the cooldown time in ticks remaining for the specified cooldown group.
