--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -66,6 +_,15 @@
     @NotNull
     public MainHand getMainHand();
 
+    // Daydream start - Settable Player Arm
+    /**
+     * 플레이어의 주로 사용하는 손을 변경합니다.
+     *
+     * @param mainHand 주로 사용하는 손
+     */
+    public void setMainHand(@NotNull MainHand mainHand);
+    // Daydream end - Settable Player Arm
+
     /**
      * If the player currently has an inventory window open, this method will
      * set a property of that window, such as the state of a progress bar.
