--- a/src/main/java/org/bukkit/event/Event.java
+++ b/src/main/java/org/bukkit/event/Event.java
@@ -51,6 +_,31 @@
         }
     }
 
+    // Daydream start - Add Event context
+    private final java.util.Map<String, Object> context = new java.util.HashMap<>();
+
+    /**
+     * 이 이벤트의 실행 컨텍스트를 가져옵니다.
+     *
+     * @return 이 이벤트의 실행 컨텍스트
+    */
+    @NotNull
+    public java.util.Map<String, Object> getEventContext() {
+        return context;
+    }
+    // Daydream end - Add Event context
+
+    // Daydream start - Event interrupt support
+    /**
+     * 이 이벤트가 인터럽트를 지원하는지 여부를 가져옵니다.
+     *
+     * @return 인터럽트 지원 여부
+     */
+    public boolean isSupportInterrupt() {
+        return false;
+    }
+    // Daydream end - Event interrupt support
+
     /**
      * Convenience method for providing a user-friendly identifier. By
      * default, it is the event's class's {@linkplain Class#getSimpleName()
