From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 28 Apr 2024 22:08:32 +0900
Subject: [PATCH] Update to 1.20.6


diff --git a/com/destroystokyo/paper/entity/ai/VanillaGoal.java b/com/destroystokyo/paper/entity/ai/VanillaGoal.java
index 7ff40b6c333b06799ed2fd2a0a39bfebd48e3580..43f4deacef349502cbb207aafc4f9cb7a75177c5 100644
--- a/com/destroystokyo/paper/entity/ai/VanillaGoal.java
+++ b/com/destroystokyo/paper/entity/ai/VanillaGoal.java
@@ -70,7 +70,7 @@ import org.jetbrains.annotations.NotNull;
         "unused",
         "SpellCheckingInspection"
 })
-@GeneratedFrom("1.20.5")
+@GeneratedFrom("1.20.6")
 public interface VanillaGoal<T extends Mob> extends Goal<T> {
     GoalKey<AbstractHorse> RANDOM_STAND = create("random_stand", AbstractHorse.class);
 
diff --git a/io/papermc/paper/registry/keys/BiomeKeys.java b/io/papermc/paper/registry/keys/BiomeKeys.java
index 3bd15016b012b8dba75787803eae3e4ca19140e1..173550253fd1c5b85e767b4cd2dae62e19259aed 100644
--- a/io/papermc/paper/registry/keys/BiomeKeys.java
+++ b/io/papermc/paper/registry/keys/BiomeKeys.java
@@ -23,7 +23,7 @@ import org.jetbrains.annotations.NotNull;
         "unused",
         "SpellCheckingInspection"
 })
-@GeneratedFrom("1.20.5")
+@GeneratedFrom("1.20.6")
 @ApiStatus.Experimental
 public final class BiomeKeys {
     /**
diff --git a/io/papermc/paper/registry/keys/GameEventKeys.java b/io/papermc/paper/registry/keys/GameEventKeys.java
index 26ce70c2ab1cacde5bd5d0036a10f03a598a7063..f220c63b8ab30f5737bf9d7ae78e1de5564f9020 100644
--- a/io/papermc/paper/registry/keys/GameEventKeys.java
+++ b/io/papermc/paper/registry/keys/GameEventKeys.java
@@ -23,7 +23,7 @@ import org.jetbrains.annotations.NotNull;
         "unused",
         "SpellCheckingInspection"
 })
-@GeneratedFrom("1.20.5")
+@GeneratedFrom("1.20.6")
 @ApiStatus.Experimental
 public final class GameEventKeys {
     /**
diff --git a/io/papermc/paper/registry/keys/StructureKeys.java b/io/papermc/paper/registry/keys/StructureKeys.java
index 8b679c044e94b5dbb02dab9528ea5fee14fe93f5..22d644aa24521e73b855dcd78d29f6fd3bfb1505 100644
--- a/io/papermc/paper/registry/keys/StructureKeys.java
+++ b/io/papermc/paper/registry/keys/StructureKeys.java
@@ -24,7 +24,7 @@ import org.jetbrains.annotations.NotNull;
         "unused",
         "SpellCheckingInspection"
 })
-@GeneratedFrom("1.20.5")
+@GeneratedFrom("1.20.6")
 @ApiStatus.Experimental
 public final class StructureKeys {
     /**
diff --git a/io/papermc/paper/registry/keys/StructureTypeKeys.java b/io/papermc/paper/registry/keys/StructureTypeKeys.java
index 8e7628c6a457d33382458006d91924b4741e5eda..bd36a385583cb2e10add1a2e146a0d6780239068 100644
--- a/io/papermc/paper/registry/keys/StructureTypeKeys.java
+++ b/io/papermc/paper/registry/keys/StructureTypeKeys.java
@@ -23,7 +23,7 @@ import org.jetbrains.annotations.NotNull;
         "unused",
         "SpellCheckingInspection"
 })
-@GeneratedFrom("1.20.5")
+@GeneratedFrom("1.20.6")
 @ApiStatus.Experimental
 public final class StructureTypeKeys {
     /**
diff --git a/io/papermc/paper/registry/keys/TrimMaterialKeys.java b/io/papermc/paper/registry/keys/TrimMaterialKeys.java
index e52bb12173760520cb3210e8292e7ff9573452cd..0635cecd49774038a9c1e80b3da3643de840ee1a 100644
--- a/io/papermc/paper/registry/keys/TrimMaterialKeys.java
+++ b/io/papermc/paper/registry/keys/TrimMaterialKeys.java
@@ -23,7 +23,7 @@ import org.jetbrains.annotations.NotNull;
         "unused",
         "SpellCheckingInspection"
 })
-@GeneratedFrom("1.20.5")
+@GeneratedFrom("1.20.6")
 @ApiStatus.Experimental
 public final class TrimMaterialKeys {
     /**
diff --git a/io/papermc/paper/registry/keys/TrimPatternKeys.java b/io/papermc/paper/registry/keys/TrimPatternKeys.java
index 72d23426af8629abf8de7142abb19ae075ac105c..b3e92fbbe281ba7943b3467a95f77429f5a0dab0 100644
--- a/io/papermc/paper/registry/keys/TrimPatternKeys.java
+++ b/io/papermc/paper/registry/keys/TrimPatternKeys.java
@@ -24,7 +24,7 @@ import org.jetbrains.annotations.NotNull;
         "unused",
         "SpellCheckingInspection"
 })
-@GeneratedFrom("1.20.5")
+@GeneratedFrom("1.20.6")
 @ApiStatus.Experimental
 public final class TrimPatternKeys {
     /**
