From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Thu, 5 Aug 2021 15:44:59 +0900
Subject: [PATCH] Add Daydream Command


diff --git a/src/main/java/org/bukkit/command/SimpleCommandMap.java b/src/main/java/org/bukkit/command/SimpleCommandMap.java
index ad1330d87223dfcf3c9da40aa5ece8c21141f0d6..47dfe4c829efc6c8c9529e74a6d48d08ab7b06cb 100644
--- a/src/main/java/org/bukkit/command/SimpleCommandMap.java
+++ b/src/main/java/org/bukkit/command/SimpleCommandMap.java
@@ -30,7 +30,7 @@ public class SimpleCommandMap implements CommandMap {
         setDefaultCommands();
     }
 
-    private void setDefaultCommands() {
+    protected void setDefaultCommands() { // Daydream - (feature) Add Daydream Command (private -> protected)
         final ReloadCommand reload = new ReloadCommand("reload");
         this.knownCommands.put("bukkit:reload", reload);
         this.knownCommands.put("bukkit:rl", reload);
diff --git a/src/main/java/org/bukkit/util/permissions/CommandPermissions.java b/src/main/java/org/bukkit/util/permissions/CommandPermissions.java
index 7763d6101ac61900db1e2310966b99584539fd0e..58582a83ba1702c0f2d2832da464f33d579c6210 100644
--- a/src/main/java/org/bukkit/util/permissions/CommandPermissions.java
+++ b/src/main/java/org/bukkit/util/permissions/CommandPermissions.java
@@ -14,10 +14,10 @@ public final class CommandPermissions {
     public static Permission registerPermissions(@NotNull Permission parent) {
         Permission commands = DefaultPermissions.registerPermission(ROOT, "Gives the user the ability to use all CraftBukkit commands", parent);
 
-        DefaultPermissions.registerPermission(PREFIX + "help", "Allows the user to view the vanilla help menu", PermissionDefault.TRUE, commands);
-        DefaultPermissions.registerPermission(PREFIX + "plugins", "Allows the user to view the list of plugins running on this server", PermissionDefault.TRUE, commands);
+        //DefaultPermissions.registerPermission(PREFIX + "help", "Allows the user to view the vanilla help menu", PermissionDefault.TRUE, commands); // Daydream - (feature) Add Daydream Command
+        //DefaultPermissions.registerPermission(PREFIX + "plugins", "Allows the user to view the list of plugins running on this server", PermissionDefault.TRUE, commands); // Daydream - (feature) Add Daydream Command
         DefaultPermissions.registerPermission(PREFIX + "reload", "Allows the user to reload the server settings", PermissionDefault.OP, commands);
-        DefaultPermissions.registerPermission(PREFIX + "version", "Allows the user to view the version of the server", PermissionDefault.TRUE, commands);
+        DefaultPermissions.registerPermission(PREFIX + "version", "Allows the user to view the version of the server", PermissionDefault.OP, commands); // Daydream - (feature) Add Daydream Command
 
         commands.recalculatePermissibles();
         return commands;
