From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 8 Aug 2021 14:54:35 +0900
Subject: [PATCH] Add more reason for PlayerKickEvent


diff --git a/src/main/java/org/bukkit/event/player/PlayerKickEvent.java b/src/main/java/org/bukkit/event/player/PlayerKickEvent.java
index 61f278a9748accd5496811beb4079d8af1895ae1..631fbeb3ed28ab228e3bf54c8d338b82ddd7bce6 100644
--- a/src/main/java/org/bukkit/event/player/PlayerKickEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerKickEvent.java
@@ -21,30 +21,49 @@ public class PlayerKickEvent extends PlayerEvent implements Cancellable {
 
     private boolean cancelled;
 
+    // Daydream start - (feature) Add more reason for PlayerKickEvent
+    private final Object value;
+
+    @ApiStatus.Internal
+    public PlayerKickEvent(@NotNull final Player playerKicked, @NotNull final Component kickReason, @NotNull final Component leaveMessage, @NotNull final Cause cause, @org.jetbrains.annotations.Nullable final Object value) {
+        super(playerKicked);
+        this.kickReason = kickReason;
+        this.leaveMessage = leaveMessage;
+        this.cause = cause;
+        this.value = value;
+    }
+
+    /**
+     * 클라이언트가 서버에 전송한 값을 가져옵니다.
+     *
+     * @return 클라이언트가 서버에 전송한 값이 올바른 값일 경우 {@code null}일 수도 있음
+     */
+    @org.jetbrains.annotations.Nullable
+    public Object getValue() {
+        return value;
+    }
+    // Daydream end - (feature) Add more reason for PlayerKickEvent
+
     @ApiStatus.Internal
     @Deprecated(forRemoval = true)
     public PlayerKickEvent(@NotNull final Player playerKicked, @NotNull final String kickReason, @NotNull final String leaveMessage) {
-        super(playerKicked);
-        this.kickReason = LegacyComponentSerializer.legacySection().deserialize(kickReason);
-        this.leaveMessage = LegacyComponentSerializer.legacySection().deserialize(leaveMessage);
-        this.cause  = Cause.UNKNOWN;
+        // Daydream start - (feature) Add more reason for PlayerKickEvent
+        this(playerKicked,
+            LegacyComponentSerializer.legacySection().deserialize(kickReason),
+            LegacyComponentSerializer.legacySection().deserialize(leaveMessage),
+            Cause.UNKNOWN, null);
+        // Daydream end - (feature) Add more reason for PlayerKickEvent
     }
 
     @ApiStatus.Internal
     @Deprecated(forRemoval = true)
     public PlayerKickEvent(@NotNull final Player playerKicked, @NotNull final Component kickReason, @NotNull final Component leaveMessage) {
-        super(playerKicked);
-        this.kickReason = kickReason;
-        this.leaveMessage = leaveMessage;
-        this.cause = Cause.UNKNOWN;
+        this(playerKicked, kickReason, leaveMessage, Cause.UNKNOWN, null); // Daydream - (feature) Add more reason for PlayerKickEvent
     }
 
     @ApiStatus.Internal
     public PlayerKickEvent(@NotNull final Player playerKicked, @NotNull final Component kickReason, @NotNull final Component leaveMessage, @NotNull final Cause cause) {
-        super(playerKicked);
-        this.kickReason = kickReason;
-        this.leaveMessage = leaveMessage;
-        this.cause = cause;
+        this(playerKicked, kickReason, leaveMessage, cause, null); // Daydream - (feature) Add more reason for PlayerKickEvent
     }
 
     /**
@@ -177,6 +196,12 @@ public class PlayerKickEvent extends PlayerEvent implements Cancellable {
         SPAM,
         ILLEGAL_ACTION,
         ILLEGAL_CHARACTERS,
+        // Daydream start - (feature) Add more reason for PlayerKickEvent
+        /**
+         * 클라이언트가 탭 자동완성을 이용하여 서버의 플러그인 목록을 조회하려고 할 때
+         */
+        TAB_COMPLETE_EXPLOIT,
+        // Daydream end - (feature) Add more reason for PlayerKickEvent
         OUT_OF_ORDER_CHAT,
         UNSIGNED_CHAT,
         CHAT_VALIDATION_FAILED,
