From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Sun, 11 Feb 2024 15:17:38 +0900
Subject: [PATCH] Add Plugin Datapack System


diff --git a/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java b/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java
index a399f87e925a72aac91b3066317c3efb61c0421a..7f4d96b56219308f82399513a8cec2db9d99e7ea 100644
--- a/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java
+++ b/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java
@@ -55,6 +55,26 @@ public interface PluginMeta extends Namespaced {
      */
     String getMainClass();
 
+    // Daydream start - (feature) Add Plugin Datapack System
+    /**
+     * 플러그인 내장 데이터팩 로더 클래스 이름을 가져옵니다.
+     *
+     * @return 클래스 이름
+     */
+    default @Nullable String getDatapackLoader() {
+        return null;
+    }
+
+    /**
+     * 플러그인 내장 데이터팩을 가지고 있는지 확인합니다.
+     *
+     * @return 내장 데이터팩 여부
+     */
+    default boolean hasEmbeddedDatapack() {
+        return false;
+    }
+    // Daydream end - (feature) Add Plugin Datapack System
+
     /**
      * Returns the phase of the server startup logic that the plugin should be loaded.
      *
diff --git a/src/main/java/moe/caramel/daydream/plugin/PluginDatapackLoader.java b/src/main/java/moe/caramel/daydream/plugin/PluginDatapackLoader.java
new file mode 100644
index 0000000000000000000000000000000000000000..778946665dfef224f766101e831b254efa3604af
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/plugin/PluginDatapackLoader.java
@@ -0,0 +1,12 @@
+package moe.caramel.daydream.plugin;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.jspecify.annotations.NullMarked;
+
+/**
+ * 서버 데이터팩을 로드할 때 인스턴스가 생성됩니다.
+ */
+@NullMarked
+@ApiStatus.OverrideOnly
+public interface PluginDatapackLoader {
+}
diff --git a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
index c14ce1e7d7a5e30cd29b406a412e47104b529511..07610b1de08d4426f08ea7b00631e1dfcd1fdcd1 100644
--- a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
+++ b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
@@ -1057,6 +1057,31 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
         return libraries;
     }
 
+    // Daydream start - (feature) Add Plugin Datapack System
+    private String datapackLoader;
+    private boolean hasEmbeddedDatapack;
+
+    @Override
+    public @Nullable String getDatapackLoader() {
+        return datapackLoader;
+    }
+
+    @ApiStatus.Internal
+    public void setDatapackLoader(final String datapackLoader) {
+        this.datapackLoader = datapackLoader;
+    }
+
+    @Override
+    public boolean hasEmbeddedDatapack() {
+        return hasEmbeddedDatapack;
+    }
+
+    @ApiStatus.Internal
+    public void setHasEmbeddedDatapack(final boolean hasEmbeddedDatapack) {
+        this.hasEmbeddedDatapack = hasEmbeddedDatapack;
+    }
+    // Daydream end - (feature) Add Plugin Datapack System
+
     /**
      * @return unused
      * @deprecated unused
@@ -1144,6 +1169,16 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
             classLoaderOf = map.get("class-loader-of").toString();
         }
 
+        // Daydream start - (feature) Add Plugin Datapack System
+        if (map.get("datapack-loader") != null) {
+            datapackLoader = map.get("datapack-loader").toString();
+        }
+
+        if (map.get("has-embedded-datapack") != null) {
+            hasEmbeddedDatapack = Boolean.parseBoolean(map.get("has-embedded-datapack").toString());
+        }
+        // Daydream end - (feature) Add Plugin Datapack System
+
         depend = makePluginNameList(map, "depend");
         softDepend = makePluginNameList(map, "softdepend");
         loadBefore = makePluginNameList(map, "loadbefore");
@@ -1340,6 +1375,16 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
             map.put("class-loader-of", classLoaderOf);
         }
 
+        // Daydream start - (feature) Add Plugin Datapack System
+        if (datapackLoader != null) {
+            map.put("datapack-loader", datapackLoader);
+        }
+
+        if (hasEmbeddedDatapack) {
+            map.put("has-embedded-datapack", true);
+        }
+        // Daydream end - (feature) Add Plugin Datapack System
+
         if (prefix != null) {
             map.put("prefix", prefix);
         }
