From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Mon, 19 Feb 2024 00:28:31 +0900
Subject: [PATCH] Add Early Loader


diff --git a/src/main/java/moe/caramel/daydream/EarlyLoader.java b/src/main/java/moe/caramel/daydream/EarlyLoader.java
new file mode 100644
index 0000000000000000000000000000000000000000..cfb3652c6fa4d455aa47c7917a33f09df093db08
--- /dev/null
+++ b/src/main/java/moe/caramel/daydream/EarlyLoader.java
@@ -0,0 +1,36 @@
+package moe.caramel.daydream;
+
+import org.bukkit.UnsafeValues;
+import org.bukkit.inventory.ItemFactory;
+import org.bukkit.inventory.meta.ItemMeta;
+import org.jetbrains.annotations.ApiStatus;
+import org.jspecify.annotations.NullMarked;
+import java.util.ServiceLoader;
+
+/**
+ * Early Loader.
+ */
+@NullMarked
+public interface EarlyLoader {
+
+    /**
+     * Early Loader Instance
+     */
+    EarlyLoader LOADER = ServiceLoader.load(EarlyLoader.class, EarlyLoader.class.getClassLoader())
+        .findFirst().orElseThrow();
+
+    /**
+     * Gets the instance of the item factory (for {@link ItemMeta}).
+     *
+     * @return the item factory
+     * @see ItemFactory
+     */
+    ItemFactory getItemFactory();
+
+    /**
+     * @return the unsafe values instance
+     * @see UnsafeValues
+     */
+    @ApiStatus.Experimental
+    UnsafeValues getUnsafe();
+}
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index c45652b31a876d2bce28584254454e3b25598139..ca343d7055469e6afcff525245463df51121eb9b 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2223,7 +2223,7 @@ public final class Bukkit {
      */
     @NotNull
     public static ItemFactory getItemFactory() {
-        return server.getItemFactory();
+        return (server == null) ? moe.caramel.daydream.EarlyLoader.LOADER.getItemFactory() : server.getItemFactory(); // Daydream - (feature) Fix errors that occur when early load
     }
 
     /**
@@ -2751,7 +2751,7 @@ public final class Bukkit {
     @Deprecated(since = "1.7.2")
     @NotNull
     public static UnsafeValues getUnsafe() {
-        return server.getUnsafe();
+        return (server == null) ? moe.caramel.daydream.EarlyLoader.LOADER.getUnsafe() : server.getUnsafe(); // Daydream - (feature) Fix errors that occur when early load
     }
 
 
diff --git a/src/main/java/org/bukkit/inventory/ItemStack.java b/src/main/java/org/bukkit/inventory/ItemStack.java
index 7d9a2f609bb43feb608cee68ffd4549c5b957d60..13a5f2ef94754f92ff94142c1b5d4eb99b2fa608 100644
--- a/src/main/java/org/bukkit/inventory/ItemStack.java
+++ b/src/main/java/org/bukkit/inventory/ItemStack.java
@@ -774,7 +774,7 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, Translat
      */
     @NotNull
     public ItemStack enchantWithLevels(final int levels, final boolean allowTreasure, final @NotNull java.util.Random random) {
-        return Bukkit.getServer().getItemFactory().enchantWithLevels(this, levels, allowTreasure, random);
+        return Bukkit.getItemFactory().enchantWithLevels(this, levels, allowTreasure, random); // Daydream - (feature) Fix errors that occur when early load
     }
 
     /**
@@ -804,7 +804,7 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, Translat
     @NotNull
     @Override
     public net.kyori.adventure.text.event.HoverEvent<net.kyori.adventure.text.event.HoverEvent.ShowItem> asHoverEvent(final @NotNull java.util.function.UnaryOperator<net.kyori.adventure.text.event.HoverEvent.ShowItem> op) {
-        return org.bukkit.Bukkit.getServer().getItemFactory().asHoverEvent(this, op);
+        return org.bukkit.Bukkit.getItemFactory().asHoverEvent(this, op); // Daydream - (feature) Fix errors that occur when early load
     }
 
     /**
@@ -815,7 +815,7 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, Translat
      * @return display name of the {@link ItemStack}
      */
     public net.kyori.adventure.text.@NotNull Component displayName() {
-        return Bukkit.getServer().getItemFactory().displayName(this);
+        return Bukkit.getItemFactory().displayName(this); // Daydream - (feature) Fix errors that occur when early load
     }
 
     /**
@@ -839,7 +839,7 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, Translat
      */
     @NotNull
     public ItemStack ensureServerConversions() {
-        return Bukkit.getServer().getItemFactory().ensureServerConversions(this);
+        return Bukkit.getItemFactory().ensureServerConversions(this); // Daydream - (feature) Fix errors that occur when early load
     }
 
     /**
@@ -963,7 +963,7 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, Translat
     @Nullable
     @Deprecated
     public String getI18NDisplayName() {
-        return Bukkit.getServer().getItemFactory().getI18NDisplayName(this);
+        return Bukkit.getItemFactory().getI18NDisplayName(this); // Daydream - (feature) Fix errors that occur when early load
     }
 
     /**
