--- a/src/main/java/org/bukkit/event/player/PlayerChangedMainHandEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerChangedMainHandEvent.java
@@ -17,7 +_,7 @@
 
     private static final HandlerList HANDLER_LIST = new HandlerList();
 
-    private final MainHand newMainHand;
+    private MainHand newMainHand; // Daydream - Settable Player Arm (remove final)
 
     @ApiStatus.Internal
     public PlayerChangedMainHandEvent(@NotNull Player player, @NotNull MainHand newMainHand) {
@@ -49,6 +_,17 @@
     public MainHand getNewMainHand() {
         return this.newMainHand;
     }
+
+    // Daydream start - Settable Player Arm
+    /**
+     * 플레이어의 주로 사용하는 손을 변경합니다.
+     *
+     * @param mainHand 주로 사용하는 손
+     */
+    public void setNewMainHand(final @NotNull MainHand mainHand) {
+        this.newMainHand = mainHand;
+    }
+    // Daydream end - Settable Player Arm
 
     @NotNull
     @Override
