--- a/src/main/java/org/bukkit/plugin/java/LibraryLoader.java
+++ b/src/main/java/org/bukkit/plugin/java/LibraryLoader.java
@@ -50,7 +_,7 @@
     public static java.util.function.Function<List<java.nio.file.Path>, List<java.nio.file.Path>> REMAPPER; // Paper - remap libraries
 
     private static List<RemoteRepository> getRepositories() {
-        return List.of(new RemoteRepository.Builder("central", "default", MavenLibraryResolver.MAVEN_CENTRAL_DEFAULT_MIRROR).build());
+        return List.of(new RemoteRepository.Builder("central", "default", MavenLibraryResolver.MAVEN_CENTRAL_DEFAULT_MIRROR).build(), new RemoteRepository.Builder("caramel-repo", "default", "https://repo.caramel.moe/repository/maven-public/").build()); // Daydream - Add caramel.moe repository to LibraryLoader
     }
 
     public LibraryLoader(@NotNull Logger logger) {
@@ -146,10 +_,12 @@
             }
 
             jarFiles.add(url);
+            if (!org.bukkit.Bukkit.getServerConfig().shouldIgnoreLibraryLoaderLog()) { // Daydream start - Suppress spam logs in LibraryLoader
             logger.log(Level.INFO, "[{0}] Loaded library {1}", new Object[]
                 {
                     java.util.Objects.requireNonNullElseGet(desc.getPrefix(), desc::getName), file // Paper - use configured log prefix
                 });
+            } // Daydream end - Suppress spam logs in LibraryLoader
         }
 
         // Paper start - rewrite reflection in libraries
